{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# *Dandelion* class\n",
    "\n",
    "![dandelion_logo](img/dandelion_logo_illustration.png)\n",
    "\n",
    "Much of the functions and utility of the `dandelion` package revolves around the `Dandelion` class object. The class will act as an intermediary object for storage and flexible interaction with other tools. This section will run through a quick primer to the `Dandelion` class."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***Import modules***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dandelion==0.1.3.post2.dev62 pandas==1.2.3 numpy==1.20.1 matplotlib==3.3.4 networkx==2.5 scipy==1.6.1 skbio==0.5.6\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "os.chdir(os.path.expanduser('/Users/kt16/Downloads/dandelion_tutorial/'))\n",
    "import dandelion as ddl\n",
    "ddl.logging.print_versions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dandelion class object with n_obs = 562 and n_contigs = 1130\n",
       "    data: 'sequence_id', 'sequence', 'rev_comp', 'productive', 'v_call', 'd_call', 'j_call', 'sequence_alignment', 'germline_alignment', 'junction', 'junction_aa', 'v_cigar', 'd_cigar', 'j_cigar', 'stop_codon', 'vj_in_frame', 'locus', 'junction_length', 'np1_length', 'np2_length', 'v_sequence_start', 'v_sequence_end', 'v_germline_start', 'v_germline_end', 'd_sequence_start', 'd_sequence_end', 'd_germline_start', 'd_germline_end', 'j_sequence_start', 'j_sequence_end', 'j_germline_start', 'j_germline_end', 'v_score', 'v_identity', 'v_support', 'd_score', 'd_identity', 'd_support', 'j_score', 'j_identity', 'j_support', 'fwr1', 'fwr2', 'fwr3', 'fwr4', 'cdr1', 'cdr2', 'cdr3', 'cell_id', 'c_call', 'consensus_count', 'umi_count', 'v_call_10x', 'd_call_10x', 'j_call_10x', 'junction_10x', 'junction_10x_aa', 'v_call_genotyped', 'germline_alignment_d_mask', 'sample_id', 'c_sequence_alignment', 'c_germline_alignment', 'c_sequence_start', 'c_sequence_end', 'c_score', 'c_identity', 'c_support', 'c_call_10x', 'junction_aa_length', 'fwr1_aa', 'fwr2_aa', 'fwr3_aa', 'fwr4_aa', 'cdr1_aa', 'cdr2_aa', 'cdr3_aa', 'sequence_alignment_aa', 'v_sequence_alignment_aa', 'd_sequence_alignment_aa', 'j_sequence_alignment_aa', 'mu_count', 'duplicate_count', 'clone_id', 'changeo_clone_id'\n",
       "    metadata: 'clone_id', 'clone_id_by_size', 'sample_id', 'locus_VDJ', 'locus_VJ', 'productive_VDJ', 'productive_VJ', 'v_call_genotyped_VDJ', 'v_call_genotyped_VJ', 'j_call_VDJ', 'j_call_VJ', 'c_call_VDJ', 'c_call_VJ', 'duplicate_count_VDJ_0', 'duplicate_count_VDJ_1', 'duplicate_count_VDJ_2', 'duplicate_count_VJ_0', 'duplicate_count_VJ_1', 'junction_aa_VDJ', 'junction_aa_VJ', 'status', 'status_summary', 'productive', 'productive_summary', 'isotype', 'isotype_summary', 'vdj_status', 'vdj_status_summary', 'constant_status_summary', 'changeo_clone_id'\n",
       "    distance: 'VDJ_0', 'VDJ_1', 'VDJ_2', 'VJ_0', 'VJ_1'\n",
       "    edges: 'source', 'target', 'weight'\n",
       "    layout: layout for 562 vertices, layout for 29 vertices\n",
       "    graph: networkx graph of 562 vertices, networkx graph of 29 vertices "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vdj = ddl.read_h5('dandelion_results.h5')\n",
    "vdj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Basically, the object can be summarized in the following illustration:\n",
    "\n",
    "![dandelion_class <](img/dandelion_class.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Essentially, the `.data` slot holds the AIRR contig table while the `.metadata` holds a collapsed version that is compatible with combining with `AnnData`'s `.obs` slot. You can retrieve these slots like a typical class object; for example, if I want the metadata:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone_id</th>\n",
       "      <th>clone_id_by_size</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>locus_VDJ</th>\n",
       "      <th>locus_VJ</th>\n",
       "      <th>productive_VDJ</th>\n",
       "      <th>productive_VJ</th>\n",
       "      <th>v_call_genotyped_VDJ</th>\n",
       "      <th>v_call_genotyped_VJ</th>\n",
       "      <th>j_call_VDJ</th>\n",
       "      <th>...</th>\n",
       "      <th>status</th>\n",
       "      <th>status_summary</th>\n",
       "      <th>productive</th>\n",
       "      <th>productive_summary</th>\n",
       "      <th>isotype</th>\n",
       "      <th>isotype_summary</th>\n",
       "      <th>vdj_status</th>\n",
       "      <th>vdj_status_summary</th>\n",
       "      <th>constant_status_summary</th>\n",
       "      <th>changeo_clone_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG</th>\n",
       "      <td>131_4_1_58</td>\n",
       "      <td>2</td>\n",
       "      <td>sc5p_v2_hs_PBMC_1k</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGK</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV4-59</td>\n",
       "      <td>IGKV4-1</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>8_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC</th>\n",
       "      <td>17_1_1_163</td>\n",
       "      <td>4</td>\n",
       "      <td>sc5p_v2_hs_PBMC_1k</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV3-15</td>\n",
       "      <td>IGLV2-23</td>\n",
       "      <td>IGHJ3</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>63_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG</th>\n",
       "      <td>146_2_1_38</td>\n",
       "      <td>8</td>\n",
       "      <td>sc5p_v2_hs_PBMC_1k</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV4-39</td>\n",
       "      <td>IGLV3-21</td>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgA</td>\n",
       "      <td>IgA</td>\n",
       "      <td>Single + Multi_VJ_j</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>67_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG</th>\n",
       "      <td>22_2_1_55</td>\n",
       "      <td>11</td>\n",
       "      <td>sc5p_v2_hs_PBMC_1k</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV1-3</td>\n",
       "      <td>IGLV1-51</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_j</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>72_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA</th>\n",
       "      <td>59_3_2_308</td>\n",
       "      <td>10</td>\n",
       "      <td>sc5p_v2_hs_PBMC_1k</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGK</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV4-59</td>\n",
       "      <td>IGKV3-15</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>57_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT</th>\n",
       "      <td>116_2_1_191</td>\n",
       "      <td>353</td>\n",
       "      <td>vdj_nextgem_hs_pbmc3</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGK</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV3-15</td>\n",
       "      <td>IGKV2-28|IGKV2D-28</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_v</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>294_537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA</th>\n",
       "      <td>40_5_2_289</td>\n",
       "      <td>382</td>\n",
       "      <td>vdj_nextgem_hs_pbmc3</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV1-2</td>\n",
       "      <td>IGLV2-14</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_j</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>465_538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA</th>\n",
       "      <td>71_7_1_249</td>\n",
       "      <td>481</td>\n",
       "      <td>vdj_nextgem_hs_pbmc3</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGK</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV3-15</td>\n",
       "      <td>IGKV1D-39|IGKV1-39</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_v</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>400_539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA</th>\n",
       "      <td>117_3_2_56</td>\n",
       "      <td>463</td>\n",
       "      <td>vdj_nextgem_hs_pbmc3</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV3-7</td>\n",
       "      <td>IGLV1-51</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_j</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>292_540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG</th>\n",
       "      <td>34_1_1_127</td>\n",
       "      <td>402</td>\n",
       "      <td>vdj_nextgem_hs_pbmc3</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV2-5</td>\n",
       "      <td>IGLV1-47</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_j</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>488_541</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>562 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          clone_id clone_id_by_size  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG     131_4_1_58                2   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC     17_1_1_163                4   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG     146_2_1_38                8   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG      22_2_1_55               11   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA     59_3_2_308               10   \n",
       "...                                            ...              ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT  116_2_1_191              353   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA   40_5_2_289              382   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA   71_7_1_249              481   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA   117_3_2_56              463   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG   34_1_1_127              402   \n",
       "\n",
       "                                                  sample_id locus_VDJ  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG      sc5p_v2_hs_PBMC_1k       IGH   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC      sc5p_v2_hs_PBMC_1k       IGH   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG      sc5p_v2_hs_PBMC_1k       IGH   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG      sc5p_v2_hs_PBMC_1k       IGH   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA      sc5p_v2_hs_PBMC_1k       IGH   \n",
       "...                                                     ...       ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT  vdj_nextgem_hs_pbmc3       IGH   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA  vdj_nextgem_hs_pbmc3       IGH   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA  vdj_nextgem_hs_pbmc3       IGH   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA  vdj_nextgem_hs_pbmc3       IGH   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG  vdj_nextgem_hs_pbmc3       IGH   \n",
       "\n",
       "                                      locus_VJ productive_VDJ productive_VJ  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG        IGK           True          True   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC        IGL           True          True   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG        IGL           True          True   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG        IGL           True          True   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA        IGK           True          True   \n",
       "...                                        ...            ...           ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT      IGK           True          True   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA      IGL           True          True   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA      IGK           True          True   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA      IGL           True          True   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG      IGL           True          True   \n",
       "\n",
       "                                      v_call_genotyped_VDJ  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG               IGHV4-59   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC               IGHV3-15   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG               IGHV4-39   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG                IGHV1-3   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA               IGHV4-59   \n",
       "...                                                    ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT             IGHV3-15   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA              IGHV1-2   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA             IGHV3-15   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA              IGHV3-7   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG              IGHV2-5   \n",
       "\n",
       "                                      v_call_genotyped_VJ j_call_VDJ  ...  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG               IGKV4-1      IGHJ6  ...   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC              IGLV2-23      IGHJ3  ...   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG              IGLV3-21      IGHJ5  ...   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG              IGLV1-51      IGHJ6  ...   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA              IGKV3-15      IGHJ4  ...   \n",
       "...                                                   ...        ...  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT  IGKV2-28|IGKV2D-28      IGHJ6  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA            IGLV2-14      IGHJ4  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA  IGKV1D-39|IGKV1-39      IGHJ4  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA            IGLV1-51      IGHJ6  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG            IGLV1-47      IGHJ4  ...   \n",
       "\n",
       "                                          status status_summary   productive  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG    IGH + IGK      IGH + IGK  True + True   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC    IGH + IGL      IGH + IGL  True + True   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG    IGH + IGL      IGH + IGL  True + True   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG    IGH + IGL      IGH + IGL  True + True   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA    IGH + IGK      IGH + IGK  True + True   \n",
       "...                                          ...            ...          ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT  IGH + IGK      IGH + IGK  True + True   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA  IGH + IGL      IGH + IGL  True + True   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA  IGH + IGK      IGH + IGK  True + True   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA  IGH + IGL      IGH + IGL  True + True   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG  IGH + IGL      IGH + IGL  True + True   \n",
       "\n",
       "                                       productive_summary  isotype  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG           True + True      IgM   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC           True + True      IgM   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG           True + True      IgA   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG           True + True      IgM   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA           True + True      IgM   \n",
       "...                                                   ...      ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT         True + True      IgM   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA         True + True      IgM   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA         True + True      IgM   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA         True + True      IgM   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG         True + True      IgM   \n",
       "\n",
       "                                       isotype_summary           vdj_status  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG                IgM      Single + Single   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC                IgM      Single + Single   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG                IgA  Single + Multi_VJ_j   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG                IgM  Single + Multi_VJ_j   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA                IgM      Single + Single   \n",
       "...                                                ...                  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT              IgM  Single + Multi_VJ_v   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA              IgM  Single + Multi_VJ_j   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA              IgM  Single + Multi_VJ_v   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA              IgM  Single + Multi_VJ_j   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG              IgM  Single + Multi_VJ_j   \n",
       "\n",
       "                                       vdj_status_summary  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG                Single   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC                Single   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG                Single   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG                Single   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA                Single   \n",
       "...                                                   ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT              Single   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA              Single   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA              Single   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA              Single   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG              Single   \n",
       "\n",
       "                                      constant_status_summary changeo_clone_id  \n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG                    Single              8_0  \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC                    Single             63_1  \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG                    Single             67_2  \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG                    Single             72_3  \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA                    Single             57_4  \n",
       "...                                                       ...              ...  \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT                  Single          294_537  \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA                  Single          465_538  \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA                  Single          400_539  \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA                  Single          292_540  \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG                  Single          488_541  \n",
       "\n",
       "[562 rows x 30 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vdj.metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### copy\n",
    "\n",
    "You can deep copy the `Dandelion` object to another variable which will inherit all slots:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone_id</th>\n",
       "      <th>clone_id_by_size</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>locus_VDJ</th>\n",
       "      <th>locus_VJ</th>\n",
       "      <th>productive_VDJ</th>\n",
       "      <th>productive_VJ</th>\n",
       "      <th>v_call_genotyped_VDJ</th>\n",
       "      <th>v_call_genotyped_VJ</th>\n",
       "      <th>j_call_VDJ</th>\n",
       "      <th>...</th>\n",
       "      <th>status</th>\n",
       "      <th>status_summary</th>\n",
       "      <th>productive</th>\n",
       "      <th>productive_summary</th>\n",
       "      <th>isotype</th>\n",
       "      <th>isotype_summary</th>\n",
       "      <th>vdj_status</th>\n",
       "      <th>vdj_status_summary</th>\n",
       "      <th>constant_status_summary</th>\n",
       "      <th>changeo_clone_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG</th>\n",
       "      <td>131_4_1_58</td>\n",
       "      <td>2</td>\n",
       "      <td>sc5p_v2_hs_PBMC_1k</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGK</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV4-59</td>\n",
       "      <td>IGKV4-1</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>8_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC</th>\n",
       "      <td>17_1_1_163</td>\n",
       "      <td>4</td>\n",
       "      <td>sc5p_v2_hs_PBMC_1k</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV3-15</td>\n",
       "      <td>IGLV2-23</td>\n",
       "      <td>IGHJ3</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>63_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG</th>\n",
       "      <td>146_2_1_38</td>\n",
       "      <td>8</td>\n",
       "      <td>sc5p_v2_hs_PBMC_1k</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV4-39</td>\n",
       "      <td>IGLV3-21</td>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgA</td>\n",
       "      <td>IgA</td>\n",
       "      <td>Single + Multi_VJ_j</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>67_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG</th>\n",
       "      <td>22_2_1_55</td>\n",
       "      <td>11</td>\n",
       "      <td>sc5p_v2_hs_PBMC_1k</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV1-3</td>\n",
       "      <td>IGLV1-51</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_j</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>72_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA</th>\n",
       "      <td>59_3_2_308</td>\n",
       "      <td>10</td>\n",
       "      <td>sc5p_v2_hs_PBMC_1k</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGK</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV4-59</td>\n",
       "      <td>IGKV3-15</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>57_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT</th>\n",
       "      <td>116_2_1_191</td>\n",
       "      <td>353</td>\n",
       "      <td>vdj_nextgem_hs_pbmc3</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGK</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV3-15</td>\n",
       "      <td>IGKV2-28|IGKV2D-28</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_v</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>294_537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA</th>\n",
       "      <td>40_5_2_289</td>\n",
       "      <td>382</td>\n",
       "      <td>vdj_nextgem_hs_pbmc3</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV1-2</td>\n",
       "      <td>IGLV2-14</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_j</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>465_538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA</th>\n",
       "      <td>71_7_1_249</td>\n",
       "      <td>481</td>\n",
       "      <td>vdj_nextgem_hs_pbmc3</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGK</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV3-15</td>\n",
       "      <td>IGKV1D-39|IGKV1-39</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>IGH + IGK</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_v</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>400_539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA</th>\n",
       "      <td>117_3_2_56</td>\n",
       "      <td>463</td>\n",
       "      <td>vdj_nextgem_hs_pbmc3</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV3-7</td>\n",
       "      <td>IGLV1-51</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_j</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>292_540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG</th>\n",
       "      <td>34_1_1_127</td>\n",
       "      <td>402</td>\n",
       "      <td>vdj_nextgem_hs_pbmc3</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGL</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>IGHV2-5</td>\n",
       "      <td>IGLV1-47</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>...</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>IGH + IGL</td>\n",
       "      <td>True + True</td>\n",
       "      <td>True + True</td>\n",
       "      <td>IgM</td>\n",
       "      <td>IgM</td>\n",
       "      <td>Single + Multi_VJ_j</td>\n",
       "      <td>Single</td>\n",
       "      <td>Single</td>\n",
       "      <td>488_541</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>562 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          clone_id clone_id_by_size  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG     131_4_1_58                2   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC     17_1_1_163                4   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG     146_2_1_38                8   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG      22_2_1_55               11   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA     59_3_2_308               10   \n",
       "...                                            ...              ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT  116_2_1_191              353   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA   40_5_2_289              382   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA   71_7_1_249              481   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA   117_3_2_56              463   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG   34_1_1_127              402   \n",
       "\n",
       "                                                  sample_id locus_VDJ  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG      sc5p_v2_hs_PBMC_1k       IGH   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC      sc5p_v2_hs_PBMC_1k       IGH   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG      sc5p_v2_hs_PBMC_1k       IGH   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG      sc5p_v2_hs_PBMC_1k       IGH   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA      sc5p_v2_hs_PBMC_1k       IGH   \n",
       "...                                                     ...       ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT  vdj_nextgem_hs_pbmc3       IGH   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA  vdj_nextgem_hs_pbmc3       IGH   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA  vdj_nextgem_hs_pbmc3       IGH   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA  vdj_nextgem_hs_pbmc3       IGH   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG  vdj_nextgem_hs_pbmc3       IGH   \n",
       "\n",
       "                                      locus_VJ productive_VDJ productive_VJ  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG        IGK           True          True   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC        IGL           True          True   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG        IGL           True          True   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG        IGL           True          True   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA        IGK           True          True   \n",
       "...                                        ...            ...           ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT      IGK           True          True   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA      IGL           True          True   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA      IGK           True          True   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA      IGL           True          True   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG      IGL           True          True   \n",
       "\n",
       "                                      v_call_genotyped_VDJ  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG               IGHV4-59   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC               IGHV3-15   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG               IGHV4-39   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG                IGHV1-3   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA               IGHV4-59   \n",
       "...                                                    ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT             IGHV3-15   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA              IGHV1-2   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA             IGHV3-15   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA              IGHV3-7   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG              IGHV2-5   \n",
       "\n",
       "                                      v_call_genotyped_VJ j_call_VDJ  ...  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG               IGKV4-1      IGHJ6  ...   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC              IGLV2-23      IGHJ3  ...   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG              IGLV3-21      IGHJ5  ...   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG              IGLV1-51      IGHJ6  ...   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA              IGKV3-15      IGHJ4  ...   \n",
       "...                                                   ...        ...  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT  IGKV2-28|IGKV2D-28      IGHJ6  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA            IGLV2-14      IGHJ4  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA  IGKV1D-39|IGKV1-39      IGHJ4  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA            IGLV1-51      IGHJ6  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG            IGLV1-47      IGHJ4  ...   \n",
       "\n",
       "                                          status status_summary   productive  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG    IGH + IGK      IGH + IGK  True + True   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC    IGH + IGL      IGH + IGL  True + True   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG    IGH + IGL      IGH + IGL  True + True   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG    IGH + IGL      IGH + IGL  True + True   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA    IGH + IGK      IGH + IGK  True + True   \n",
       "...                                          ...            ...          ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT  IGH + IGK      IGH + IGK  True + True   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA  IGH + IGL      IGH + IGL  True + True   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA  IGH + IGK      IGH + IGK  True + True   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA  IGH + IGL      IGH + IGL  True + True   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG  IGH + IGL      IGH + IGL  True + True   \n",
       "\n",
       "                                       productive_summary  isotype  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG           True + True      IgM   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC           True + True      IgM   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG           True + True      IgA   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG           True + True      IgM   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA           True + True      IgM   \n",
       "...                                                   ...      ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT         True + True      IgM   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA         True + True      IgM   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA         True + True      IgM   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA         True + True      IgM   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG         True + True      IgM   \n",
       "\n",
       "                                       isotype_summary           vdj_status  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG                IgM      Single + Single   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC                IgM      Single + Single   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG                IgA  Single + Multi_VJ_j   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG                IgM  Single + Multi_VJ_j   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA                IgM      Single + Single   \n",
       "...                                                ...                  ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT              IgM  Single + Multi_VJ_v   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA              IgM  Single + Multi_VJ_j   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA              IgM  Single + Multi_VJ_v   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA              IgM  Single + Multi_VJ_j   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG              IgM  Single + Multi_VJ_j   \n",
       "\n",
       "                                       vdj_status_summary  \\\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG                Single   \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC                Single   \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG                Single   \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG                Single   \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA                Single   \n",
       "...                                                   ...   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT              Single   \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA              Single   \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA              Single   \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA              Single   \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG              Single   \n",
       "\n",
       "                                      constant_status_summary changeo_clone_id  \n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG                    Single              8_0  \n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC                    Single             63_1  \n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG                    Single             67_2  \n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG                    Single             72_3  \n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA                    Single             57_4  \n",
       "...                                                       ...              ...  \n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT                  Single          294_537  \n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA                  Single          465_538  \n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA                  Single          400_539  \n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA                  Single          292_540  \n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG                  Single          488_541  \n",
       "\n",
       "[562 rows x 30 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vdj2 = vdj.copy()\n",
    "vdj2.metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retrieving entries with `update_metadata`\n",
    "\n",
    "The `.metadata` slot in Dandelion class automatically initializes whenever the `.data` slot is filled. However, it only returns a standard number of columns that are pre-specified. To retrieve other columns from the `.data` slot, we can update the metadata with `ddl.update_metadata` and specify the option `retrieve`. \n",
    "\n",
    "The following options determine how the retrieval is completed:\n",
    "\n",
    "`split` - splits the retrieval into heavy and light chains calls.\n",
    "\n",
    "`split_locus` - smiliar to `split` but splits the retrieval to `IGH/IGK/IGL`.\n",
    "\n",
    "`collapse` - Adds a `|` to separate every element.\n",
    "\n",
    "`combine` - similar to `collapse` but only retains unique elements (separated by a `|` if multiple are found)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***Example 1 : retrieving junction amino acid sequences***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dandelion class object with n_obs = 562 and n_contigs = 1130\n",
       "    data: 'sequence_id', 'sequence', 'rev_comp', 'productive', 'v_call', 'd_call', 'j_call', 'sequence_alignment', 'germline_alignment', 'junction', 'junction_aa', 'v_cigar', 'd_cigar', 'j_cigar', 'stop_codon', 'vj_in_frame', 'locus', 'junction_length', 'np1_length', 'np2_length', 'v_sequence_start', 'v_sequence_end', 'v_germline_start', 'v_germline_end', 'd_sequence_start', 'd_sequence_end', 'd_germline_start', 'd_germline_end', 'j_sequence_start', 'j_sequence_end', 'j_germline_start', 'j_germline_end', 'v_score', 'v_identity', 'v_support', 'd_score', 'd_identity', 'd_support', 'j_score', 'j_identity', 'j_support', 'fwr1', 'fwr2', 'fwr3', 'fwr4', 'cdr1', 'cdr2', 'cdr3', 'cell_id', 'c_call', 'consensus_count', 'umi_count', 'v_call_10x', 'd_call_10x', 'j_call_10x', 'junction_10x', 'junction_10x_aa', 'v_call_genotyped', 'germline_alignment_d_mask', 'sample_id', 'c_sequence_alignment', 'c_germline_alignment', 'c_sequence_start', 'c_sequence_end', 'c_score', 'c_identity', 'c_support', 'c_call_10x', 'junction_aa_length', 'fwr1_aa', 'fwr2_aa', 'fwr3_aa', 'fwr4_aa', 'cdr1_aa', 'cdr2_aa', 'cdr3_aa', 'sequence_alignment_aa', 'v_sequence_alignment_aa', 'd_sequence_alignment_aa', 'j_sequence_alignment_aa', 'mu_count', 'duplicate_count', 'clone_id', 'changeo_clone_id'\n",
       "    metadata: 'clone_id', 'clone_id_by_size', 'sample_id', 'locus_VDJ', 'locus_VJ', 'productive_VDJ', 'productive_VJ', 'v_call_genotyped_VDJ', 'v_call_genotyped_VJ', 'j_call_VDJ', 'j_call_VJ', 'c_call_VDJ', 'c_call_VJ', 'duplicate_count_VDJ_0', 'duplicate_count_VDJ_1', 'duplicate_count_VDJ_2', 'duplicate_count_VJ_0', 'duplicate_count_VJ_1', 'junction_aa_VDJ', 'junction_aa_VJ', 'status', 'status_summary', 'productive', 'productive_summary', 'isotype', 'isotype_summary', 'vdj_status', 'vdj_status_summary', 'constant_status_summary', 'changeo_clone_id', 'd_call_VDJ', 'd_call_VJ'\n",
       "    distance: 'VDJ_0', 'VDJ_1', 'VDJ_2', 'VJ_0', 'VJ_1'\n",
       "    edges: 'source', 'target', 'weight'\n",
       "    layout: layout for 562 vertices, layout for 29 vertices\n",
       "    graph: networkx graph of 562 vertices, networkx graph of 29 vertices "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ddl.update_metadata(vdj, retrieve = 'd_call')\n",
    "vdj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note the additional `d_call` heavy and light columns in the metadata slot."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By default, `dandelion` will not try to merge numerical columns as it can create mixed dtype columns."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***Example 2 : editing clone_id column***\n",
    "\n",
    "Perhaps you want to have a bit more control with how clones are called. We can edit this directly from the `.data` slot and retrieve accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clone_id</th>\n",
       "      <th>clone_id_heavy_only</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG</th>\n",
       "      <td>131_4_1_58</td>\n",
       "      <td>131_4_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC</th>\n",
       "      <td>17_1_1_163</td>\n",
       "      <td>17_1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG</th>\n",
       "      <td>146_2_1_38</td>\n",
       "      <td>146_2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG</th>\n",
       "      <td>22_2_1_55</td>\n",
       "      <td>22_2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA</th>\n",
       "      <td>59_3_2_308</td>\n",
       "      <td>59_3_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT</th>\n",
       "      <td>116_2_1_191</td>\n",
       "      <td>116_2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA</th>\n",
       "      <td>40_5_2_289</td>\n",
       "      <td>40_5_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA</th>\n",
       "      <td>71_7_1_249</td>\n",
       "      <td>71_7_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA</th>\n",
       "      <td>117_3_2_56</td>\n",
       "      <td>117_3_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG</th>\n",
       "      <td>34_1_1_127</td>\n",
       "      <td>34_1_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>562 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          clone_id clone_id_heavy_only\n",
       "sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG     131_4_1_58             131_4_1\n",
       "sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC     17_1_1_163              17_1_1\n",
       "sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG     146_2_1_38             146_2_1\n",
       "sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG      22_2_1_55              22_2_1\n",
       "sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA     59_3_2_308              59_3_2\n",
       "...                                            ...                 ...\n",
       "vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT  116_2_1_191             116_2_1\n",
       "vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA   40_5_2_289              40_5_2\n",
       "vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA   71_7_1_249              71_7_1\n",
       "vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA   117_3_2_56             117_3_2\n",
       "vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG   34_1_1_127              34_1_1\n",
       "\n",
       "[562 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# if we only want to keep the light chain clone assignment \n",
    "clones = []\n",
    "for clone in vdj.data['clone_id']:\n",
    "    if '|' in clone: # this is because clones were merged into the the same column if they have different pairing of BCR combinations\n",
    "        clone_list = clone.split('|')\n",
    "        clones.append('|'.join(list(set([clone_2.rsplit('_', 1)[0] if clone_2.count('_') == 3 else clone_2 for clone_2 in clone_list]))))\n",
    "    else:\n",
    "        if clone.count('_') == 3: # this means it's looking for X_X_X_X, 3 underscores\n",
    "            clones.append(clone.rsplit('_', 1)[0]) # split the 3rd underscore but only keep the first entry\n",
    "        else:\n",
    "            clones.append(clone)\n",
    "vdj.data['clone_id_heavy_only'] = clones\n",
    "ddl.update_metadata(vdj, retrieve = 'clone_id_heavy_only', split = False, collapse = True)\n",
    "vdj.metadata[['clone_id', 'clone_id_heavy_only']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `concat`enating multiple objects\n",
    "\n",
    "This is a simple function to concatenate (append) two or more `Dandelion` class, or `pandas` dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dandelion class object with n_obs = 562 and n_contigs = 1130\n",
       "    data: 'sequence_id', 'sequence', 'rev_comp', 'productive', 'v_call', 'd_call', 'j_call', 'sequence_alignment', 'germline_alignment', 'junction', 'junction_aa', 'v_cigar', 'd_cigar', 'j_cigar', 'stop_codon', 'vj_in_frame', 'locus', 'junction_length', 'np1_length', 'np2_length', 'v_sequence_start', 'v_sequence_end', 'v_germline_start', 'v_germline_end', 'd_sequence_start', 'd_sequence_end', 'd_germline_start', 'd_germline_end', 'j_sequence_start', 'j_sequence_end', 'j_germline_start', 'j_germline_end', 'v_score', 'v_identity', 'v_support', 'd_score', 'd_identity', 'd_support', 'j_score', 'j_identity', 'j_support', 'fwr1', 'fwr2', 'fwr3', 'fwr4', 'cdr1', 'cdr2', 'cdr3', 'cell_id', 'c_call', 'consensus_count', 'umi_count', 'v_call_10x', 'd_call_10x', 'j_call_10x', 'junction_10x', 'junction_10x_aa', 'v_call_genotyped', 'germline_alignment_d_mask', 'sample_id', 'c_sequence_alignment', 'c_germline_alignment', 'c_sequence_start', 'c_sequence_end', 'c_score', 'c_identity', 'c_support', 'c_call_10x', 'junction_aa_length', 'fwr1_aa', 'fwr2_aa', 'fwr3_aa', 'fwr4_aa', 'cdr1_aa', 'cdr2_aa', 'cdr3_aa', 'sequence_alignment_aa', 'v_sequence_alignment_aa', 'd_sequence_alignment_aa', 'j_sequence_alignment_aa', 'mu_count', 'duplicate_count', 'clone_id', 'changeo_clone_id', 'clone_id_heavy_only'\n",
       "    metadata: 'clone_id', 'clone_id_by_size', 'sample_id', 'locus_VDJ', 'locus_VJ', 'productive_VDJ', 'productive_VJ', 'v_call_genotyped_VDJ', 'v_call_genotyped_VJ', 'j_call_VDJ', 'j_call_VJ', 'c_call_VDJ', 'c_call_VJ', 'duplicate_count_VDJ_0', 'duplicate_count_VDJ_1', 'duplicate_count_VDJ_2', 'duplicate_count_VJ_0', 'duplicate_count_VJ_1', 'junction_aa_VDJ', 'junction_aa_VJ', 'status', 'status_summary', 'productive', 'productive_summary', 'isotype', 'isotype_summary', 'vdj_status', 'vdj_status_summary', 'constant_status_summary', 'changeo_clone_id', 'd_call_VDJ', 'd_call_VJ', 'clone_id_heavy_only'\n",
       "    distance: 'VDJ_0', 'VDJ_1', 'VDJ_2', 'VJ_0', 'VJ_1'\n",
       "    edges: 'source', 'target', 'weight'\n",
       "    layout: layout for 562 vertices, layout for 29 vertices\n",
       "    graph: networkx graph of 562 vertices, networkx graph of 29 vertices "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for example, the original dandelion class has 838 unique cell barcodes and 1700 contigs\n",
    "vdj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dandelion class object with n_obs = 562 and n_contigs = 3390\n",
       "    data: 'sequence_id', 'sequence', 'rev_comp', 'productive', 'v_call', 'd_call', 'j_call', 'sequence_alignment', 'germline_alignment', 'junction', 'junction_aa', 'v_cigar', 'd_cigar', 'j_cigar', 'stop_codon', 'vj_in_frame', 'locus', 'junction_length', 'np1_length', 'np2_length', 'v_sequence_start', 'v_sequence_end', 'v_germline_start', 'v_germline_end', 'd_sequence_start', 'd_sequence_end', 'd_germline_start', 'd_germline_end', 'j_sequence_start', 'j_sequence_end', 'j_germline_start', 'j_germline_end', 'v_score', 'v_identity', 'v_support', 'd_score', 'd_identity', 'd_support', 'j_score', 'j_identity', 'j_support', 'fwr1', 'fwr2', 'fwr3', 'fwr4', 'cdr1', 'cdr2', 'cdr3', 'cell_id', 'c_call', 'consensus_count', 'umi_count', 'v_call_10x', 'd_call_10x', 'j_call_10x', 'junction_10x', 'junction_10x_aa', 'v_call_genotyped', 'germline_alignment_d_mask', 'sample_id', 'c_sequence_alignment', 'c_germline_alignment', 'c_sequence_start', 'c_sequence_end', 'c_score', 'c_identity', 'c_support', 'c_call_10x', 'junction_aa_length', 'fwr1_aa', 'fwr2_aa', 'fwr3_aa', 'fwr4_aa', 'cdr1_aa', 'cdr2_aa', 'cdr3_aa', 'sequence_alignment_aa', 'v_sequence_alignment_aa', 'd_sequence_alignment_aa', 'j_sequence_alignment_aa', 'mu_count', 'duplicate_count', 'clone_id', 'changeo_clone_id', 'clone_id_heavy_only'\n",
       "    metadata: 'clone_id', 'clone_id_by_size', 'sample_id', 'locus_VDJ', 'locus_VJ', 'productive_VDJ', 'productive_VJ', 'v_call_genotyped_VDJ', 'v_call_genotyped_VJ', 'j_call_VDJ', 'j_call_VJ', 'c_call_VDJ', 'c_call_VJ', 'duplicate_count_VDJ_0', 'duplicate_count_VDJ_1', 'duplicate_count_VDJ_2', 'duplicate_count_VDJ_3', 'duplicate_count_VDJ_4', 'duplicate_count_VDJ_5', 'duplicate_count_VDJ_6', 'duplicate_count_VDJ_7', 'duplicate_count_VDJ_8', 'duplicate_count_VJ_0', 'duplicate_count_VJ_1', 'duplicate_count_VJ_2', 'duplicate_count_VJ_3', 'duplicate_count_VJ_4', 'duplicate_count_VJ_5', 'junction_aa_VDJ', 'junction_aa_VJ', 'status', 'status_summary', 'productive', 'productive_summary', 'isotype', 'isotype_summary', 'vdj_status', 'vdj_status_summary', 'constant_status_summary'\n",
       "    distance: None\n",
       "    edges: None\n",
       "    layout: None\n",
       "    graph: None"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now it has 5100 contigs instead, and the metadata should also be properly populated\n",
    "vdj_concat = ddl.concat([vdj, vdj, vdj])\n",
    "vdj_concat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### read/write\n",
    "\n",
    "`Dandelion` class can be saved using `.write_h5` and `.write_pkl` functions with accompanying compression methods. `write_h5` primarily uses pandas `to_hdf` library and `write_pkl` just uses pickle. `read_h5` and `read_pkl` functions will read the respective file formats accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kt16/miniconda3/envs/dandelion/lib/python3.7/site-packages/pandas/core/generic.py:2621: PerformanceWarning: \n",
      "your performance may suffer as PyTables will pickle object types that it cannot\n",
      "map directly to c-types [inferred_type->mixed,key->block3_values] [items->Index(['sequence_id', 'sequence', 'v_call', 'd_call', 'j_call',\n",
      "       'sequence_alignment', 'germline_alignment', 'junction', 'junction_aa',\n",
      "       'v_cigar', 'd_cigar', 'j_cigar', 'locus', 'np2_length',\n",
      "       'd_sequence_start', 'd_sequence_end', 'd_germline_start',\n",
      "       'd_germline_end', 'd_score', 'd_identity', 'd_support', 'fwr1', 'fwr2',\n",
      "       'fwr3', 'fwr4', 'cdr1', 'cdr2', 'cdr3', 'cell_id', 'c_call',\n",
      "       'v_call_10x', 'd_call_10x', 'j_call_10x', 'junction_10x',\n",
      "       'junction_10x_aa', 'v_call_genotyped', 'germline_alignment_d_mask',\n",
      "       'sample_id', 'c_sequence_alignment', 'c_germline_alignment',\n",
      "       'c_sequence_start', 'c_sequence_end', 'c_score', 'c_identity',\n",
      "       'c_support', 'c_call_10x', 'fwr1_aa', 'fwr2_aa', 'fwr3_aa', 'fwr4_aa',\n",
      "       'cdr1_aa', 'cdr2_aa', 'cdr3_aa', 'sequence_alignment_aa',\n",
      "       'v_sequence_alignment_aa', 'd_sequence_alignment_aa',\n",
      "       'j_sequence_alignment_aa', 'clone_id', 'changeo_clone_id',\n",
      "       'clone_id_heavy_only'],\n",
      "      dtype='object')]\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.05 s, sys: 51.5 ms, total: 1.1 s\n",
      "Wall time: 1.12 s\n"
     ]
    }
   ],
   "source": [
    "%time vdj.write_h5('dandelion_results.h5', complib = 'bzip2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 419 ms, sys: 37.9 ms, total: 457 ms\n",
      "Wall time: 458 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Dandelion class object with n_obs = 562 and n_contigs = 1130\n",
       "    data: 'sequence_id', 'sequence', 'rev_comp', 'productive', 'v_call', 'd_call', 'j_call', 'sequence_alignment', 'germline_alignment', 'junction', 'junction_aa', 'v_cigar', 'd_cigar', 'j_cigar', 'stop_codon', 'vj_in_frame', 'locus', 'junction_length', 'np1_length', 'np2_length', 'v_sequence_start', 'v_sequence_end', 'v_germline_start', 'v_germline_end', 'd_sequence_start', 'd_sequence_end', 'd_germline_start', 'd_germline_end', 'j_sequence_start', 'j_sequence_end', 'j_germline_start', 'j_germline_end', 'v_score', 'v_identity', 'v_support', 'd_score', 'd_identity', 'd_support', 'j_score', 'j_identity', 'j_support', 'fwr1', 'fwr2', 'fwr3', 'fwr4', 'cdr1', 'cdr2', 'cdr3', 'cell_id', 'c_call', 'consensus_count', 'umi_count', 'v_call_10x', 'd_call_10x', 'j_call_10x', 'junction_10x', 'junction_10x_aa', 'v_call_genotyped', 'germline_alignment_d_mask', 'sample_id', 'c_sequence_alignment', 'c_germline_alignment', 'c_sequence_start', 'c_sequence_end', 'c_score', 'c_identity', 'c_support', 'c_call_10x', 'junction_aa_length', 'fwr1_aa', 'fwr2_aa', 'fwr3_aa', 'fwr4_aa', 'cdr1_aa', 'cdr2_aa', 'cdr3_aa', 'sequence_alignment_aa', 'v_sequence_alignment_aa', 'd_sequence_alignment_aa', 'j_sequence_alignment_aa', 'mu_count', 'duplicate_count', 'clone_id', 'changeo_clone_id', 'clone_id_heavy_only'\n",
       "    metadata: 'clone_id', 'clone_id_by_size', 'sample_id', 'locus_VDJ', 'locus_VJ', 'productive_VDJ', 'productive_VJ', 'v_call_genotyped_VDJ', 'v_call_genotyped_VJ', 'j_call_VDJ', 'j_call_VJ', 'c_call_VDJ', 'c_call_VJ', 'duplicate_count_VDJ_0', 'duplicate_count_VDJ_1', 'duplicate_count_VDJ_2', 'duplicate_count_VJ_0', 'duplicate_count_VJ_1', 'junction_aa_VDJ', 'junction_aa_VJ', 'status', 'status_summary', 'productive', 'productive_summary', 'isotype', 'isotype_summary', 'vdj_status', 'vdj_status_summary', 'constant_status_summary', 'changeo_clone_id', 'd_call_VDJ', 'd_call_VJ', 'clone_id_heavy_only'\n",
       "    distance: 'VDJ_0', 'VDJ_1', 'VDJ_2', 'VJ_0', 'VJ_1'\n",
       "    edges: 'source', 'target', 'weight'\n",
       "    layout: layout for 562 vertices, layout for 29 vertices\n",
       "    graph: networkx graph of 562 vertices, networkx graph of 29 vertices "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time vdj_1 = ddl.read_h5('dandelion_results.h5')\n",
    "vdj_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The read/write times using `pickle` can be situationally faster/slower and file sizes can also be situationally smaller/larger (depending on which compression is used)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4.44 s, sys: 30.9 ms, total: 4.47 s\n",
      "Wall time: 4.55 s\n"
     ]
    }
   ],
   "source": [
    "%time vdj.write_pkl('dandelion_results.pkl.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 58.7 ms, sys: 6.65 ms, total: 65.3 ms\n",
      "Wall time: 68.5 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Dandelion class object with n_obs = 562 and n_contigs = 1130\n",
       "    data: 'sequence_id', 'sequence', 'rev_comp', 'productive', 'v_call', 'd_call', 'j_call', 'sequence_alignment', 'germline_alignment', 'junction', 'junction_aa', 'v_cigar', 'd_cigar', 'j_cigar', 'stop_codon', 'vj_in_frame', 'locus', 'junction_length', 'np1_length', 'np2_length', 'v_sequence_start', 'v_sequence_end', 'v_germline_start', 'v_germline_end', 'd_sequence_start', 'd_sequence_end', 'd_germline_start', 'd_germline_end', 'j_sequence_start', 'j_sequence_end', 'j_germline_start', 'j_germline_end', 'v_score', 'v_identity', 'v_support', 'd_score', 'd_identity', 'd_support', 'j_score', 'j_identity', 'j_support', 'fwr1', 'fwr2', 'fwr3', 'fwr4', 'cdr1', 'cdr2', 'cdr3', 'cell_id', 'c_call', 'consensus_count', 'umi_count', 'v_call_10x', 'd_call_10x', 'j_call_10x', 'junction_10x', 'junction_10x_aa', 'v_call_genotyped', 'germline_alignment_d_mask', 'sample_id', 'c_sequence_alignment', 'c_germline_alignment', 'c_sequence_start', 'c_sequence_end', 'c_score', 'c_identity', 'c_support', 'c_call_10x', 'junction_aa_length', 'fwr1_aa', 'fwr2_aa', 'fwr3_aa', 'fwr4_aa', 'cdr1_aa', 'cdr2_aa', 'cdr3_aa', 'sequence_alignment_aa', 'v_sequence_alignment_aa', 'd_sequence_alignment_aa', 'j_sequence_alignment_aa', 'mu_count', 'duplicate_count', 'clone_id', 'changeo_clone_id', 'clone_id_heavy_only'\n",
       "    metadata: 'clone_id', 'clone_id_by_size', 'sample_id', 'locus_VDJ', 'locus_VJ', 'productive_VDJ', 'productive_VJ', 'v_call_genotyped_VDJ', 'v_call_genotyped_VJ', 'j_call_VDJ', 'j_call_VJ', 'c_call_VDJ', 'c_call_VJ', 'duplicate_count_VDJ_0', 'duplicate_count_VDJ_1', 'duplicate_count_VDJ_2', 'duplicate_count_VJ_0', 'duplicate_count_VJ_1', 'junction_aa_VDJ', 'junction_aa_VJ', 'status', 'status_summary', 'productive', 'productive_summary', 'isotype', 'isotype_summary', 'vdj_status', 'vdj_status_summary', 'constant_status_summary', 'changeo_clone_id', 'd_call_VDJ', 'd_call_VJ', 'clone_id_heavy_only'\n",
       "    distance: 'VDJ_0', 'VDJ_1', 'VDJ_2', 'VJ_0', 'VJ_1'\n",
       "    edges: 'source', 'target', 'weight'\n",
       "    layout: layout for 562 vertices, layout for 29 vertices\n",
       "    graph: networkx graph of 562 vertices, networkx graph of 29 vertices "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time vdj_2 = ddl.read_pkl('dandelion_results.pkl.gz')\n",
    "vdj_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (dandelion)",
   "language": "python",
   "name": "dandelion"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
