

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Pre-processing (Re-annotation) &mdash; dandelion  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../_static/dandelion_logo.png"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Filtering" href="2_dandelion_filtering-10x_data.html" />
    <link rel="prev" title="&lt;no title&gt;" href="../tutorial_short.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/dandelion_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../README.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#required-database">Required database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#external-softwares">External softwares</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#basic-requirements">Basic requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#acknowledgements">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#citation">Citation</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="singularity_preprocessing.html">Dandelion preprocessing with Singularity</a></li>
<li class="toctree-l1"><a class="reference internal" href="gamma_delta.html">Dandelion for TCR gamma/delta reannotation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Pre-processing (Re-annotation)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Foreword">Foreword</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-1:-Formatting-the-headers-of-the-Cell-Ranger-fasta-file">Step 1: Formatting the headers of the Cell Ranger fasta file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-2:-Reannotate-the-V/D/J-genes-with-igblastn">Step 2: Reannotate the V/D/J genes with <em>igblastn</em></a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-3-:-Reassigning-heavy-chain-V-gene-alleles-(optional-but-recommended)">Step 3 : Reassigning heavy chain V gene alleles <em>(optional but recommended)</em></a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-4:-Assigning-constant-region-calls">Step 4: Assigning constant region calls</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-5:-Quantify-mutations-(optional).">Step 5: Quantify mutations <em>(optional)</em>.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="2_dandelion_filtering-10x_data.html">Filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="1b_dandelion_noreannotation-10x_data.html">Reading 10X Cell Ranger output directly</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_dandelion_findingclones-10x_data.html">BCR clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="1c_dandelion_scirpy.html">Interoperability with <code class="docutils literal notranslate"><span class="pre">scirpy</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="4_dandelion_visualization-10x_data.html">Visualizing BCR data</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_dandelion_diversity_and_mutation-10x_data.html">Calculating diversity and mutation</a></li>
<li class="toctree-l1"><a class="reference internal" href="0_dandelion_primer.html"><em>Dandelion</em> class</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_dandelion_running_from_R-10x_data.html">Running from R</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_dandelion_TCR_data_10x_data.html">Analyzing TCR data</a></li>
</ul>
<p class="caption"><span class="caption-text">API:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">Preprocessing: <cite>pp</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.preprocessing.external">Preprocessing (external): <cite>pp.external</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.tools">Tools: <cite>tl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.plotting">Plotting: <cite>pl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.utilities">Utilities: <cite>utl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.Dandelion">Dandelion</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.logging">Logging</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">dandelion</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../tutorial_short.html">&lt;no title&gt;</a> &raquo;</li>
        
      <li>Pre-processing (Re-annotation)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/notebooks/1_dandelion_preprocessing-10x_data.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Pre-processing-(Re-annotation)">
<h1>Pre-processing (Re-annotation)<a class="headerlink" href="#Pre-processing-(Re-annotation)" title="Permalink to this headline">¶</a></h1>
<img alt="dandelion_logo" src="../_images/dandelion_logo_illustration.png" />
<div class="section" id="Foreword">
<h2>Foreword<a class="headerlink" href="#Foreword" title="Permalink to this headline">¶</a></h2>
<p><strong>dandelion</strong> is written in <code class="docutils literal notranslate"><span class="pre">python=3.7.6</span></code> and is primarily a single-cell BCR-seq analysis package. It makes use of some tools from the fantastic <a class="reference external" href="https://immcantation.readthedocs.io/">immcantation suite</a>, implementing a workflow to streamline the pre-processing and exploratory stages for analyzing single-cell BCR-seq data from 10X Genomics. Post-processed data from <strong>dandelion</strong> can be smoothly transferred to <a class="reference external" href="https://scanpy.readthedocs.io/">scanpy</a>/AnnData object for integration and
exploration of BCR-seq data and RNA-seq data. I hope to be able to introduce some new single-cell BCR-seq exploratory tools down the road through <em>dandelion</em>.</p>
<p>This section will cover the initial pre-processing of files after 10X’s <code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">Ranger</span> <span class="pre">vdj</span></code> immune profiling data analysis pipeline <strong>manually</strong>. As mentioned, there is now a <a class="reference external" href="https://sc-dandelion.readthedocs.io/en/master/notebooks/singularity_preprocessing.html">singularity container</a> that can automate the first few steps outlined below.</p>
<p>We will download the 10X data sets to process for this tutorial:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># create sub-folders</span>
mkdir -p dandelion_tutorial/vdj_nextgem_hs_pbmc3
mkdir -p dandelion_tutorial/vdj_v1_hs_pbmc3
mkdir -p dandelion_tutorial/sc5p_v2_hs_PBMC_10k
mkdir -p dandelion_tutorial/sc5p_v2_hs_PBMC_1k

<span class="c1"># change into each directory and download the necessary files</span>
<span class="nb">cd</span> dandelion_tutorial/vdj_v1_hs_pbmc3<span class="p">;</span>
wget -O filtered_feature_bc_matrix.h5 https://cf.10xgenomics.com/samples/cell-vdj/3.1.0/vdj_v1_hs_pbmc3/vdj_v1_hs_pbmc3_filtered_feature_bc_matrix.h5<span class="p">;</span>
wget -O filtered_contig_annotations.csv https://cf.10xgenomics.com/samples/cell-vdj/3.1.0/vdj_v1_hs_pbmc3/vdj_v1_hs_pbmc3_b_filtered_contig_annotations.csv<span class="p">;</span>
wget -O filtered_contig.fasta https://cf.10xgenomics.com/samples/cell-vdj/3.1.0/vdj_v1_hs_pbmc3/vdj_v1_hs_pbmc3_b_filtered_contig.fasta<span class="p">;</span>

<span class="nb">cd</span> ../vdj_nextgem_hs_pbmc3
wget -O filtered_feature_bc_matrix.h5 https://cf.10xgenomics.com/samples/cell-vdj/3.1.0/vdj_nextgem_hs_pbmc3/vdj_nextgem_hs_pbmc3_filtered_feature_bc_matrix.h5<span class="p">;</span>
wget -O filtered_contig_annotations.csv https://cf.10xgenomics.com/samples/cell-vdj/3.1.0/vdj_nextgem_hs_pbmc3/vdj_nextgem_hs_pbmc3_b_filtered_contig_annotations.csv<span class="p">;</span>
wget -O filtered_contig.fasta https://cf.10xgenomics.com/samples/cell-vdj/3.1.0/vdj_nextgem_hs_pbmc3/vdj_nextgem_hs_pbmc3_b_filtered_contig.fasta<span class="p">;</span>

<span class="nb">cd</span> ../sc5p_v2_hs_PBMC_10k<span class="p">;</span>
wget -O filtered_feature_bc_matrix.h5 https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_filtered_feature_bc_matrix.h5<span class="p">;</span>
wget -O filtered_contig_annotations.csv https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_b_filtered_contig_annotations.csv<span class="p">;</span>
wget -O filtered_contig.fasta https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_b_filtered_contig.fasta<span class="p">;</span>

<span class="nb">cd</span> ../sc5p_v2_hs_PBMC_1k<span class="p">;</span>
wget -O filtered_feature_bc_matrix.h5 wget https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_PBMC_1k/sc5p_v2_hs_PBMC_1k_filtered_feature_bc_matrix.h5<span class="p">;</span>
wget -O filtered_contig_annotations.csv wget https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_PBMC_1k/sc5p_v2_hs_PBMC_1k_b_filtered_contig_annotations.csv<span class="p">;</span>
wget -O filtered_contig.fasta https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_PBMC_1k/sc5p_v2_hs_PBMC_1k_b_filtered_contig.fasta<span class="p">;</span>
</pre></div>
</div>
<p><strong>dandelion</strong>’s reannotation workflow requires the Cell Ranger fasta files and annotation files to start, particularly either <em>all_contig.fasta</em> or <em>filtered_contig.fasta</em> and corresponding <em>all_contig_annotations.csv</em> and <em>filtered_contig_annotations.csv</em>.</p>
<p>I’m running everything with the <em>filtered_contig</em> files as a standard analysis set up. I’m using a standard laptop for the analysis here: entry level 2017 Macbook Pro with 2.3 GHz Intel Core i5 processor and 16 GB 2133 MHz LPDDR3 ram.</p>
<p>If you followed the installation instructions, you should have the requisite auxillary softwares installed already. Otherwise, you can download them manually: <a class="reference external" href="https://ftp.ncbi.nih.gov/blast/executables/igblast/release/LATEST/">blast+</a> and <a class="reference external" href="https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">igblast</a>. For tigger-genotype, you can download it <a class="reference external" href="https://bitbucket.org/kleinstein/immcantation/src/default/pipelines/">here</a>. Just note that I made some minor modifications to this
file, hence there is a version that comes with this package.</p>
<p>For convenience, in <strong>shell</strong>, export the path to the database folders like as follows:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># bash/shell</span>
<span class="nb">echo</span> <span class="s2">&quot;export GERMLINE=/Users/kt16/Documents/Github/dandelion/database/germlines/&quot;</span> &gt;&gt; ~/.bash_profile
<span class="nb">echo</span> <span class="s2">&quot;export IGDATA=/Users/kt16/Documents/Github/dandelion/database/igblast/&quot;</span> &gt;&gt; ~/.bash_profile
<span class="nb">echo</span> <span class="s2">&quot;export BLASTDB=/Users/kt16/Documents/Github/dandelion/database/blast/&quot;</span> &gt;&gt; ~/.bash_profile
<span class="c1"># reload</span>
<span class="nb">source</span> ~/.bash_profile
</pre></div>
</div>
<p>The databases for igblast are basically setup using <a class="reference external" href="https://changeo.readthedocs.io/en/stable/examples/igblast.html">changeo’s instructions</a>.</p>
<p>If you are using a jupyter notebook initialized via jupyterhub instance, you might want to try the fix to a known issue where pathing requires some adjustments <a class="reference external" href="https://github.com/zktuong/dandelion/issues/66">https://github.com/zktuong/dandelion/issues/66</a>.</p>
<p>For reannotation of constant genes, reference fasta files were downloaded from IMGT and only sequences corresponding to <em>CH1</em> region for each constant gene/allele were retained. The headers were trimmed to only keep the gene and allele information. Links to find the sequences can be found here : <a class="reference external" href="http://www.imgt.org/genedb/GENElect?query=7.2+IGHC&amp;species=Homo+sapiens">human</a> and <a class="reference external" href="http://www.imgt.org/genedb/GENElect?query=7.2+IGHC&amp;species=Mus">mouse</a>.</p>
<p>The utility function <code class="docutils literal notranslate"><span class="pre">utl.makeblastdb</span></code> is a wrapper for:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># bash/shell</span>
makeblastdb -dbtype nucl -parse_seqids -in <span class="nv">$BLASTDB</span>/human/human_BCR_C.fasta
</pre></div>
</div>
<p>So effectively, this does the same thing:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># python</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">utl</span><span class="o">.</span><span class="n">makeblastdb</span><span class="p">(</span><span class="s1">&#39;/Users/kt16/Documents/Github/dandelion/database/blast/human/human_BCR_C.fasta&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>If you have cloned the directory from dandelion’s github, you should have all the databases ready to go and would not need to run makeblastdb.</p>
<p>This section will now demonstrate how I batch process multiple samples/files from the same donor, as it will become important later on.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Import modules</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">dandelion</span> <span class="k">as</span> <span class="nn">ddl</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_versions</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
dandelion==0.1.3.post2.dev62 pandas==1.2.3 numpy==1.20.1 matplotlib==3.3.4 networkx==2.5 scipy==1.6.1 skbio==0.5.6
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># change directory to somewhere more workable</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s1">&#39;/Users/kt16/Downloads/dandelion_tutorial/&#39;</span><span class="p">))</span>
<span class="c1"># print current working directory</span>
<span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;/Users/kt16/Downloads/dandelion_tutorial&#39;
</pre></div></div>
</div>
</div>
<div class="section" id="Step-1:-Formatting-the-headers-of-the-Cell-Ranger-fasta-file">
<h2>Step 1: Formatting the headers of the Cell Ranger fasta file<a class="headerlink" href="#Step-1:-Formatting-the-headers-of-the-Cell-Ranger-fasta-file" title="Permalink to this headline">¶</a></h2>
<p>Here, I’m adding a prefix to the headers of each contig in the fasta files, via the function <code class="docutils literal notranslate"><span class="pre">pp.format_fastas</span></code>. The prefix is basically just the folder name, so in this case it’s: <code class="docutils literal notranslate"><span class="pre">sc5p_v2_hs_PBMC_1k</span></code>, <code class="docutils literal notranslate"><span class="pre">sc5p_v2_hs_PBMC_10k</span></code>, <code class="docutils literal notranslate"><span class="pre">vdj_v1_hs_pbmc3</span></code> and <code class="docutils literal notranslate"><span class="pre">vdj_nextgem_hs_pbmc3</span></code>.</p>
<p>The function will also create sub-folders where a new fasta file and all subsequent files will be located. The function will also add the prefix to the corresponding annotation file automatically and create a copy in the same folder as the formated fasta file.</p>
<p>This is to ensure that the barcodes are consistent throughout so as not to interfere with subsequent integration with the gene expression data, which will be covered in subsequent sections.</p>
<p>The file structure should look something like this later on if the settings are left as default. The tmp directory can be deleted once the initial preprocessing has completed.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp"># </span>bash/shell
<span class="gp gp-VirtualEnv">(dandelion)</span> <span class="gp">mib113557i:dandelion_tutorial kt16$ </span>tree sc5p_v2_hs_PBMC_1k
<span class="go">sc5p_v2_hs_PBMC_1k</span>
<span class="go">├── dandelion</span>
<span class="go">│   ├── filtered_contig.fasta</span>
<span class="go">│   ├── filtered_contig_annotations.csv</span>
<span class="go">│   ├── filtered_contig_igblast_db-pass_genotyped.tsv</span>
<span class="go">│   └── tmp</span>
<span class="go">│       ├── filtered_contig_igblast.fmt7</span>
<span class="go">│       ├── filtered_contig_igblast.tsv</span>
<span class="go">│       ├── filtered_contig_igblast_db-pass.blastsummary.txt</span>
<span class="go">│       ├── filtered_contig_igblast_db-pass.tsv</span>
<span class="go">│       ├── filtered_contig_igblast_db-pass.xml</span>
<span class="go">│       ├── filtered_contig_igblast_db-pass_genotyped.tsv</span>
<span class="go">│       ├── filtered_contig_igblast_db-pass_heavy_parse-select.tsv</span>
<span class="go">│       └── filtered_contig_igblast_db-pass_light_parse-select.tsv</span>
<span class="go">├── filtered_contig.fasta</span>
<span class="go">├── filtered_contig_annotations.csv</span>
<span class="go">└── filtered_feature_bc_matrix.h5</span>
</pre></div>
</div>
<p>The first option of <code class="docutils literal notranslate"><span class="pre">pp.format_fastas</span></code> accepts a list of the fasta file paths to reformat, or list of names of folders containing the fasta files; each folder should only contain 1 fasta file, and 1 contig_annotation.csv. Make sure there’s no hidden files and delete those if present.</p>
<p>You can provide <code class="docutils literal notranslate"><span class="pre">prefixes</span></code> and/or <code class="docutils literal notranslate"><span class="pre">suffixes</span></code> to add the the cell/contig barcodes as a list and they will be formatted accordingly. The prefixes/suffixes will be separated by an underscore (<code class="docutils literal notranslate"><span class="pre">_</span></code>) if left as default but that can be adjusted with the <code class="docutils literal notranslate"><span class="pre">sep</span></code> option.</p>
<p>If you choose not to provide a prefix/suffix, then the function will simply make a copy of the original files and place it in the <code class="docutils literal notranslate"><span class="pre">dandelion</span></code> sub-folders.</p>
<p>For more complex experimental setups, such as with data from multiplexed experiments, please contact me (<a class="reference external" href="mailto:kt16&#37;&#52;&#48;sanger&#46;ac&#46;uk">kt16<span>&#64;</span>sanger<span>&#46;</span>ac<span>&#46;</span>uk</a>) and I can walk you through a slightly more advanced set up.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># the first option is a list of fasta files to format and the second option is the list of prefix to add to each file.</span>
<span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sc5p_v2_hs_PBMC_1k&#39;</span><span class="p">,</span> <span class="s1">&#39;sc5p_v2_hs_PBMC_10k&#39;</span><span class="p">,</span> <span class="s1">&#39;vdj_v1_hs_pbmc3&#39;</span><span class="p">,</span> <span class="s1">&#39;vdj_nextgem_hs_pbmc3&#39;</span><span class="p">]</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">format_fastas</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">prefix</span> <span class="o">=</span> <span class="n">samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Formating fasta(s) : 100%|██████████| 4/4 [00:01&lt;00:00,  3.23it/s]
</pre></div></div>
</div>
<hr class="docutils" />
<p><strong>Non-standard (filtered) input files</strong>:</p>
<p>If you are not using <code class="docutils literal notranslate"><span class="pre">filtered</span></code> files, e.g. <code class="docutils literal notranslate"><span class="pre">all_contig.fasta</span></code>, please specify the <code class="docutils literal notranslate"><span class="pre">filename_prefix</span></code> option for all preprocessing functions below (except for quantify_mutations).</p>
<p>For example, use <code class="docutils literal notranslate"><span class="pre">filename_prefix</span> <span class="pre">=</span> <span class="pre">'all'</span></code> for <code class="docutils literal notranslate"><span class="pre">all_contig.fasta</span></code>, or <code class="docutils literal notranslate"><span class="pre">filename_prefix</span> <span class="pre">=</span> <span class="pre">&lt;insertprefix&gt;</span></code> for any files that are named <code class="docutils literal notranslate"><span class="pre">&lt;insertprefix&gt;_contig.fasta</span></code>.</p>
<p>If you are running more than 1 sample and if each filename prefix needs to be specified, <code class="docutils literal notranslate"><span class="pre">filename_prefix</span></code> in <code class="docutils literal notranslate"><span class="pre">ddl.pp.format_fastas</span></code> will accept a list of prefixes i.e. <code class="docutils literal notranslate"><span class="pre">filename_prefix</span> <span class="pre">=</span> <span class="pre">['all',</span> <span class="pre">'filtered',</span> <span class="pre">...]</span></code>.</p>
<hr class="docutils" />
</div>
<div class="section" id="Step-2:-Reannotate-the-V/D/J-genes-with-igblastn">
<h2>Step 2: Reannotate the V/D/J genes with <em>igblastn</em><a class="headerlink" href="#Step-2:-Reannotate-the-V/D/J-genes-with-igblastn" title="Permalink to this headline">¶</a></h2>
<p>Like immcantation, we will reannotate the V(D)J genes with igblastn using the latest IMGT reference databases. <code class="docutils literal notranslate"><span class="pre">pp.reannotate_genes</span></code> uses <a class="reference external" href="https://changeo.readthedocs.io/en/stable/examples/10x.html">changeo</a>’s <code class="docutils literal notranslate"><span class="pre">AssignGenes.py</span></code> and <code class="docutils literal notranslate"><span class="pre">MakeDB.py</span></code> to call <em>igblastn</em> to reannotate the fasta files. All the column headers are now adhereing to the <a class="reference external" href="http://docs.airr-community.org/">AIRR</a> standard.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">reannotate_genes</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Assigning genes : 100%|██████████| 4/4 [07:19&lt;00:00, 109.87s/it]
</pre></div></div>
</div>
<p>If you did not set a path to the igblast or germline paths in the environment above, you need to specify the path to the folders containing the fasta files directly.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">reannotate_genes</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">igblast_db</span> <span class="o">=</span> <span class="s2">&quot;database/igblast/&quot;</span><span class="p">,</span> <span class="n">germline</span> <span class="o">=</span> <span class="s2">&quot;database/germlines/imgt/human/vdj/&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="Step-3-:-Reassigning-heavy-chain-V-gene-alleles-(optional-but-recommended)">
<h2>Step 3 : Reassigning heavy chain V gene alleles <em>(optional but recommended)</em><a class="headerlink" href="#Step-3-:-Reassigning-heavy-chain-V-gene-alleles-(optional-but-recommended)" title="Permalink to this headline">¶</a></h2>
<p>Next, we use <em>immcantation’s TIgGER</em> method to reassign allelic calls for heavy chain V genes with <code class="docutils literal notranslate"><span class="pre">pp.reassign_alleles</span></code>. As stated in TIgGER’s <a class="reference external" href="https://tigger.readthedocs.io/en/stable/">website</a> and <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/25675496/">manuscript</a>, <em>‘TIgGER is a computational method that significantly improves V(D)J allele assignments by first determining the complete set of gene segments carried by an individual (including novel alleles) from V(D)J-rearrange sequences. TIgGER
can then infer a subject’s genotype from these sequences, and use this genotype to correct the initial V(D)J allele assignments.’</em></p>
<p>This impacts on how contigs are chosen for finding clones later. It is also important when considering to do mutational analysis. For convenience, germline sequences are reconstructed at this step using the corrected V-gene alleles. Therefore, it is highly recommended to run it.</p>
<p>However, this will only work properly if there is sufficient contigs. An ideal scenario would be to run it on multiple samples from the same subject to allow for more information to be used to confidently assign a genotyped <em>v_call</em>. In this tutorial, I’m assuming the four samples can be split into two sets where sets of two corresponds to a different/single individual. So while important, this step can be skipped if you don’t have enough data to do this.</p>
<p><code class="docutils literal notranslate"><span class="pre">pp.reassign_alleles</span></code> requires the <code class="docutils literal notranslate"><span class="pre">combined_folder</span></code> option to be specified so that a merged/concatenated file can be produced for running TIgGER.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># reassigning alleles on the first set of samples</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">reassign_alleles</span><span class="p">(</span><span class="n">samples</span><span class="p">[:</span><span class="mi">2</span><span class="p">],</span> <span class="n">combined_folder</span> <span class="o">=</span> <span class="s1">&#39;tutorial_scgp1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Processing data file(s) : 100%|██████████| 2/2 [00:02&lt;00:00,  1.27s/it]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Concatenating objects
      Running tigger-genotype with novel allele discovery.
      Reassigning alleles
            Reconstructing heavy chain dmask germline sequences with v_call_genotyped.
            Reconstructing light chain dmask germline sequences with v_call.
      For convenience, entries for light chain `v_call` are copied to `v_call_genotyped`.
Returning summary plot
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/kt16/miniconda3/envs/dandelion/lib/python3.7/site-packages/plotnine/facets/facet_grid.py:136: FutureWarning: Index.__and__ operating as a set operation is deprecated, in the future this will be a logical operation matching Series.__and__.  Use index.intersection(other) instead
/Users/kt16/miniconda3/envs/dandelion/lib/python3.7/site-packages/plotnine/facets/facet_grid.py:137: FutureWarning: Index.__and__ operating as a set operation is deprecated, in the future this will be a logical operation matching Series.__and__.  Use index.intersection(other) instead
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_1_dandelion_preprocessing-10x_data_10_3.png" src="../_images/notebooks_1_dandelion_preprocessing-10x_data_10_3.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (342143629)&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Writing out to individual folders : 100%|██████████| 2/2 [00:00&lt;00:00,  5.94it/s]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># reassigning alleles on the second set of samples</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">reassign_alleles</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="mi">2</span><span class="p">:],</span> <span class="n">combined_folder</span> <span class="o">=</span> <span class="s1">&#39;tutorial_scgp2&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Processing data file(s) : 100%|██████████| 2/2 [00:01&lt;00:00,  1.04it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Concatenating objects
      Running tigger-genotype with novel allele discovery.
      Reassigning alleles
            Reconstructing heavy chain dmask germline sequences with v_call_genotyped.
            Reconstructing light chain dmask germline sequences with v_call.
      For convenience, entries for light chain `v_call` are copied to `v_call_genotyped`.
Returning summary plot
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/kt16/miniconda3/envs/dandelion/lib/python3.7/site-packages/plotnine/facets/facet_grid.py:136: FutureWarning: Index.__and__ operating as a set operation is deprecated, in the future this will be a logical operation matching Series.__and__.  Use index.intersection(other) instead
/Users/kt16/miniconda3/envs/dandelion/lib/python3.7/site-packages/plotnine/facets/facet_grid.py:137: FutureWarning: Index.__and__ operating as a set operation is deprecated, in the future this will be a logical operation matching Series.__and__.  Use index.intersection(other) instead
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_1_dandelion_preprocessing-10x_data_11_3.png" src="../_images/notebooks_1_dandelion_preprocessing-10x_data_11_3.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (344788537)&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Writing out to individual folders : 100%|██████████| 2/2 [00:00&lt;00:00,  3.90it/s]
</pre></div></div>
</div>
<p>We can see that most of the original ambiguous V calls have now been corrected and only a few remain.</p>
<p>Similar to above, if you you can specify the path to the folder containing the fasta files accordingly:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">reassign_alleles</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="mi">2</span><span class="p">:],</span> <span class="n">combined_folder</span> <span class="o">=</span> <span class="s1">&#39;tutorial_scgp2&#39;</span><span class="p">,</span> <span class="n">germline</span> <span class="o">=</span> <span class="s2">&quot;database/germlines/imgt/human/vdj&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="Step-4:-Assigning-constant-region-calls">
<h2>Step 4: Assigning constant region calls<a class="headerlink" href="#Step-4:-Assigning-constant-region-calls" title="Permalink to this headline">¶</a></h2>
<p>Cell Ranger’s annotation files provides a <em>c_gene</em> column, but rather than simply relying on Cell Ranger’s annotation, it is common to use <a class="reference external" href="https://presto.readthedocs.io/en/version-0.5.3---license-change/tools/MaskPrimers.html">immcantation-presto’s MaskPrimers.py</a> with a custom primer list.</p>
<p>As an alternative, <code class="docutils literal notranslate"><span class="pre">dandelion</span></code> includes a pre-processing function, <code class="docutils literal notranslate"><span class="pre">pp.assign_isotypes</span></code>, to use <em>blast</em> to annotate constant region calls for all contigs and retrieves the call, merging it with the tsv files. This function will overwrite the output from previous steps and add a <em>c_call</em> column at the end, or replace the existing column if it already exists. The Cell Ranger calls are returned as <code class="docutils literal notranslate"><span class="pre">c_call_10x</span></code>.</p>
<p>Further, to deal with incorrect constant gene calls due to insufficient length, a pairwise alignment will be run against <a class="reference external" href="https://immunology.sciencemag.org/content/6/56/eabe6291">curated sequences</a> that were deemed to be highly specific in distinguishing <code class="docutils literal notranslate"><span class="pre">IGHA1</span></code> vs <code class="docutils literal notranslate"><span class="pre">IGHA2</span></code>, and <code class="docutils literal notranslate"><span class="pre">IGHG1</span></code> to <code class="docutils literal notranslate"><span class="pre">IGHG4</span></code>. I have also curated sets of sequences that should help deal with <code class="docutils literal notranslate"><span class="pre">IGLC3/6/7</span></code> as these are problematic too. If there is insufficient info, the <code class="docutils literal notranslate"><span class="pre">c_call</span></code> will be returned as a combination
of the most aligned sets of sequences. Because of how similar the lambda light chains are, extremely ambiguous calls (only able to map to a common sequence across the light chains) will be returned as <code class="docutils literal notranslate"><span class="pre">IGLC</span></code>. This typically occurs when the constant sequence is very short. Those that have equal alignment scores between <code class="docutils literal notranslate"><span class="pre">IGLC3/6/7</span></code> sequences and the common sequence will be returned as a concatenated call; for example, a contig initially annotated as <code class="docutils literal notranslate"><span class="pre">IGLC3</span></code> will be returned as
<code class="docutils literal notranslate"><span class="pre">IGLC,IGLC3</span></code>.</p>
<hr class="docutils" />
<p>The curated sequences can be updated/replaced with a dict-of-dict-of-dict style dictionary via the option <code class="docutils literal notranslate"><span class="pre">correction_dict</span></code>. The provided dictionary should be a nested dictionary like the following:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">primer_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;IGHG&#39;</span><span class="p">:{</span>
        <span class="s1">&#39;IGHG1&#39;</span><span class="p">:</span><span class="s1">&#39;GCCTCCACCAAGGGCCCATCGGTCTTCCCCCTGGCACCCTCCTCCAAGAGCACCTCTGGGGGCACAGCGGCCCTGGGC&#39;</span><span class="p">,</span>
        <span class="s1">&#39;IGHG2&#39;</span><span class="p">:</span><span class="s1">&#39;GCCTCCACCAAGGGCCCATCGGTCTTCCCCCTGGCGCCCTGCTCCAGGAGCACCTCCGAGAGCACAGCGGCCCTGGGC&#39;</span><span class="p">,</span>
        <span class="s1">&#39;IGHG3&#39;</span><span class="p">:</span><span class="s1">&#39;GCTTCCACCAAGGGCCCATCGGTCTTCCCCCTGGCGCCCTGCTCCAGGAGCACCTCTGGGGGCACAGCGGCCCTGGGC&#39;</span><span class="p">,</span>
        <span class="s1">&#39;IGHG4&#39;</span><span class="p">:</span><span class="s1">&#39;GCTTCCACCAAGGGCCCATCCGTCTTCCCCCTGGCGCCCTGCTCCAGGAGCACCTCCGAGAGCACAGCCGCCCTGGGC&#39;</span><span class="p">}}</span>
</pre></div>
</div>
<p>The key for the first level of the dictionary is used for searching whether the string pattern exists in the <code class="docutils literal notranslate"><span class="pre">c_call</span></code>, and the second level holds the dictionary for the the reference sequences to align to. The keys in the second level are used for replacing the existing <code class="docutils literal notranslate"><span class="pre">c_call</span></code> annotation if it is returned with the highest alignment score. The function currently only accepts 2-4 reference sequences for the pairwise alignment.</p>
<hr class="docutils" />
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">assign_isotypes</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Retrieving contant region calls, parallelizing with 3 cpus : 100%|██████████| 2059/2059 [00:18&lt;00:00, 113.46it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_1_dandelion_preprocessing-10x_data_15_1.png" src="../_images/notebooks_1_dandelion_preprocessing-10x_data_15_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (349671201)&gt;
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Retrieving contant region calls, parallelizing with 3 cpus : 100%|██████████| 2601/2601 [00:28&lt;00:00, 90.70it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_1_dandelion_preprocessing-10x_data_15_4.png" src="../_images/notebooks_1_dandelion_preprocessing-10x_data_15_4.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (349452577)&gt;
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Retrieving contant region calls, parallelizing with 3 cpus : 100%|██████████| 2059/2059 [00:18&lt;00:00, 113.83it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_1_dandelion_preprocessing-10x_data_15_7.png" src="../_images/notebooks_1_dandelion_preprocessing-10x_data_15_7.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (347611561)&gt;
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Retrieving contant region calls, parallelizing with 3 cpus : 100%|██████████| 3222/3222 [00:42&lt;00:00, 75.61it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_1_dandelion_preprocessing-10x_data_15_10.png" src="../_images/notebooks_1_dandelion_preprocessing-10x_data_15_10.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (352347573)&gt;
</pre></div></div>
</div>
<p>Should you want to use a different reference fasta file for this step, run with the following option:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">assign_isotypes</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">blastdb</span> <span class="o">=</span> <span class="s2">&quot;path/to/custom_BCR_constant.fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>This may take a while when dealing with large files; the number of cpus to size of file isn’t exactly linear. Nevertheless, I have enabled parallelization as default because there were noticeable improvements in processing speeds with the smaller files. It should work faster with more cpus. The default option will return a summary plot that can be disabled with <code class="docutils literal notranslate"><span class="pre">plot</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p>
<p>Finally, it’s worthwhile to manually check the the sequences for constant calls returned as IGHA1-2, IGHG1-4 and the light chains and manually correct them if necessary.</p>
</div>
<div class="section" id="Step-5:-Quantify-mutations-(optional).">
<h2>Step 5: Quantify mutations <em>(optional)</em>.<a class="headerlink" href="#Step-5:-Quantify-mutations-(optional)." title="Permalink to this headline">¶</a></h2>
<p>At this stage, with all the necessary columns in the files, you can quantify the basic mutational load with <code class="docutils literal notranslate"><span class="pre">pp.quantify_mutations</span></code>, a wrapper of <code class="docutils literal notranslate"><span class="pre">SHaZaM</span></code>’s basic mutational analysis in R, before subsequent analyses. This will be covered again later in the <code class="docutils literal notranslate"><span class="pre">Calculating</span> <span class="pre">diversity</span> <span class="pre">and</span> <span class="pre">mutation</span></code> section.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="c1"># quantify mutations</span>
<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">desc</span> <span class="o">=</span> <span class="s1">&#39;Basic mutational load analysis &#39;</span><span class="p">):</span>
    <span class="n">filePath</span> <span class="o">=</span> <span class="n">s</span><span class="o">+</span><span class="s1">&#39;/dandelion/filtered_contig_igblast_db-pass_genotyped.tsv&#39;</span>
    <span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">quantify_mutations</span><span class="p">(</span><span class="n">filePath</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Basic mutational load analysis : 100%|██████████| 4/4 [00:42&lt;00:00, 10.74s/it]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="2_dandelion_filtering-10x_data.html" class="btn btn-neutral float-right" title="Filtering" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../tutorial_short.html" class="btn btn-neutral float-left" title="&lt;no title&gt;" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, zktuong.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>