

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>BCR clustering &mdash; dandelion  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../_static/dandelion_logo.png"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Interoperability with scirpy" href="1c_dandelion_scirpy.html" />
    <link rel="prev" title="Reading 10X Cell Ranger output directly" href="1b_dandelion_noreannotation-10x_data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/dandelion_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../README.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#required-database">Required database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#external-softwares">External softwares</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#basic-requirements">Basic requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#acknowledgements">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#citation">Citation</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="singularity_preprocessing.html">Dandelion preprocessing with Singularity</a></li>
<li class="toctree-l1"><a class="reference internal" href="gamma_delta.html">Dandelion for TCR gamma/delta reannotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="1_dandelion_preprocessing-10x_data.html">Pre-processing (Re-annotation)</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_dandelion_filtering-10x_data.html">Filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="1b_dandelion_noreannotation-10x_data.html">Reading 10X Cell Ranger output directly</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">BCR clustering</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Finding-clones">Finding clones</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Running-tl.find_clones">Running <code class="docutils literal notranslate"><span class="pre">tl.find_clones</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#Alternative-:-Running-tl.define_clones">Alternative : Running <code class="docutils literal notranslate"><span class="pre">tl.define_clones</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Generation-of-V(D)J-network">Generation of V(D)J network</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="1c_dandelion_scirpy.html">Interoperability with <code class="docutils literal notranslate"><span class="pre">scirpy</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="4_dandelion_visualization-10x_data.html">Visualizing BCR data</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_dandelion_diversity_and_mutation-10x_data.html">Calculating diversity and mutation</a></li>
<li class="toctree-l1"><a class="reference internal" href="0_dandelion_primer.html"><em>Dandelion</em> class</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_dandelion_running_from_R-10x_data.html">Running from R</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_dandelion_TCR_data_10x_data.html">Analyzing TCR data</a></li>
</ul>
<p class="caption"><span class="caption-text">API:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">Preprocessing: <cite>pp</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.preprocessing.external">Preprocessing (external): <cite>pp.external</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.tools">Tools: <cite>tl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.plotting">Plotting: <cite>pl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.utilities">Utilities: <cite>utl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.Dandelion">Dandelion</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.logging">Logging</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">dandelion</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../tutorial_short.html">&lt;no title&gt;</a> &raquo;</li>
        
      <li>BCR clustering</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/notebooks/3_dandelion_findingclones-10x_data.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="BCR-clustering">
<h1>BCR clustering<a class="headerlink" href="#BCR-clustering" title="Permalink to this headline">¶</a></h1>
<img alt="dandelion_logo" src="../_images/dandelion_logo_illustration.png" />
<p>On the topic of finding clones/clonotypes, there are many ways used for clustering BCRs, almost all involving some measure based on sequence similarity. There are also a lot of very well established guidelines and criterias maintained by the BCR community. For example, <em>immcantation</em> uses a number of model-based <a class="reference external" href="https://changeo.readthedocs.io/en/stable/methods/clustering.html">methods</a> to group clones based on the distribution of length-normalised junctional hamming distance while others use
the whole BCR V(D)J sequence to define clones as shown in this recent <a class="reference external" href="https://www.nature.com/articles/s41586-019-1595-3.pdf">paper</a>.</p>
<p><strong>Import modules</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">dandelion</span> <span class="k">as</span> <span class="nn">ddl</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
dandelion==0.1.3.post2.dev62 pandas==1.2.3 numpy==1.20.1 matplotlib==3.3.4 networkx==2.5 scipy==1.6.1 skbio==0.5.6
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># change directory to somewhere more workable</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s1">&#39;/Users/kt16/Downloads/dandelion_tutorial/&#39;</span><span class="p">))</span>
<span class="c1"># I&#39;m importing scanpy here to make use of its logging module.</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">3</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
scanpy==1.6.0 anndata==0.7.6 umap==0.4.6 numpy==1.20.1 scipy==1.6.1 pandas==1.2.3 scikit-learn==0.24.1 statsmodels==0.12.1 python-igraph==0.8.3 leidenalg==0.8.3
</pre></div></div>
</div>
<p><strong>Read in the previously saved files</strong></p>
<p>I will work with the same example from the previous section since I have the filtered V(D)J data stored in a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> class.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">read_h5</span><span class="p">(</span><span class="s1">&#39;dandelion_results.h5&#39;</span><span class="p">)</span>
<span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 562 and n_contigs = 1130
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_count&#39;, &#39;duplicate_count&#39;
    metadata: &#39;sample_id&#39;, &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_genotyped_VDJ&#39;, &#39;v_call_genotyped_VJ&#39;, &#39;j_call_VDJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;duplicate_count_VDJ_0&#39;, &#39;duplicate_count_VDJ_1&#39;, &#39;duplicate_count_VDJ_2&#39;, &#39;duplicate_count_VJ_0&#39;, &#39;duplicate_count_VJ_1&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;status&#39;, &#39;status_summary&#39;, &#39;productive&#39;, &#39;productive_summary&#39;, &#39;isotype&#39;, &#39;isotype_summary&#39;, &#39;vdj_status&#39;, &#39;vdj_status_summary&#39;, &#39;constant_status_summary&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<div class="section" id="Finding-clones">
<h2>Finding clones<a class="headerlink" href="#Finding-clones" title="Permalink to this headline">¶</a></h2>
<p>The following is <em>dandelion</em>’s implementation of a rather conventional method to define clones, <code class="docutils literal notranslate"><span class="pre">tl.find_clones</span></code>.</p>
<p>Clone definition is based on the following criterias:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="n">Identical</span> <span class="n">IGH</span> <span class="n">V</span><span class="o">-</span><span class="n">J</span> <span class="n">gene</span> <span class="n">usage</span><span class="o">.</span>

<span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="n">Identical</span> <span class="n">CDR3</span> <span class="n">junctional</span> <span class="n">sequence</span> <span class="n">length</span><span class="o">.</span>

<span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="n">CDR3</span> <span class="n">Junctional</span> <span class="n">sequences</span> <span class="n">attains</span> <span class="n">a</span> <span class="n">minimum</span> <span class="n">of</span> <span class="o">%</span> <span class="n">sequence</span> <span class="n">similarity</span><span class="p">,</span> <span class="n">based</span> <span class="n">on</span> <span class="n">hamming</span> <span class="n">distance</span><span class="o">.</span> <span class="n">The</span> <span class="n">similarity</span> <span class="n">cut</span><span class="o">-</span><span class="n">off</span> <span class="ow">is</span> <span class="n">tunable</span> <span class="p">(</span><span class="n">default</span> <span class="ow">is</span> <span class="mi">85</span><span class="o">%</span><span class="p">)</span><span class="o">.</span>

<span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="n">Light</span> <span class="n">chain</span> <span class="n">usage</span><span class="o">.</span> <span class="n">If</span> <span class="n">cells</span> <span class="n">within</span> <span class="n">clones</span> <span class="n">use</span> <span class="n">different</span> <span class="n">light</span> <span class="n">chains</span><span class="p">,</span> <span class="n">the</span> <span class="n">clone</span> <span class="n">will</span> <span class="n">be</span> <span class="n">splitted</span> <span class="n">following</span> <span class="n">the</span> <span class="n">same</span> <span class="n">conditions</span> <span class="k">for</span> <span class="n">heavy</span> <span class="n">chains</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="mi">3</span><span class="p">)</span> <span class="k">as</span> <span class="n">above</span><span class="o">.</span>
</pre></div>
</div>
<p>The ‘clone_id’ name follows a <code class="docutils literal notranslate"><span class="pre">{A}_{B}_{C}_{D}</span></code> format and largely reflects the conditions above where:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">A</span><span class="p">}</span> <span class="n">indicates</span> <span class="k">if</span> <span class="n">the</span> <span class="n">contigs</span> <span class="n">use</span> <span class="n">the</span> <span class="n">same</span> <span class="n">IGH</span> <span class="n">V</span><span class="o">/</span><span class="n">J</span> <span class="n">genes</span><span class="o">.</span>

<span class="p">{</span><span class="n">B</span><span class="p">}</span> <span class="n">indicates</span> <span class="k">if</span> <span class="n">IGH</span> <span class="n">junctional</span> <span class="n">sequences</span> <span class="n">are</span> <span class="n">equal</span> <span class="ow">in</span> <span class="n">length</span><span class="o">.</span>

<span class="p">{</span><span class="n">C</span><span class="p">}</span> <span class="n">indicates</span> <span class="k">if</span> <span class="n">clones</span> <span class="n">are</span> <span class="n">splitted</span> <span class="n">based</span> <span class="n">on</span> <span class="n">junctional</span> <span class="n">hamming</span> <span class="n">distance</span> <span class="n">threshold</span>

<span class="p">{</span><span class="n">D</span><span class="p">}</span> <span class="n">indicates</span> <span class="n">light</span> <span class="n">chain</span> <span class="n">pairing</span><span class="o">.</span>
</pre></div>
</div>
<p>The last position will not be annotated if there’s only one group of light chains usage detected in the clone.</p>
<div class="section" id="Running-tl.find_clones">
<h3>Running <code class="docutils literal notranslate"><span class="pre">tl.find_clones</span></code><a class="headerlink" href="#Running-tl.find_clones" title="Permalink to this headline">¶</a></h3>
<p>The function will take a file path, a pandas <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code> (for example if you’ve used pandas to read in the filtered file already), or a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> class object. The default mode for calculation of junctional hamming distance is to use the CDR3 junction amino acid sequences, specified via the <code class="docutils literal notranslate"><span class="pre">key</span></code> option (<code class="docutils literal notranslate"><span class="pre">None</span></code> defaults to <code class="docutils literal notranslate"><span class="pre">junction_aa</span></code>). You can switch it to using CDR3 junction nucleotide sequences (<code class="docutils literal notranslate"><span class="pre">key</span> <span class="pre">=</span> <span class="pre">'junction'</span></code>, or even the full V(D)J amino acid sequence
(<code class="docutils literal notranslate"><span class="pre">key</span> <span class="pre">=</span> <span class="pre">'sequence_alignment_aa</span></code>), as long as the column name exists in the <code class="docutils literal notranslate"><span class="pre">.data</span></code> slot.</p>
<p>If you want to use the alleles for defining V-J gene usuage, specify:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">by_alleles</span> <span class="o">=</span> <span class="kc">True</span>
</pre></div>
</div>
<p>Clustering TCR is potentially possible with the same setup, although some parameters may need to be tweaked.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">find_clones</span><span class="p">(</span><span class="n">vdj</span><span class="p">)</span>
<span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finding clonotypes
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Finding clones based on VDJ chains : 100%|██████████| 153/153 [00:00&lt;00:00, 4781.42it/s]
Refining clone assignment based on VJ chain pairing : 100%|██████████| 562/562 [00:00&lt;00:00, 426796.82it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
 finished: Updated Dandelion object:
   &#39;data&#39;, contig-indexed clone table
   &#39;metadata&#39;, cell-indexed clone table
 (0:00:00)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 562 and n_contigs = 1130
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_count&#39;, &#39;duplicate_count&#39;, &#39;clone_id&#39;
    metadata: &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;sample_id&#39;, &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_genotyped_VDJ&#39;, &#39;v_call_genotyped_VJ&#39;, &#39;j_call_VDJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;duplicate_count_VDJ_0&#39;, &#39;duplicate_count_VDJ_1&#39;, &#39;duplicate_count_VDJ_2&#39;, &#39;duplicate_count_VJ_0&#39;, &#39;duplicate_count_VJ_1&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;status&#39;, &#39;status_summary&#39;, &#39;productive&#39;, &#39;productive_summary&#39;, &#39;isotype&#39;, &#39;isotype_summary&#39;, &#39;vdj_status&#39;, &#39;vdj_status_summary&#39;, &#39;constant_status_summary&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<p>This will return a new column with the column name <code class="docutils literal notranslate"><span class="pre">'clone_id'</span></code> as per convention. If a file path is provided as input, it will also save the file automatically into the base directory of the file name. Otherwise, a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object will be returned.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span><span class="o">.</span><span class="n">metadata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>clone_id</th>
      <th>clone_id_by_size</th>
      <th>sample_id</th>
      <th>locus_VDJ</th>
      <th>locus_VJ</th>
      <th>productive_VDJ</th>
      <th>productive_VJ</th>
      <th>v_call_genotyped_VDJ</th>
      <th>v_call_genotyped_VJ</th>
      <th>j_call_VDJ</th>
      <th>...</th>
      <th>junction_aa_VJ</th>
      <th>status</th>
      <th>status_summary</th>
      <th>productive</th>
      <th>productive_summary</th>
      <th>isotype</th>
      <th>isotype_summary</th>
      <th>vdj_status</th>
      <th>vdj_status_summary</th>
      <th>constant_status_summary</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_ACACTGATCGGTTCGG</th>
      <td>131_4_1_58</td>
      <td>2</td>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IGH</td>
      <td>IGK</td>
      <td>True</td>
      <td>True</td>
      <td>IGHV4-59</td>
      <td>IGKV4-1</td>
      <td>IGHJ6</td>
      <td>...</td>
      <td>CQQYYSTPRTF</td>
      <td>IGH + IGK</td>
      <td>IGH + IGK</td>
      <td>True + True</td>
      <td>True + True</td>
      <td>IgM</td>
      <td>IgM</td>
      <td>Single + Single</td>
      <td>Single</td>
      <td>Single</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_CCGGTAGGTCAGAAGC</th>
      <td>17_1_1_163</td>
      <td>4</td>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IGH</td>
      <td>IGL</td>
      <td>True</td>
      <td>True</td>
      <td>IGHV3-15</td>
      <td>IGLV2-23</td>
      <td>IGHJ3</td>
      <td>...</td>
      <td>CCAYAGSSTFPYVF</td>
      <td>IGH + IGL</td>
      <td>IGH + IGL</td>
      <td>True + True</td>
      <td>True + True</td>
      <td>IgM</td>
      <td>IgM</td>
      <td>Single + Single</td>
      <td>Single</td>
      <td>Single</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_CGATCGGAGATGTCGG</th>
      <td>146_2_1_38</td>
      <td>8</td>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IGH</td>
      <td>IGL</td>
      <td>True</td>
      <td>True</td>
      <td>IGHV4-39</td>
      <td>IGLV3-21</td>
      <td>IGHJ5</td>
      <td>...</td>
      <td>CQIWHDNYHVVF</td>
      <td>IGH + IGL</td>
      <td>IGH + IGL</td>
      <td>True + True</td>
      <td>True + True</td>
      <td>IgA</td>
      <td>IgA</td>
      <td>Single + Multi_VJ_j</td>
      <td>Single</td>
      <td>Single</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_CGGACGTGTTGATTCG</th>
      <td>22_2_1_55</td>
      <td>11</td>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IGH</td>
      <td>IGL</td>
      <td>True</td>
      <td>True</td>
      <td>IGHV1-3</td>
      <td>IGLV1-51</td>
      <td>IGHJ6</td>
      <td>...</td>
      <td>CGTWDSSLSASF</td>
      <td>IGH + IGL</td>
      <td>IGH + IGL</td>
      <td>True + True</td>
      <td>True + True</td>
      <td>IgM</td>
      <td>IgM</td>
      <td>Single + Multi_VJ_j</td>
      <td>Single</td>
      <td>Single</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_GTACGTACAGCCTATA</th>
      <td>59_3_2_308</td>
      <td>10</td>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IGH</td>
      <td>IGK</td>
      <td>True</td>
      <td>True</td>
      <td>IGHV4-59</td>
      <td>IGKV3-15</td>
      <td>IGHJ4</td>
      <td>...</td>
      <td>CQQYDNWPRTF</td>
      <td>IGH + IGK</td>
      <td>IGH + IGK</td>
      <td>True + True</td>
      <td>True + True</td>
      <td>IgM</td>
      <td>IgM</td>
      <td>Single + Single</td>
      <td>Single</td>
      <td>Single</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTCTCAACACATCTTT</th>
      <td>116_2_1_191</td>
      <td>353</td>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IGH</td>
      <td>IGK</td>
      <td>True</td>
      <td>True</td>
      <td>IGHV3-15</td>
      <td>IGKV2-28|IGKV2D-28</td>
      <td>IGHJ6</td>
      <td>...</td>
      <td>CMQALQTRYTF</td>
      <td>IGH + IGK</td>
      <td>IGH + IGK</td>
      <td>True + True</td>
      <td>True + True</td>
      <td>IgM</td>
      <td>IgM</td>
      <td>Single + Multi_VJ_v</td>
      <td>Single</td>
      <td>Single</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTCTCCTAGGGCACTA</th>
      <td>40_5_2_289</td>
      <td>382</td>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IGH</td>
      <td>IGL</td>
      <td>True</td>
      <td>True</td>
      <td>IGHV1-2</td>
      <td>IGLV2-14</td>
      <td>IGHJ4</td>
      <td>...</td>
      <td>CSSYASSSTVVF</td>
      <td>IGH + IGL</td>
      <td>IGH + IGL</td>
      <td>True + True</td>
      <td>True + True</td>
      <td>IgM</td>
      <td>IgM</td>
      <td>Single + Multi_VJ_j</td>
      <td>Single</td>
      <td>Single</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTCTTAGCAAACGCGA</th>
      <td>71_7_1_249</td>
      <td>481</td>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IGH</td>
      <td>IGK</td>
      <td>True</td>
      <td>True</td>
      <td>IGHV3-15</td>
      <td>IGKV1D-39|IGKV1-39</td>
      <td>IGHJ4</td>
      <td>...</td>
      <td>CQQSYSMSTF</td>
      <td>IGH + IGK</td>
      <td>IGH + IGK</td>
      <td>True + True</td>
      <td>True + True</td>
      <td>IgM</td>
      <td>IgM</td>
      <td>Single + Multi_VJ_v</td>
      <td>Single</td>
      <td>Single</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA</th>
      <td>117_3_2_56</td>
      <td>463</td>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IGH</td>
      <td>IGL</td>
      <td>True</td>
      <td>True</td>
      <td>IGHV3-7</td>
      <td>IGLV1-51</td>
      <td>IGHJ6</td>
      <td>...</td>
      <td>CGTWDSSLSAVVF</td>
      <td>IGH + IGL</td>
      <td>IGH + IGL</td>
      <td>True + True</td>
      <td>True + True</td>
      <td>IgM</td>
      <td>IgM</td>
      <td>Single + Multi_VJ_j</td>
      <td>Single</td>
      <td>Single</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG</th>
      <td>34_1_1_127</td>
      <td>402</td>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IGH</td>
      <td>IGL</td>
      <td>True</td>
      <td>True</td>
      <td>IGHV2-5</td>
      <td>IGLV1-47</td>
      <td>IGHJ4</td>
      <td>...</td>
      <td>CAAWDDSLTGVVF</td>
      <td>IGH + IGL</td>
      <td>IGH + IGL</td>
      <td>True + True</td>
      <td>True + True</td>
      <td>IgM</td>
      <td>IgM</td>
      <td>Single + Multi_VJ_j</td>
      <td>Single</td>
      <td>Single</td>
    </tr>
  </tbody>
</table>
<p>562 rows × 29 columns</p>
</div></div>
</div>
</div>
<div class="section" id="Alternative-:-Running-tl.define_clones">
<h3>Alternative : Running <code class="docutils literal notranslate"><span class="pre">tl.define_clones</span></code><a class="headerlink" href="#Alternative-:-Running-tl.define_clones" title="Permalink to this headline">¶</a></h3>
<p>Alternatively, a wrapper to call <em>changeo’s</em> <code class="docutils literal notranslate"><span class="pre">DefineClones.py</span></code> is also included. To run it, you need to choose the distance threshold for clonal assignment. To facilitate this, the function <code class="docutils literal notranslate"><span class="pre">pp.calculate_threshold</span></code> will run <a class="reference external" href="https://shazam.readthedocs.io/en/stable/topics/distToNearest/">shazam’s distToNearest</a> function and return a plot showing the length normalized hamming distance distribution and automated threshold value.</p>
<p>Again, <code class="docutils literal notranslate"><span class="pre">pp.calculate_threshold</span></code> will take a file path, pandas <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code> or <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object as input. If a dandelion object is provided, the threshold value will be inserted into the <code class="docutils literal notranslate"><span class="pre">.threshold</span></code> slot. For more fine control, please use <a class="reference external" href="https://shazam.readthedocs.io/en/stable/topics/distToNearest/">shazam’s distToNearest</a> and <a class="reference external" href="https://immcantation.readthedocs.io/en/stable/tutorials/10x_tutorial.html">changeo’s DefineClones.py</a> functions directly.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_threshold</span><span class="p">(</span><span class="n">vdj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating threshold
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
R[write to console]: Error in (function (db, sequenceColumn = &#34;junction&#34;, vCallColumn = &#34;v_call&#34;,  :
  2 cell(s) with multiple heavy chains found. One heavy chain per cell is expected.

</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Rerun this after filtering. For now, switching to heavy mode.
      Threshold method &#39;density&#39; did not return with any values. Switching to method = &#39;gmm&#39;.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones-10x_data_13_3.png" src="../_images/notebooks_3_dandelion_findingclones-10x_data_13_3.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (348688929)&gt;
 finished: Updated Dandelion object:
   &#39;threshold&#39;, threshold value for tuning clonal assignment
 (0:03:08)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># see the actual value in .threshold slot</span>
<span class="n">vdj</span><span class="o">.</span><span class="n">threshold</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.3984713368194217
</pre></div></div>
</div>
<p>You can also manually select a value as the threshold if you wish. Note that rerunning this with <code class="docutils literal notranslate"><span class="pre">manual_threshold</span></code> is just for reproducing the plot but with the line at 0.1 in this tutorial. You can just edit <code class="docutils literal notranslate"><span class="pre">vdj.threshold</span></code> directly if you wish.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_threshold</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">manual_threshold</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating threshold
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
R[write to console]: Error in (function (db, sequenceColumn = &#34;junction&#34;, vCallColumn = &#34;v_call&#34;,  :
  2 cell(s) with multiple heavy chains found. One heavy chain per cell is expected.

</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Rerun this after filtering. For now, switching to heavy mode.
      Threshold method &#39;density&#39; did not return with any values. Switching to method = &#39;gmm&#39;.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones-10x_data_16_3.png" src="../_images/notebooks_3_dandelion_findingclones-10x_data_16_3.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (357826549)&gt;
 finished: Updated Dandelion object:
   &#39;threshold&#39;, threshold value for tuning clonal assignment
 (0:04:38)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># see the updated .threshold slot</span>
<span class="n">vdj</span><span class="o">.</span><span class="n">threshold</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.1
</pre></div></div>
</div>
<p>We can run <code class="docutils literal notranslate"><span class="pre">tl.define_clones</span></code> to call <em>changeo’s</em> <code class="docutils literal notranslate"><span class="pre">DefineClones.py</span></code>; see <a class="reference external" href="https://changeo.readthedocs.io/en/stable/methods/clustering.html">here</a> for more info. Note, if a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> or file path is provided as the input, the value in dist option (corresponds to threshold value) needs to be manually supplied. If a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object is provided, it will automatically retrieve it from the threshold slot.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">define_clones</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">key_added</span> <span class="o">=</span> <span class="s1">&#39;changeo_clone_id&#39;</span><span class="p">)</span>
<span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finding clones
 finished: Updated Dandelion object:
   &#39;data&#39;, contig-indexed clone table
   &#39;metadata&#39;, cell-indexed clone table
 (0:00:05)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 562 and n_contigs = 1130
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_count&#39;, &#39;duplicate_count&#39;, &#39;clone_id&#39;, &#39;changeo_clone_id&#39;
    metadata: &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;sample_id&#39;, &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_genotyped_VDJ&#39;, &#39;v_call_genotyped_VJ&#39;, &#39;j_call_VDJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;duplicate_count_VDJ_0&#39;, &#39;duplicate_count_VDJ_1&#39;, &#39;duplicate_count_VDJ_2&#39;, &#39;duplicate_count_VJ_0&#39;, &#39;duplicate_count_VJ_1&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;status&#39;, &#39;status_summary&#39;, &#39;productive&#39;, &#39;productive_summary&#39;, &#39;isotype&#39;, &#39;isotype_summary&#39;, &#39;vdj_status&#39;, &#39;vdj_status_summary&#39;, &#39;constant_status_summary&#39;, &#39;changeo_clone_id&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<p>Note that I specified the option <code class="docutils literal notranslate"><span class="pre">key_added</span></code> and this adds the output from <code class="docutils literal notranslate"><span class="pre">tl.define_clones</span></code> into a separate column. If left as default (<code class="docutils literal notranslate"><span class="pre">None</span></code>), it will write into <code class="docutils literal notranslate"><span class="pre">clone_id</span></code> column. The same option can be specified in <code class="docutils literal notranslate"><span class="pre">tl.find_clones</span></code> earlier.</p>
</div>
</div>
<div class="section" id="Generation-of-V(D)J-network">
<h2>Generation of V(D)J network<a class="headerlink" href="#Generation-of-V(D)J-network" title="Permalink to this headline">¶</a></h2>
<p><em>dandelion</em> generates a network to facilitate visualisation of results. This uses the full V(D)J contig sequences instead of just the junctional sequences to chart a tree-like network for each clone. The actual visualization will be achieved through <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> later.</p>
<p><strong>tl.generate_network</strong></p>
<p>First we need to generate the network. <code class="docutils literal notranslate"><span class="pre">tl.generate_network</span></code> will take a V(D)J table that has clones defined, specifically under the <code class="docutils literal notranslate"><span class="pre">'clone_id'</span></code> column. The default mode is to use amino acid sequences for constructing Levenshtein distance matrices, but can be toggled using the <code class="docutils literal notranslate"><span class="pre">key</span></code> option.</p>
<p>If you have a pre-processed table parsed from immcantation’s <a class="reference external" href="https://changeo.readthedocs.io/en/stable/examples/10x.html">method</a>, or any other method as long as it’s in a <em>AIRR</em> format, the table can be used as well.</p>
<p>You can specify the <code class="docutils literal notranslate"><span class="pre">clone_key</span></code> option for generating the network for the clone id definition of choice as long as it exists as a column in the <code class="docutils literal notranslate"><span class="pre">.data</span></code> slot.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">generate_network</span><span class="p">(</span><span class="n">vdj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Generating network
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Calculating distances... : 100%|██████████| 5/5 [00:01&lt;00:00,  3.24it/s]
Generating edge list : 100%|██████████| 12/12 [00:00&lt;00:00, 884.50it/s]
Linking edges : 100%|██████████| 545/545 [00:00&lt;00:00, 5747.54it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
generating network layout
 finished: Updated Dandelion object:
   &#39;data&#39;, contig-indexed clone table
   &#39;metadata&#39;, cell-indexed clone table
   &#39;distance&#39;, heavy and light chain distance matrices
   &#39;edges&#39;, network edges
   &#39;layout&#39;, network layout
   &#39;graph&#39;, network (0:00:06)
</pre></div></div>
</div>
<p>This step works reasonably fast here but will take quite a while when a lot of contigs are provided.</p>
<p>You can also downsample the number of cells. This will return a new object as a downsampled copy of the original with it’s own distance matrix.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj_downsample</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">generate_network</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">downsample</span> <span class="o">=</span> <span class="mi">500</span><span class="p">)</span>
<span class="n">vdj_downsample</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Generating network
Downsampling to 500 cells.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Calculating distances... : 100%|██████████| 4/4 [00:01&lt;00:00,  3.35it/s]
Generating edge list : 100%|██████████| 11/11 [00:00&lt;00:00, 901.70it/s]
Linking edges : 100%|██████████| 483/483 [00:00&lt;00:00, 1622.05it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
generating network layout
 finished: Updated Dandelion object:
   &#39;data&#39;, contig-indexed clone table
   &#39;metadata&#39;, cell-indexed clone table
   &#39;distance&#39;, heavy and light chain distance matrices
   &#39;edges&#39;, network edges
   &#39;layout&#39;, network layout
   &#39;graph&#39;, network (0:00:08)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 498 and n_contigs = 1001
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_count&#39;, &#39;duplicate_count&#39;, &#39;clone_id&#39;, &#39;changeo_clone_id&#39;
    metadata: &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;sample_id&#39;, &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_genotyped_VDJ&#39;, &#39;v_call_genotyped_VJ&#39;, &#39;j_call_VDJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;duplicate_count_VDJ_0&#39;, &#39;duplicate_count_VDJ_1&#39;, &#39;duplicate_count_VJ_0&#39;, &#39;duplicate_count_VJ_1&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;status&#39;, &#39;status_summary&#39;, &#39;productive&#39;, &#39;productive_summary&#39;, &#39;isotype&#39;, &#39;isotype_summary&#39;, &#39;vdj_status&#39;, &#39;vdj_status_summary&#39;, &#39;constant_status_summary&#39;
    distance: &#39;VDJ_0&#39;, &#39;VDJ_1&#39;, &#39;VJ_0&#39;, &#39;VJ_1&#39;
    edges: &#39;source&#39;, &#39;target&#39;, &#39;weight&#39;
    layout: layout for 498 vertices, layout for 26 vertices
    graph: networkx graph of 498 vertices, networkx graph of 26 vertices
</pre></div></div>
</div>
<p><strong>check the newly re-initialized Dandelion object</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 562 and n_contigs = 1130
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_count&#39;, &#39;duplicate_count&#39;, &#39;clone_id&#39;, &#39;changeo_clone_id&#39;
    metadata: &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;sample_id&#39;, &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_genotyped_VDJ&#39;, &#39;v_call_genotyped_VJ&#39;, &#39;j_call_VDJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;duplicate_count_VDJ_0&#39;, &#39;duplicate_count_VDJ_1&#39;, &#39;duplicate_count_VDJ_2&#39;, &#39;duplicate_count_VJ_0&#39;, &#39;duplicate_count_VJ_1&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;status&#39;, &#39;status_summary&#39;, &#39;productive&#39;, &#39;productive_summary&#39;, &#39;isotype&#39;, &#39;isotype_summary&#39;, &#39;vdj_status&#39;, &#39;vdj_status_summary&#39;, &#39;constant_status_summary&#39;, &#39;changeo_clone_id&#39;
    distance: &#39;VDJ_0&#39;, &#39;VDJ_1&#39;, &#39;VDJ_2&#39;, &#39;VJ_0&#39;, &#39;VJ_1&#39;
    edges: &#39;source&#39;, &#39;target&#39;, &#39;weight&#39;
    layout: layout for 562 vertices, layout for 29 vertices
    graph: networkx graph of 562 vertices, networkx graph of 29 vertices
</pre></div></div>
</div>
<p>The graph/networks can be accessed through the <code class="docutils literal notranslate"><span class="pre">.graph</span></code> slot as an <code class="docutils literal notranslate"><span class="pre">networkx</span></code> graph object if you want to extract the data for network statistics or make any changes to the network.</p>
<p>At this point, we can save the dandelion object; the file can be quite big because the distance matrix is not sparse. I reccomend some form of compression (I use <code class="docutils literal notranslate"><span class="pre">bzip2</span></code> below but that can impact on read/write times significantly). See <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_hdf.html">here</a> for options compression options.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span><span class="o">.</span><span class="n">write_h5</span><span class="p">(</span><span class="s1">&#39;dandelion_results.h5&#39;</span><span class="p">,</span> <span class="n">complib</span> <span class="o">=</span> <span class="s1">&#39;bzip2&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="1c_dandelion_scirpy.html" class="btn btn-neutral float-right" title="Interoperability with scirpy" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="1b_dandelion_noreannotation-10x_data.html" class="btn btn-neutral float-left" title="Reading 10X Cell Ranger output directly" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, zktuong.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>