

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Analyzing TCR data &mdash; dandelion  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../_static/dandelion_logo.png"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Preprocessing: pp" href="../api.html" />
    <link rel="prev" title="Running from R" href="6_dandelion_running_from_R-10x_data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/dandelion_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../README.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#required-database">Required database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#external-softwares">External softwares</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#basic-requirements">Basic requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#acknowledgements">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#citation">Citation</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="singularity_preprocessing.html">Dandelion preprocessing with Singularity</a></li>
<li class="toctree-l1"><a class="reference internal" href="gamma_delta.html">Dandelion for TCR gamma/delta reannotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="1_dandelion_preprocessing-10x_data.html">Pre-processing (Re-annotation)</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_dandelion_filtering-10x_data.html">Filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="1b_dandelion_noreannotation-10x_data.html">Reading 10X Cell Ranger output directly</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_dandelion_findingclones-10x_data.html">BCR clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="1c_dandelion_scirpy.html">Interoperability with <code class="docutils literal notranslate"><span class="pre">scirpy</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="4_dandelion_visualization-10x_data.html">Visualizing BCR data</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_dandelion_diversity_and_mutation-10x_data.html">Calculating diversity and mutation</a></li>
<li class="toctree-l1"><a class="reference internal" href="0_dandelion_primer.html"><em>Dandelion</em> class</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_dandelion_running_from_R-10x_data.html">Running from R</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Analyzing TCR data</a></li>
</ul>
<p class="caption"><span class="caption-text">API:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">Preprocessing: <cite>pp</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.preprocessing.external">Preprocessing (external): <cite>pp.external</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.tools">Tools: <cite>tl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.plotting">Plotting: <cite>pl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.utilities">Utilities: <cite>utl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.Dandelion">Dandelion</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.logging">Logging</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">dandelion</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Analyzing TCR data</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/notebooks/7_dandelion_TCR_data_10x_data.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Analyzing-TCR-data">
<h1>Analyzing TCR data<a class="headerlink" href="#Analyzing-TCR-data" title="Permalink to this headline">¶</a></h1>
<img alt="dandelion_logo" src="../_images/dandelion_logo_illustration.png" />
<p>With <code class="docutils literal notranslate"><span class="pre">dandelion&gt;=1.3</span></code> onwards, there will be the ability to start analyzing 10x single-cell TCR data with the existing setup for both alpha-beta and gamma-delta TCR data formats. Currently, the alpha-beta and gamma-delta data sets have to be analyzed separately.</p>
<p>The workflow may be a bit buggy because it’s not optimised for TCR data; do let me know if you find any bugs and I can see if i can try to fix the issue(s).</p>
<p>We will download the various input formats of TCR files from 10x’s <a class="reference external" href="https://www.10xgenomics.com/resources/datasets">resource page</a> as part of this tutorial:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># bash</span>
mkdir -p dandelion_tutorial/sc5p_v2_hs_PBMC_10k<span class="p">;</span>
mkdir -p dandelion_tutorial/sc5p_v1p1_hs_melanoma_10k<span class="p">;</span>

<span class="nb">cd</span> dandelion_tutorial/sc5p_v2_hs_PBMC_10k<span class="p">;</span>
wget https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_filtered_feature_bc_matrix.h5<span class="p">;</span>
wget https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_t_airr_rearrangement.tsv<span class="p">;</span>
wget https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_t_filtered_contig_annotations.csv<span class="p">;</span>
wget https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_t_filtered_contig.fasta<span class="p">;</span>

<span class="nb">cd</span> ../sc5p_v1p1_hs_melanoma_10k<span class="p">;</span>
wget https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v1p1_hs_melanoma_10k/sc5p_v1p1_hs_melanoma_10k_filtered_feature_bc_matrix.h5<span class="p">;</span>
wget https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v1p1_hs_melanoma_10k/sc5p_v1p1_hs_melanoma_10k_t_airr_rearrangement.tsv<span class="p">;</span>
wget https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v1p1_hs_melanoma_10k/sc5p_v1p1_hs_melanoma_10k_t_filtered_contig_annotations.csv<span class="p">;</span>
wget https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v1p1_hs_melanoma_10k/sc5p_v1p1_hs_melanoma_10k_t_filtered_contig.fasta<span class="p">;</span>
</pre></div>
</div>
<p><strong>Import dandelion module</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">dandelion</span> <span class="k">as</span> <span class="nn">ddl</span>
<span class="c1"># change directory to somewhere more workable</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s1">&#39;/Users/kt16/Downloads/dandelion_tutorial/&#39;</span><span class="p">))</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_versions</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
dandelion==0.1.3.post2.dev62 pandas==1.2.3 numpy==1.20.1 matplotlib==3.3.4 networkx==2.5 scipy==1.6.1 skbio==0.5.6
</pre></div></div>
</div>
<p>I’m showing two examples for reading in the data: with or without reannotation.</p>
<p><strong>Read in AIRR format</strong></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># read in the airr_rearrangement.tsv file</span>
<span class="n">file1</span> <span class="o">=</span> <span class="s1">&#39;sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_t_airr_rearrangement.tsv&#39;</span>
<span class="n">file2</span> <span class="o">=</span> <span class="s1">&#39;sc5p_v1p1_hs_melanoma_10k/sc5p_v1p1_hs_melanoma_10k_t_airr_rearrangement.tsv&#39;</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj1</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">read_10x_airr</span><span class="p">(</span><span class="n">file1</span><span class="p">)</span>
<span class="n">vdj1</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 5245 and n_contigs = 10860
    data: &#39;cell_id&#39;, &#39;clone_id&#39;, &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;sequence_aa&#39;, &#39;productive&#39;, &#39;rev_comp&#39;, &#39;v_call&#39;, &#39;v_cigar&#39;, &#39;d_call&#39;, &#39;d_cigar&#39;, &#39;j_call&#39;, &#39;j_cigar&#39;, &#39;c_call&#39;, &#39;c_cigar&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;junction_length&#39;, &#39;junction_aa_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;consensus_count&#39;, &#39;duplicate_count&#39;, &#39;is_cell&#39;, &#39;locus&#39;
    metadata: &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_VDJ&#39;, &#39;v_call_VJ&#39;, &#39;j_call_VDJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;duplicate_count_VDJ_0&#39;, &#39;duplicate_count_VDJ_1&#39;, &#39;duplicate_count_VJ_0&#39;, &#39;duplicate_count_VJ_1&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;status&#39;, &#39;status_summary&#39;, &#39;productive&#39;, &#39;productive_summary&#39;, &#39;isotype&#39;, &#39;isotype_summary&#39;, &#39;vdj_status&#39;, &#39;vdj_status_summary&#39;, &#39;constant_status_summary&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj2</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">read_10x_airr</span><span class="p">(</span><span class="n">file2</span><span class="p">)</span>
<span class="n">vdj2</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 1540 and n_contigs = 2755
    data: &#39;cell_id&#39;, &#39;clone_id&#39;, &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;sequence_aa&#39;, &#39;productive&#39;, &#39;rev_comp&#39;, &#39;v_call&#39;, &#39;v_cigar&#39;, &#39;d_call&#39;, &#39;d_cigar&#39;, &#39;j_call&#39;, &#39;j_cigar&#39;, &#39;c_call&#39;, &#39;c_cigar&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;junction_length&#39;, &#39;junction_aa_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;consensus_count&#39;, &#39;duplicate_count&#39;, &#39;is_cell&#39;, &#39;locus&#39;
    metadata: &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_VDJ&#39;, &#39;v_call_VJ&#39;, &#39;j_call_VDJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;duplicate_count_VDJ_0&#39;, &#39;duplicate_count_VDJ_1&#39;, &#39;duplicate_count_VJ_0&#39;, &#39;duplicate_count_VJ_1&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;status&#39;, &#39;status_summary&#39;, &#39;productive&#39;, &#39;productive_summary&#39;, &#39;isotype&#39;, &#39;isotype_summary&#39;, &#39;vdj_status&#39;, &#39;vdj_status_summary&#39;, &#39;constant_status_summary&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># combine into a singular object</span>
<span class="c1"># let&#39;s add the sample_id to each cell barcode so that we don&#39;t end up overlapping later on</span>
<span class="n">sample_id</span> <span class="o">=</span> <span class="s1">&#39;sc5p_v2_hs_PBMC_10k&#39;</span>
<span class="n">vdj1</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;sample_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sample_id</span>
<span class="n">vdj1</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;cell_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample_id</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">vdj1</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;cell_id&#39;</span><span class="p">]]</span>
<span class="n">vdj1</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;sequence_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample_id</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">s</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">vdj1</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;sequence_id&#39;</span><span class="p">]]</span>

<span class="n">sample_id</span> <span class="o">=</span> <span class="s1">&#39;sc5p_v1p1_hs_melanoma_10k&#39;</span>
<span class="n">vdj2</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;sample_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sample_id</span>
<span class="n">vdj2</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;cell_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample_id</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">vdj2</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;cell_id&#39;</span><span class="p">]]</span>
<span class="n">vdj2</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;sequence_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample_id</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">s</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">vdj2</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;sequence_id&#39;</span><span class="p">]]</span>

<span class="c1"># combine into a singular object</span>
<span class="n">vdj</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">vdj1</span><span class="p">,</span> <span class="n">vdj2</span><span class="p">])</span>
<span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 6785 and n_contigs = 13615
    data: &#39;cell_id&#39;, &#39;clone_id&#39;, &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;sequence_aa&#39;, &#39;productive&#39;, &#39;rev_comp&#39;, &#39;v_call&#39;, &#39;v_cigar&#39;, &#39;d_call&#39;, &#39;d_cigar&#39;, &#39;j_call&#39;, &#39;j_cigar&#39;, &#39;c_call&#39;, &#39;c_cigar&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;junction_length&#39;, &#39;junction_aa_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;consensus_count&#39;, &#39;duplicate_count&#39;, &#39;is_cell&#39;, &#39;locus&#39;, &#39;sample_id&#39;
    metadata: &#39;sample_id&#39;, &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_VDJ&#39;, &#39;v_call_VJ&#39;, &#39;j_call_VDJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;duplicate_count_VDJ_0&#39;, &#39;duplicate_count_VDJ_1&#39;, &#39;duplicate_count_VJ_0&#39;, &#39;duplicate_count_VJ_1&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;status&#39;, &#39;status_summary&#39;, &#39;productive&#39;, &#39;productive_summary&#39;, &#39;isotype&#39;, &#39;isotype_summary&#39;, &#39;vdj_status&#39;, &#39;vdj_status_summary&#39;, &#39;constant_status_summary&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<p><strong>Read in with reannotation</strong></p>
<p>We specify the <code class="docutils literal notranslate"><span class="pre">filename_prefix</span></code> option because they have different prefixes that precedes <code class="docutils literal notranslate"><span class="pre">_contig.fasta</span></code> and <code class="docutils literal notranslate"><span class="pre">_contig_annotations.csv</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sc5p_v2_hs_PBMC_10k&#39;</span><span class="p">,</span> <span class="s1">&#39;sc5p_v1p1_hs_melanoma_10k&#39;</span><span class="p">]</span>
<span class="n">filename_prefixes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sc5p_v2_hs_PBMC_10k_t_filtered&#39;</span><span class="p">,</span> <span class="s1">&#39;sc5p_v1p1_hs_melanoma_10k_t_filtered&#39;</span><span class="p">]</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">format_fastas</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">prefix</span> <span class="o">=</span> <span class="n">samples</span><span class="p">,</span> <span class="n">filename_prefix</span> <span class="o">=</span> <span class="n">filename_prefixes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Formating fasta(s) : 100%|██████████| 2/2 [00:02&lt;00:00,  1.05s/it]
</pre></div></div>
</div>
<p>Make sure to toggle <code class="docutils literal notranslate"><span class="pre">loci</span> <span class="pre">=</span> <span class="pre">'tr'</span></code> for TCR data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">reannotate_genes</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">loci</span> <span class="o">=</span> <span class="s1">&#39;tr&#39;</span><span class="p">,</span> <span class="n">filename_prefix</span> <span class="o">=</span> <span class="n">filename_prefixes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Assigning genes : 100%|██████████| 2/2 [08:42&lt;00:00, 261.29s/it]
</pre></div></div>
</div>
<p>There’s no need to run the the rest of the preprocessing steps.</p>
<p>We’ll read in the reannotated files like as follow:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">tcr_files</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">:</span>
    <span class="n">file_location</span> <span class="o">=</span> <span class="n">sample</span> <span class="o">+</span><span class="s1">&#39;/dandelion/&#39;</span> <span class="o">+</span> <span class="n">sample</span> <span class="o">+</span> <span class="s1">&#39;_t_filtered_contig_igblast_db-pass.tsv&#39;</span>
    <span class="n">tcr_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_location</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">))</span>
<span class="n">tcr</span> <span class="o">=</span> <span class="n">tcr_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tcr_files</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
<span class="n">tcr</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">tcr</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sequence_id</th>
      <th>sequence</th>
      <th>rev_comp</th>
      <th>productive</th>
      <th>v_call</th>
      <th>d_call</th>
      <th>j_call</th>
      <th>sequence_alignment</th>
      <th>germline_alignment</th>
      <th>junction</th>
      <th>...</th>
      <th>cdr3</th>
      <th>cell_id</th>
      <th>c_call</th>
      <th>consensus_count</th>
      <th>umi_count</th>
      <th>v_call_10x</th>
      <th>d_call_10x</th>
      <th>j_call_10x</th>
      <th>junction_10x</th>
      <th>junction_10x_aa</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>sc5p_v2_hs_PBMC_10k_AAACCTGAGCGATAGC_contig_1</td>
      <td>CTGGAAGACCACCTGGGCTGTCATTGAGCTCTGGTGCCAGGAGGAA...</td>
      <td>F</td>
      <td>T</td>
      <td>TRAV23/DV6*02</td>
      <td>NaN</td>
      <td>TRAJ22*01</td>
      <td>CAGCAGCAGGTGAAACAAAGTCCTCAATCTTTGATAGTCCAGAAAG...</td>
      <td>CAGCAGCAGGTGAAACAAAGTCCTCAATCTTTGATAGTCCAGAAAG...</td>
      <td>TGTGCAGCAAGCAAGGGTTCTGCAAGGCAACTGACCTTT</td>
      <td>...</td>
      <td>GCAGCAAGCAAGGGTTCTGCAAGGCAACTGACC</td>
      <td>sc5p_v2_hs_PBMC_10k_AAACCTGAGCGATAGC</td>
      <td>TRAC</td>
      <td>2699</td>
      <td>11</td>
      <td>TRAV23/DV6</td>
      <td>NaN</td>
      <td>TRAJ22</td>
      <td>TGTGCAGCAAGCAAGGGTTCTGCAAGGCAACTGACCTTT</td>
      <td>CAASKGSARQLTF</td>
    </tr>
    <tr>
      <th>1</th>
      <td>sc5p_v2_hs_PBMC_10k_AAACCTGAGCGATAGC_contig_2</td>
      <td>GAGAGTCCTGCTCCCCTTTCATCAATGCACAGATACAGAAGACCCC...</td>
      <td>F</td>
      <td>T</td>
      <td>TRBV6-5*01</td>
      <td>TRBD2*01</td>
      <td>TRBJ2-3*01</td>
      <td>AATGCTGGTGTCACTCAGACCCCAAAATTCCAGGTCCTGAAGACAG...</td>
      <td>AATGCTGGTGTCACTCAGACCCCAAAATTCCAGGTCCTGAAGACAG...</td>
      <td>TGTGCCAGCAGTTACCGGGGGGGATCGGAAGATACGCAGTATTTT</td>
      <td>...</td>
      <td>GCCAGCAGTTACCGGGGGGGATCGGAAGATACGCAGTAT</td>
      <td>sc5p_v2_hs_PBMC_10k_AAACCTGAGCGATAGC</td>
      <td>TRBC2</td>
      <td>3202</td>
      <td>9</td>
      <td>TRBV6-5</td>
      <td>NaN</td>
      <td>TRBJ2-3</td>
      <td>TGTGCCAGCAGTTACCGGGGGGGATCGGAAGATACGCAGTATTTT</td>
      <td>CASSYRGGSEDTQYF</td>
    </tr>
    <tr>
      <th>2</th>
      <td>sc5p_v2_hs_PBMC_10k_AAACCTGAGTCACGCC_contig_1</td>
      <td>GTAGCTCGTTGATATCTGTGTGGATAGGGAGCTGTGACGAGGGCAA...</td>
      <td>F</td>
      <td>T</td>
      <td>TRAV8-6*02</td>
      <td>NaN</td>
      <td>TRAJ8*01</td>
      <td>GCCCAGTCTGTGACCCAGCTTGACAGCCAAGTCCCTGTCTTTGAAG...</td>
      <td>GCCCAGTCTGTGACCCAGCTTGACAGCCAAGTCCCTGTCTTTGAAG...</td>
      <td>TGTGCTGTGAGTGCGTTTTTTCAGAAACTTGTATTT</td>
      <td>...</td>
      <td>GCTGTGAGTGCGTTTTTTCAGAAACTTGTA</td>
      <td>sc5p_v2_hs_PBMC_10k_AAACCTGAGTCACGCC</td>
      <td>TRAC</td>
      <td>983</td>
      <td>9</td>
      <td>TRAV8-6</td>
      <td>NaN</td>
      <td>TRAJ8</td>
      <td>TGTGCTGTGAGTGCGTTTTTTCAGAAACTTGTATTT</td>
      <td>CAVSAFFQKLVF</td>
    </tr>
    <tr>
      <th>3</th>
      <td>sc5p_v2_hs_PBMC_10k_AAACCTGAGTCACGCC_contig_2</td>
      <td>CCTTTTCACCAATGCACAGACCCAGAGGACCCCTCCATCCTGCAGT...</td>
      <td>F</td>
      <td>T</td>
      <td>TRBV6-2*01,TRBV6-3*01</td>
      <td>TRBD1*01</td>
      <td>TRBJ2-6*01</td>
      <td>AATGCTGGTGTCACTCAGACCCCAAAATTCCGGGTCCTGAAGACAG...</td>
      <td>AATGCTGGTGTCACTCAGACCCCAAAATTCCGGGTCCTGAAGACAG...</td>
      <td>TGTGCCAGCAGTTATCTCCCCCGTAGACAGGACAGGGAATCCTCTG...</td>
      <td>...</td>
      <td>GCCAGCAGTTATCTCCCCCGTAGACAGGACAGGGAATCCTCTGGGG...</td>
      <td>sc5p_v2_hs_PBMC_10k_AAACCTGAGTCACGCC</td>
      <td>TRBC2</td>
      <td>4032</td>
      <td>12</td>
      <td>TRBV6-2</td>
      <td>NaN</td>
      <td>TRBJ2-6</td>
      <td>TGTGCCAGCAGTTATCTCCCCCGTAGACAGGACAGGGAATCCTCTG...</td>
      <td>CASSYLPRRQDRESSGANVLTF</td>
    </tr>
    <tr>
      <th>4</th>
      <td>sc5p_v2_hs_PBMC_10k_AAACCTGCACGTCAGC_contig_1</td>
      <td>CCCACATGAAGTGTCTACCTTCTGCAGACTCCAATGGCTCAGGAAC...</td>
      <td>F</td>
      <td>T</td>
      <td>TRAV1-2*01</td>
      <td>NaN</td>
      <td>TRAJ33*01</td>
      <td>GGACAAAACATTGACCAG...CCCACTGAGATGACAGCTACGGAAG...</td>
      <td>GGACAAAACATTGACCAG...CCCACTGAGATGACAGCTACGGAAG...</td>
      <td>TGTGCTGTCATGGATAGCAACTATCAGTTAATCTGG</td>
      <td>...</td>
      <td>GCTGTCATGGATAGCAACTATCAGTTAATC</td>
      <td>sc5p_v2_hs_PBMC_10k_AAACCTGCACGTCAGC</td>
      <td>TRAC</td>
      <td>881</td>
      <td>5</td>
      <td>TRAV1-2</td>
      <td>NaN</td>
      <td>TRAJ33</td>
      <td>TGTGCTGTCATGGATAGCAACTATCAGTTAATCTGG</td>
      <td>CAVMDSNYQLIW</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>15552</th>
      <td>sc5p_v1p1_hs_melanoma_10k_TTTGGTTTCACAGGCC_con...</td>
      <td>AATGGCTCAGGAACTGGGAATGCAGTGCCAGGCTCGTGGTATCCTG...</td>
      <td>F</td>
      <td>T</td>
      <td>TRAV1-2*01</td>
      <td>NaN</td>
      <td>TRAJ20*01</td>
      <td>GGACAAAACATTGACCAG...CCCACTGAGATGACAGCTACGGAAG...</td>
      <td>GGACAAAACATTGACCAG...CCCACTGAGATGACAGCTACGGAAG...</td>
      <td>TGTGCTGTGATGGGGGACTACAAGCTCAGCTTT</td>
      <td>...</td>
      <td>GCTGTGATGGGGGACTACAAGCTCAGC</td>
      <td>sc5p_v1p1_hs_melanoma_10k_TTTGGTTTCACAGGCC</td>
      <td>TRAC</td>
      <td>3174</td>
      <td>2</td>
      <td>TRAV1-2</td>
      <td>NaN</td>
      <td>TRAJ20</td>
      <td>TGTGCTGTGATGGGGGACTACAAGCTCAGCTTT</td>
      <td>CAVMGDYKLSF</td>
    </tr>
    <tr>
      <th>15553</th>
      <td>sc5p_v1p1_hs_melanoma_10k_TTTGGTTTCTTTACGT_con...</td>
      <td>TTCCTCTGCTCTGGCAGCAGATCTCCCAGAGGGAGCAGCCTGACCA...</td>
      <td>F</td>
      <td>T</td>
      <td>TRBV30*01</td>
      <td>TRBD2*01,TRBD2*02</td>
      <td>TRBJ2-5*01</td>
      <td>TCTCAGACTATTCATCAATGGCCAGCGACCCTGGTGCAGCCTGTGG...</td>
      <td>TCTCAGACTATTCATCAATGGCCAGCGACCCTGGTGCAGCCTGTGG...</td>
      <td>TGTGCCTGGAGTGAGCTAGCGGCCCAAGAGACCCAGTACTTC</td>
      <td>...</td>
      <td>GCCTGGAGTGAGCTAGCGGCCCAAGAGACCCAGTAC</td>
      <td>sc5p_v1p1_hs_melanoma_10k_TTTGGTTTCTTTACGT</td>
      <td>TRBC2</td>
      <td>11477</td>
      <td>8</td>
      <td>TRBV30</td>
      <td>NaN</td>
      <td>TRBJ2-5</td>
      <td>TGTGCCTGGAGTGAGCTAGCGGCCCAAGAGACCCAGTACTTC</td>
      <td>CAWSELAAQETQYF</td>
    </tr>
    <tr>
      <th>15554</th>
      <td>sc5p_v1p1_hs_melanoma_10k_TTTGGTTTCTTTACGT_con...</td>
      <td>GATCTTAATTGGGAAGAACAAGGATGACATCCATTCGAGCTGTATT...</td>
      <td>F</td>
      <td>F</td>
      <td>TRAV13-1*02</td>
      <td>NaN</td>
      <td>TRAJ43*01</td>
      <td>GGAGAGAATGTGGAGCAGCATCCTTCAACCCTGAGTGTCCAGGAGG...</td>
      <td>GGAGAGAATGTGGAGCAGCATCCTTCAACCCTGAGTGTCCAGGAGG...</td>
      <td>TGTGCAGCAAGTACAACCCGAAGGTTAGGCGGGGTGGGGTCAAAAA...</td>
      <td>...</td>
      <td>GCAGCAAGTACAACCCGAAGGTTAGGCGGGGTGGGGTCAAAAAAAT...</td>
      <td>sc5p_v1p1_hs_melanoma_10k_TTTGGTTTCTTTACGT</td>
      <td>TRAC</td>
      <td>7783</td>
      <td>6</td>
      <td>TRAV13-1</td>
      <td>NaN</td>
      <td>TRAJ43</td>
      <td>TGTGCAGCAAGTACAACCCGAAGGTTAGGC</td>
      <td>CAASTTRRLG</td>
    </tr>
    <tr>
      <th>15555</th>
      <td>sc5p_v1p1_hs_melanoma_10k_TTTGTCACATTTCACT_con...</td>
      <td>TCTGGGGATGTTCACAGAGGGCCTGGTCTGGAATATTCCACATCTG...</td>
      <td>F</td>
      <td>T</td>
      <td>TRBV12-4*01</td>
      <td>TRBD2*02</td>
      <td>TRBJ1-1*01</td>
      <td>GATGCTGGAGTTATCCAGTCACCCCGGCACGAGGTGACAGAGATGG...</td>
      <td>GATGCTGGAGTTATCCAGTCACCCCGGCACGAGGTGACAGAGATGG...</td>
      <td>TGTGCCAGCAGTTTAGGATGGGGAGACGGCACTGAAGCTTTCTTT</td>
      <td>...</td>
      <td>GCCAGCAGTTTAGGATGGGGAGACGGCACTGAAGCTTTC</td>
      <td>sc5p_v1p1_hs_melanoma_10k_TTTGTCACATTTCACT</td>
      <td>TRBC1</td>
      <td>14181</td>
      <td>7</td>
      <td>TRBV12-4</td>
      <td>NaN</td>
      <td>TRBJ1-1</td>
      <td>TGTGCCAGCAGTTTAGGATGGGGAGACGGCACTGAAGCTTTCTTT</td>
      <td>CASSLGWGDGTEAFF</td>
    </tr>
    <tr>
      <th>15556</th>
      <td>sc5p_v1p1_hs_melanoma_10k_TTTGTCACATTTCACT_con...</td>
      <td>AGATCAGAAGAGGAGGCTTCTCACCCTGCAGCAGGGACCTGTGAGC...</td>
      <td>F</td>
      <td>T</td>
      <td>TRAV38-2/DV8*01</td>
      <td>NaN</td>
      <td>TRAJ32*01,TRAJ32*02</td>
      <td>GCTCAGACAGTCACTCAGTCTCAACCAGAGATGTCTGTGCAGGAGG...</td>
      <td>GCTCAGACAGTCACTCAGTCTCAACCAGAGATGTCTGTGCAGGAGG...</td>
      <td>TGTGCTTATAGGAGCACCCAGATCCCCCAGCTCATCTTT</td>
      <td>...</td>
      <td>GCTTATAGGAGCACCCAGATCCCCCAGCTCATC</td>
      <td>sc5p_v1p1_hs_melanoma_10k_TTTGTCACATTTCACT</td>
      <td>TRAC</td>
      <td>8837</td>
      <td>4</td>
      <td>TRAV38-2/DV8</td>
      <td>NaN</td>
      <td>TRAJ32</td>
      <td>TGTGCTTATAGGAGCACCCAGATCCCCCAGCTCATCTTT</td>
      <td>CAYRSTQIPQLIF</td>
    </tr>
  </tbody>
</table>
<p>15557 rows × 57 columns</p>
</div></div>
</div>
<p>The reannotated file can be used with dandelion as per the BCR tutorial.</p>
<p>For the rest of the tutorial, I’m going to show how to proceed with 10x’s AIRR format file instead as there are some minor differences.</p>
<p><strong>Import modules for use with scanpy</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">functools</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">scipy.stats</span>
<span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
scanpy==1.6.0 anndata==0.7.6 umap==0.4.6 numpy==1.20.1 scipy==1.6.1 pandas==1.2.3 scikit-learn==0.24.1 statsmodels==0.12.1 python-igraph==0.8.3 leidenalg==0.8.3
</pre></div></div>
</div>
<p><strong>Import the transcriptome data</strong></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">gex_files</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;sc5p_v2_hs_PBMC_10k&#39;</span><span class="p">:</span><span class="s1">&#39;sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_filtered_feature_bc_matrix.h5&#39;</span><span class="p">,</span>
    <span class="s1">&#39;sc5p_v1p1_hs_melanoma_10k&#39;</span><span class="p">:</span><span class="s1">&#39;sc5p_v1p1_hs_melanoma_10k/sc5p_v1p1_hs_melanoma_10k_filtered_feature_bc_matrix.h5&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">gex_files</span><span class="p">:</span>
    <span class="n">adata_tmp</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_10x_h5</span><span class="p">(</span><span class="n">gex_files</span><span class="p">[</span><span class="n">f</span><span class="p">],</span> <span class="n">gex_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">adata_tmp</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;sample_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">f</span>
    <span class="n">adata_tmp</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">adata_tmp</span><span class="o">.</span><span class="n">obs_names</span><span class="p">]</span>
    <span class="n">adata_tmp</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
    <span class="n">adata_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">adata_tmp</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">adata_list</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Variable names are not unique. To make them unique, call `.var_names_make_unique`.
Variable names are not unique. To make them unique, call `.var_names_make_unique`.
Variable names are not unique. To make them unique, call `.var_names_make_unique`.
Variable names are not unique. To make them unique, call `.var_names_make_unique`.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 17275 × 36601
    obs: &#39;sample_id&#39;
</pre></div></div>
</div>
<p><strong>Run QC on the transcriptome data.</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">recipe_scanpy_qc</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 17275 × 36601
    obs: &#39;sample_id&#39;, &#39;scrublet_score&#39;, &#39;n_genes&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;gmm_pct_count_clusters&#39;, &#39;gmm_pct_count_clusters_keep&#39;, &#39;is_doublet&#39;, &#39;filter_rna&#39;
</pre></div></div>
</div>
<p><strong>Filtering TCR data.</strong></p>
<p>Note that I’m using the <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object as input rather than the pandas dataframe (yes both types of input will works. In fact, a file path to the .tsv will work too).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># The function will return both objects.</span>
<span class="n">vdj</span><span class="p">,</span> <span class="n">adata</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_contigs</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">locus</span> <span class="o">=</span> <span class="s1">&#39;tr-ab&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Scanning for poor quality/ambiguous contigs: 100%|██████████| 6911/6911 [03:22&lt;00:00, 34.13it/s]
Annotating in anndata obs slot : 100%|██████████| 17275/17275 [00:00&lt;00:00, 48137.08it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finishing up filtering
Initializing Dandelion object
</pre></div></div>
</div>
<p><strong>Check the output V(D)J table</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 4776 and n_contigs = 8801
    data: &#39;cell_id&#39;, &#39;clone_id&#39;, &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;sequence_aa&#39;, &#39;productive&#39;, &#39;rev_comp&#39;, &#39;v_call&#39;, &#39;v_cigar&#39;, &#39;d_call&#39;, &#39;d_cigar&#39;, &#39;j_call&#39;, &#39;j_cigar&#39;, &#39;c_call&#39;, &#39;c_cigar&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;junction_length&#39;, &#39;junction_aa_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;consensus_count&#39;, &#39;duplicate_count&#39;, &#39;is_cell&#39;, &#39;locus&#39;, &#39;sample_id&#39;, &#39;umi_count&#39;
    metadata: &#39;sample_id&#39;, &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_VDJ&#39;, &#39;v_call_VJ&#39;, &#39;j_call_VDJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;duplicate_count_VDJ_0&#39;, &#39;duplicate_count_VJ_0&#39;, &#39;duplicate_count_VJ_1&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;status&#39;, &#39;status_summary&#39;, &#39;productive&#39;, &#39;productive_summary&#39;, &#39;isotype&#39;, &#39;isotype_summary&#39;, &#39;vdj_status&#39;, &#39;vdj_status_summary&#39;, &#39;constant_status_summary&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<p><strong>Check the AnnData object as well</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 17275 × 36601
    obs: &#39;sample_id&#39;, &#39;scrublet_score&#39;, &#39;n_genes&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;gmm_pct_count_clusters&#39;, &#39;gmm_pct_count_clusters_keep&#39;, &#39;is_doublet&#39;, &#39;filter_rna&#39;, &#39;has_contig&#39;, &#39;filter_contig_quality&#39;, &#39;filter_contig_VDJ&#39;, &#39;filter_contig_VJ&#39;, &#39;contig_QC_pass&#39;, &#39;filter_contig&#39;
</pre></div></div>
</div>
<p><strong>The number of cells that actually has a matching BCR can be tabluated.</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pd</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;has_contig&#39;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;filter_contig&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>filter_contig</th>
      <th>False</th>
      <th>True</th>
    </tr>
    <tr>
      <th>has_contig</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>True</th>
      <td>4874</td>
      <td>1974</td>
    </tr>
    <tr>
      <th>No_contig</th>
      <td>6452</td>
      <td>3975</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pd</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;has_contig&#39;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;contig_QC_pass&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>contig_QC_pass</th>
      <th>False</th>
      <th>True</th>
      <th>No_contig</th>
    </tr>
    <tr>
      <th>has_contig</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>True</th>
      <td>2072</td>
      <td>4776</td>
      <td>0</td>
    </tr>
    <tr>
      <th>No_contig</th>
      <td>0</td>
      <td>0</td>
      <td>10427</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pd</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;contig_QC_pass&#39;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;filter_contig&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>filter_contig</th>
      <th>False</th>
      <th>True</th>
    </tr>
    <tr>
      <th>contig_QC_pass</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>False</th>
      <td>98</td>
      <td>1974</td>
    </tr>
    <tr>
      <th>True</th>
      <td>4776</td>
      <td>0</td>
    </tr>
    <tr>
      <th>No_contig</th>
      <td>6452</td>
      <td>3975</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p><strong>Now actually filter the AnnData object and run through a standard workflow starting by filtering genes and normalizing the data</strong></p>
<p>Because the ‘filtered’ <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object was returned as a filtered but otherwise unprocessed object, we still need to normalize and run through the usual process here. The following is just a standard scanpy workflow.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># filter genes</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="c1"># Normalize the counts</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="c1"># Logarithmize the data</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="c1"># Stash the normalised counts</span>
<span class="n">adata</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="n">adata</span>
</pre></div>
</div>
</div>
<p><strong>Identify highly-variable genes</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_mean</span><span class="o">=</span><span class="mf">0.0125</span><span class="p">,</span> <span class="n">max_mean</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">min_disp</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_7_dandelion_TCR_data_10x_data_35_0.png" src="../_images/notebooks_7_dandelion_TCR_data_10x_data_35_0.png" />
</div>
</div>
<p><strong>Filter the genes to only those marked as highly-variable</strong></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">highly_variable</span><span class="p">]</span>
</pre></div>
</div>
</div>
<p><strong>Regress out effects of total counts per cell and the percentage of mitochondrial genes expressed. Scale the data to unit variance.</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">regress_out</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Trying to set attribute `.obs` of view, copying.
... storing &#39;sample_id&#39; as categorical
</pre></div></div>
</div>
<p><strong>Run PCA</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca_variance_ratio</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_pcs</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_7_dandelion_TCR_data_10x_data_41_0.png" src="../_images/notebooks_7_dandelion_TCR_data_10x_data_41_0.png" />
</div>
</div>
<p><strong>Computing the neighborhood graph, umap and clusters</strong></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Computing the neighborhood graph</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="c1"># Embedding the neighborhood graph</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="c1"># Clustering the neighborhood graph</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p><strong>Visualizing the clusters and whether or not there’s a corresponding contig</strong>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;contig_QC_pass&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_7_dandelion_TCR_data_10x_data_45_0.png" src="../_images/notebooks_7_dandelion_TCR_data_10x_data_45_0.png" />
</div>
</div>
<p><strong>Visualizing some T cell genes.</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD3E&#39;</span><span class="p">,</span> <span class="s1">&#39;CD8B&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_7_dandelion_TCR_data_10x_data_47_0.png" src="../_images/notebooks_7_dandelion_TCR_data_10x_data_47_0.png" />
</div>
</div>
<p><strong>Find clones.</strong></p>
<p>Here we specify <code class="docutils literal notranslate"><span class="pre">identity</span> <span class="pre">=</span> <span class="pre">1</span></code> so only cells with identical CDR3 amino acid sequences are grouped into clones/clonotypes. Also specify <code class="docutils literal notranslate"><span class="pre">locus</span> <span class="pre">=</span> <span class="pre">'tr-ab</span></code> so that <code class="docutils literal notranslate"><span class="pre">dandelion</span></code> knows how to initialise the TCR data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">find_clones</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">identity</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">locus</span> <span class="o">=</span> <span class="s1">&#39;tr-ab&#39;</span><span class="p">)</span>
<span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Finding clones based on VDJ chains : 100%|██████████| 501/501 [00:00&lt;00:00, 1936.87it/s]
Refining clone assignment based on VJ chain pairing : 100%|██████████| 4776/4776 [00:00&lt;00:00, 194890.31it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 4776 and n_contigs = 8801
    data: &#39;cell_id&#39;, &#39;clone_id&#39;, &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;sequence_aa&#39;, &#39;productive&#39;, &#39;rev_comp&#39;, &#39;v_call&#39;, &#39;v_cigar&#39;, &#39;d_call&#39;, &#39;d_cigar&#39;, &#39;j_call&#39;, &#39;j_cigar&#39;, &#39;c_call&#39;, &#39;c_cigar&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;junction_length&#39;, &#39;junction_aa_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;consensus_count&#39;, &#39;duplicate_count&#39;, &#39;is_cell&#39;, &#39;locus&#39;, &#39;sample_id&#39;, &#39;umi_count&#39;
    metadata: &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;sample_id&#39;, &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_VDJ&#39;, &#39;v_call_VJ&#39;, &#39;j_call_VDJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;duplicate_count_VDJ_0&#39;, &#39;duplicate_count_VJ_0&#39;, &#39;duplicate_count_VJ_1&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;status&#39;, &#39;status_summary&#39;, &#39;productive&#39;, &#39;productive_summary&#39;, &#39;isotype&#39;, &#39;isotype_summary&#39;, &#39;vdj_status&#39;, &#39;vdj_status_summary&#39;, &#39;constant_status_summary&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<p><strong>Generate TCR network.</strong></p>
<p>The 10x-provided AIRR file is missing columns like <code class="docutils literal notranslate"><span class="pre">sequence_alignment</span></code> and <code class="docutils literal notranslate"><span class="pre">sequence_alignment_aa</span></code> so we will use the next best thing, which is <code class="docutils literal notranslate"><span class="pre">sequence</span></code> or <code class="docutils literal notranslate"><span class="pre">sequence_aa</span></code>. Note that these columns are not-gapped.</p>
<p>Specify <code class="docutils literal notranslate"><span class="pre">key</span> <span class="pre">=</span> <span class="pre">'sequence_aa'</span></code> to toggle this behavior. Can also try <code class="docutils literal notranslate"><span class="pre">junction</span></code> or <code class="docutils literal notranslate"><span class="pre">junction_aa</span></code> if just want to visualise the CDR3 linkage.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">generate_network</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;sequence_aa&#39;</span><span class="p">,</span> <span class="n">locus</span> <span class="o">=</span> <span class="s1">&#39;tr-ab&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Calculating distances... : 100%|██████████| 3/3 [02:33&lt;00:00, 51.33s/it]
Generating edge list : 100%|██████████| 26/26 [00:00&lt;00:00, 645.24it/s]
Linking edges : 100%|██████████| 4745/4745 [00:19&lt;00:00, 244.23it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
generating network layout
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 4776 and n_contigs = 8801
    data: &#39;cell_id&#39;, &#39;clone_id&#39;, &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;sequence_aa&#39;, &#39;productive&#39;, &#39;rev_comp&#39;, &#39;v_call&#39;, &#39;v_cigar&#39;, &#39;d_call&#39;, &#39;d_cigar&#39;, &#39;j_call&#39;, &#39;j_cigar&#39;, &#39;c_call&#39;, &#39;c_cigar&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;junction_length&#39;, &#39;junction_aa_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;consensus_count&#39;, &#39;duplicate_count&#39;, &#39;is_cell&#39;, &#39;locus&#39;, &#39;sample_id&#39;, &#39;umi_count&#39;
    metadata: &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;sample_id&#39;, &#39;locus_VDJ&#39;, &#39;locus_VJ&#39;, &#39;productive_VDJ&#39;, &#39;productive_VJ&#39;, &#39;v_call_VDJ&#39;, &#39;v_call_VJ&#39;, &#39;j_call_VDJ&#39;, &#39;j_call_VJ&#39;, &#39;c_call_VDJ&#39;, &#39;c_call_VJ&#39;, &#39;duplicate_count_VDJ_0&#39;, &#39;duplicate_count_VJ_0&#39;, &#39;duplicate_count_VJ_1&#39;, &#39;junction_aa_VDJ&#39;, &#39;junction_aa_VJ&#39;, &#39;status&#39;, &#39;status_summary&#39;, &#39;productive&#39;, &#39;productive_summary&#39;, &#39;isotype&#39;, &#39;isotype_summary&#39;, &#39;vdj_status&#39;, &#39;vdj_status_summary&#39;, &#39;constant_status_summary&#39;
    distance: &#39;VDJ_0&#39;, &#39;VJ_0&#39;, &#39;VJ_1&#39;
    edges: &#39;source&#39;, &#39;target&#39;, &#39;weight&#39;
    layout: layout for 4776 vertices, layout for 57 vertices
    graph: networkx graph of 4776 vertices, networkx graph of 57 vertices
</pre></div></div>
</div>
<p><strong>Plotting in ``scanpy``.</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">transfer</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">vdj</span><span class="p">)</span> <span class="c1"># this will include singletons. To show only expanded clones, specify expanded_only=True</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
converting matrices
Updating anndata slots
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">])</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                     <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample_id&#39;</span><span class="p">],</span>
                     <span class="n">edges_width</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;clone_id&#39; as categorical
... storing &#39;locus_VDJ&#39; as categorical
... storing &#39;locus_VJ&#39; as categorical
... storing &#39;productive_VDJ&#39; as categorical
... storing &#39;productive_VJ&#39; as categorical
... storing &#39;v_call_VDJ&#39; as categorical
... storing &#39;v_call_VJ&#39; as categorical
... storing &#39;j_call_VDJ&#39; as categorical
... storing &#39;j_call_VJ&#39; as categorical
... storing &#39;c_call_VDJ&#39; as categorical
... storing &#39;c_call_VJ&#39; as categorical
... storing &#39;junction_aa_VDJ&#39; as categorical
... storing &#39;junction_aa_VJ&#39; as categorical
... storing &#39;status&#39; as categorical
... storing &#39;status_summary&#39; as categorical
... storing &#39;productive&#39; as categorical
... storing &#39;productive_summary&#39; as categorical
... storing &#39;isotype&#39; as categorical
... storing &#39;isotype_summary&#39; as categorical
... storing &#39;vdj_status&#39; as categorical
... storing &#39;vdj_status_summary&#39; as categorical
... storing &#39;constant_status_summary&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_7_dandelion_TCR_data_10x_data_55_1.png" class="no-scaled-link" src="../_images/notebooks_7_dandelion_TCR_data_10x_data_55_1.png" style="width: 639px; height: 417px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">])</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                     <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">,</span> <span class="s1">&#39;vdj_status&#39;</span><span class="p">],</span>
                     <span class="n">ncols</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                     <span class="n">legend_fontoutline</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
                     <span class="n">edges_width</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_7_dandelion_TCR_data_10x_data_56_0.png" class="no-scaled-link" src="../_images/notebooks_7_dandelion_TCR_data_10x_data_56_0.png" style="width: 657px; height: 840px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">transfer</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">vdj</span><span class="p">,</span> <span class="n">expanded_only</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
converting matrices
Updating anndata slots
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">])</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                     <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample_id&#39;</span><span class="p">],</span>
                     <span class="n">edges_width</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                     <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_7_dandelion_TCR_data_10x_data_58_0.png" class="no-scaled-link" src="../_images/notebooks_7_dandelion_TCR_data_10x_data_58_0.png" style="width: 639px; height: 417px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">])</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                     <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">,</span> <span class="s1">&#39;vdj_status&#39;</span><span class="p">],</span>
                     <span class="n">ncols</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                     <span class="n">legend_fontoutline</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
                     <span class="n">edges_width</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                     <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_7_dandelion_TCR_data_10x_data_59_0.png" class="no-scaled-link" src="../_images/notebooks_7_dandelion_TCR_data_10x_data_59_0.png" style="width: 657px; height: 840px;" />
</div>
</div>
<p><strong>Finish.</strong></p>
<p>We can save the files.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;adata_tcr.h5ad&#39;</span><span class="p">,</span> <span class="n">compression</span> <span class="o">=</span> <span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span><span class="o">.</span><span class="n">write_h5</span><span class="p">(</span><span class="s1">&#39;dandelion_results_tcr.h5&#39;</span><span class="p">,</span> <span class="n">complib</span> <span class="o">=</span> <span class="s1">&#39;blosc:lz4&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="../api.html" class="btn btn-neutral float-right" title="Preprocessing: pp" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="6_dandelion_running_from_R-10x_data.html" class="btn btn-neutral float-left" title="Running from R" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, zktuong.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>