

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>dandelion.preprocessing.calculate_threshold &mdash; dandelion  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../_static/dandelion_logo.png"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="dandelion.preprocessing.create_germlines" href="dandelion.preprocessing.create_germlines.html" />
    <link rel="prev" title="dandelion.preprocessing.assign_isotypes" href="dandelion.preprocessing.assign_isotypes.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/dandelion_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../README.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#required-database">Required database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#external-softwares">External softwares</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#basic-requirements">Basic requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#acknowledgements">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#citation">Citation</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/singularity_preprocessing.html">Dandelion preprocessing with Singularity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/gamma_delta.html">Dandelion for TCR gamma/delta reannotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/1_dandelion_preprocessing-10x_data.html">Pre-processing (Re-annotation)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/2_dandelion_filtering-10x_data.html">Filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/1b_dandelion_noreannotation-10x_data.html">Reading 10X Cell Ranger output directly</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/3_dandelion_findingclones-10x_data.html">BCR clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/1c_dandelion_scirpy.html">Interoperability with <code class="docutils literal notranslate"><span class="pre">scirpy</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/4_dandelion_visualization-10x_data.html">Visualizing BCR data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/5_dandelion_diversity_and_mutation-10x_data.html">Calculating diversity and mutation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/0_dandelion_primer.html"><em>Dandelion</em> class</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/6_dandelion_running_from_R-10x_data.html">Running from R</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/7_dandelion_TCR_data_10x_data.html">Analyzing TCR data</a></li>
</ul>
<p class="caption"><span class="caption-text">API:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">Preprocessing: <cite>pp</cite></a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.assign_isotype.html">dandelion.preprocessing.assign_isotype</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.assign_isotypes.html">dandelion.preprocessing.assign_isotypes</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">dandelion.preprocessing.calculate_threshold</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.create_germlines.html">dandelion.preprocessing.create_germlines</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.filter_contigs.html">dandelion.preprocessing.filter_contigs</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.format_fasta.html">dandelion.preprocessing.format_fasta</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.format_fastas.html">dandelion.preprocessing.format_fastas</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.quantify_mutations.html">dandelion.preprocessing.quantify_mutations</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.reannotate_genes.html">dandelion.preprocessing.reannotate_genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.reassign_alleles.html">dandelion.preprocessing.reassign_alleles</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.preprocessing.external">Preprocessing (external): <cite>pp.external</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.tools">Tools: <cite>tl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.plotting">Plotting: <cite>pl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.utilities">Utilities: <cite>utl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.Dandelion">Dandelion</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.logging">Logging</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">dandelion</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../api.html">Preprocessing: <cite>pp</cite></a> &raquo;</li>
        
      <li>dandelion.preprocessing.calculate_threshold</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/modules/dandelion.preprocessing.calculate_threshold.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="dandelion-preprocessing-calculate-threshold">
<h1>dandelion.preprocessing.calculate_threshold<a class="headerlink" href="#dandelion-preprocessing-calculate-threshold" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt id="dandelion.preprocessing.calculate_threshold">
<code class="sig-prename descclassname"><span class="pre">dandelion.preprocessing.</span></code><code class="sig-name descname"><span class="pre">calculate_threshold</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'single-cell'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">manual_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">VJthenLen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">onlyHeavy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cross</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subsample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold_model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specificity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(4.5,</span> <span class="pre">2.5)</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dandelion/preprocessing/_preprocessing.html#calculate_threshold"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dandelion.preprocessing.calculate_threshold" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculating nearest neighbor distances for tuning clonal assignment with <cite>shazam</cite>.</p>
<p>&lt;<a class="reference external" href="https://shazam.readthedocs.io/en/stable/vignettes/DistToNearest-Vignette/">https://shazam.readthedocs.io/en/stable/vignettes/DistToNearest-Vignette/</a>&gt;`__.
Runs the following:</p>
<dl class="simple">
<dt>distToNearest</dt><dd><p>Get non-zero distance of every heavy chain (IGH) sequence (as defined by sequenceColumn) to its nearest sequence
in a partition of heavy chains sharing the same V gene, J gene, and junction length (VJL), or in a partition of
single cells with heavy chains sharing the same heavy chain VJL combination, or of single cells with heavy and
light chains sharing the same heavy chain VJL and light chain VJL combinations.</p>
</dd>
<dt>findThreshold</dt><dd><p>automtically determines an optimal threshold for clonal assignment of Ig sequences using a vector of nearest
neighbor distances. It provides two alternative methods using either a Gamma/Gaussian Mixture Model fit
(threshold_method=”gmm”) or kernel density fit (threshold_method=”density”).</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> (<em>Dandelion, DataFrame, str</em>) – <cite>Dandelion</cite> object, pandas <cite>DataFrame</cite> in changeo/airr format, or file path to changeo/airr file after clones
have been determined.</p></li>
<li><p><strong>mode</strong> (<em>Literal, str</em>) – accepts one of “heavy” or “single-cell”.
Refer to <a class="reference external" href="https://shazam.readthedocs.io/en/stable/vignettes/DistToNearest-Vignette">https://shazam.readthedocs.io/en/stable/vignettes/DistToNearest-Vignette</a>.</p></li>
<li><p><strong>manual_threshold</strong> (<em>float, optional</em>) – value to manually plot in histogram.</p></li>
<li><p><strong>VJthenLen</strong> (<em>bool</em>) – logical value specifying whether to perform partitioning as a 2-stage process.
If True, partitions are made first based on V and J gene, and then further split
based on junction lengths corresponding to sequenceColumn.
If False, perform partition as a 1-stage process during which V gene, J gene, and junction length
are used to create partitions simultaneously.
Defaults to False.</p></li>
<li><p><strong>onlyHeavy</strong> (<em>bool</em>) – use only the IGH (BCR) or TRB/TRD (TCR) sequences for grouping. Only applicable to single-cell mode.
See groupGenes for further details.</p></li>
<li><p><strong>model</strong> (<em>str, optional</em>) – underlying SHM model, which must be one of “ham”,”aa”,”hh_s1f”,”hh_s5f”,”mk_rs1nf”,”hs1f_compat”,”m1n_compat”.</p></li>
<li><p><strong>normalize_method</strong> (<em>str, optional</em>) – method of normalization. The default is “len”, which divides the distance by the length of the sequence group.
If “none” then no normalization if performed.</p></li>
<li><p><strong>threshold_method</strong> (<em>str, optional</em>) – string defining the method to use for determining the optimal threshold. One of “gmm” or “density”.</p></li>
<li><p><strong>edge</strong> (<em>float, optional</em>) – upper range as a fraction of the data density to rule initialization of Gaussian fit parameters.
Default value is 0.9 (or 90). Applies only when threshold_method=”density”.</p></li>
<li><p><strong>cross</strong> (<em>Sequence, optional</em>) – supplementary nearest neighbor distance vector output from distToNearest for initialization of the Gaussian fit
parameters. Applies only when method=”gmm”.</p></li>
<li><p><strong>subsample</strong> (<em>int, optional</em>) – maximum number of distances to subsample to before threshold detection.</p></li>
<li><p><strong>threshold_model</strong> (<em>str, optional</em>) – allows the user to choose among four possible combinations of fitting curves: “norm-norm”, “norm-gamma”,
“gamma-norm”, and “gamma-gamma”. Applies only when method=”gmm”.</p></li>
<li><p><strong>cutoff</strong> (<em>str, optional</em>) – method to use for threshold selection: the optimal threshold “optimal”, the intersection point of the two fitted
curves “intersect”, or a value defined by user for one of the sensitivity or specificity “user”. Applies only
when method=”gmm”.</p></li>
<li><p><strong>sensitivity</strong> (<em>float, optional</em>) – sensitivity required. Applies only when method=”gmm” and cutoff=”user”.</p></li>
<li><p><strong>specificity</strong> (<em>float, optional</em>) – specificity required. Applies only when method=”gmm” and cutoff=”user”.</p></li>
<li><p><strong>ncpu</strong> (<em>int, optional</em>) – number of cpus for parallelization. Default is all available cpus.</p></li>
<li><p><strong>plot</strong> (<em>bool</em>) – whether or not to return plot.</p></li>
<li><p><strong>plot_group</strong> (<em>str, optional</em>) – determines the fill color and facets.</p></li>
<li><p><strong>figsize</strong> (<em>Tuple[Union[int,float], Union[int,float]]</em>) – size of plot. Default is (4.5, 2.5).</p></li>
<li><p><strong>**kwargs</strong> – passed to shazam’s <a class="reference external" href="https://shazam.readthedocs.io/en/stable/topics/distToNearest/">distToNearest</a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Dandelion</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><cite>Dandelion</cite> object object with distance threshold value in <cite>.threshold</cite>.</p></li>
<li><p><em>If plot = True,plotnine plot showing histogram of length normalized ham model distance threshold.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="dandelion.preprocessing.create_germlines.html" class="btn btn-neutral float-right" title="dandelion.preprocessing.create_germlines" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="dandelion.preprocessing.assign_isotypes.html" class="btn btn-neutral float-left" title="dandelion.preprocessing.assign_isotypes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, zktuong.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>