

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>dandelion.preprocessing.filter_bcr &mdash; dandelion 0.0.26 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../_static/dandelion_logo.png"/>
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="dandelion.preprocessing.quantify_mutations" href="dandelion.preprocessing.quantify_mutations.html" />
    <link rel="prev" title="dandelion.preprocessing.create_germlines" href="dandelion.preprocessing.create_germlines.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/dandelion_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../README.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#required-database">Required database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#external-softwares">External softwares</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#basic-requirements">Basic requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#acknowledgements">Acknowledgements</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/1_dandelion_preprocessing-10x%20data.html">Pre-processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/2_dandelion_filtering-10x%20data.html">Filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/3_dandelion_findingclones_and_analysis-10x%20data.html">BCR clustering and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/5_dandelion_running_from_R.html">Running from R</a></li>
</ul>
<p class="caption"><span class="caption-text">API:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">Preprocessing: <cite>pp</cite></a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.format_fasta.html">dandelion.preprocessing.format_fasta</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.format_fastas.html">dandelion.preprocessing.format_fastas</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.assign_isotype.html">dandelion.preprocessing.assign_isotype</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.assign_isotypes.html">dandelion.preprocessing.assign_isotypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.reannotate_genes.html">dandelion.preprocessing.reannotate_genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.reassign_alleles.html">dandelion.preprocessing.reassign_alleles</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.create_germlines.html">dandelion.preprocessing.create_germlines</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">dandelion.preprocessing.filter_bcr</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.quantify_mutations.html">dandelion.preprocessing.quantify_mutations</a></li>
<li class="toctree-l2"><a class="reference internal" href="dandelion.preprocessing.calculate_threshold.html">dandelion.preprocessing.calculate_threshold</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.preprocessing.external">Preprocessing (external): <cite>pp.external</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.tools">Tools: <cite>tl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.plotting">Plotting: <cite>pl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#module-dandelion.utilities">Utilities: <cite>utl</cite></a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">dandelion</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../api.html">Preprocessing: <cite>pp</cite></a> &raquo;</li>
        
      <li>dandelion.preprocessing.filter_bcr</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/_generate/dandelion.preprocessing.filter_bcr.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="dandelion-preprocessing-filter-bcr">
<h1>dandelion.preprocessing.filter_bcr<a class="headerlink" href="#dandelion-preprocessing-filter-bcr" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt id="dandelion.preprocessing.filter_bcr">
<code class="sig-prename descclassname">dandelion.preprocessing.</code><code class="sig-name descname">filter_bcr</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">adata</span></em>, <em class="sig-param"><span class="n">filter_bcr</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">filter_rna</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">filter_poorqualitybcr</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">rescue_igh</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">umi_foldchange_cutoff</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">filter_lightchains</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">filter_missing</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">parallel</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">ncpu</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dandelion/preprocessing/_preprocessing.html#filter_bcr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dandelion.preprocessing.filter_bcr" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters doublets and poor quality cells and corresponding contigs based on provided V(D)J <cite>DataFrame</cite> and <cite>AnnData</cite> objects. Depends on a <cite>AnnData</cite>.obs slot populated with ‘filter_rna’ column.
If the aligned sequence is an exact match between contigs, the contigs will be merged into the one with the highest umi count, adding the summing the umi count of the duplicated contigs to duplicate_count column. After this check, if there are still multiple contigs, cells with multiple IGH contigs are filtered unless <cite>rescue_igh</cite> is True, where by the umi counts for each IGH contig will then be compared. The contig with the highest umi that is &gt; umi_foldchange_cutoff (default is empirically set at 5) from the lowest will be retained.
If there’s multiple contigs that survive the ‘rescue’, then all contigs will be filtered. The default behaviour is to also filter cells with multiple lightchains but this may sometimes be a true biological occurrence; toggling filter_lightchains to False will rescue the mutltiplet light chains.
Lastly, contigs with no corresponding cell barcode in the AnnData object is filtered if filter_missing is True. However, this may be useful to toggle to False if more contigs are preferred to be kept or for integrating with bulk reperotire seq data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataDrame, str</em>) – V(D)J airr/changeo data to filter. Can be pandas <cite>DataFrame</cite> object or file path as string.</p></li>
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object to filter.</p></li>
<li><p><strong>filter_bcr</strong> (<em>bool</em>) – If True, V(D)J <cite>DataFrame</cite> object returned will be filtered. Default is True.</p></li>
<li><p><strong>filter_rna</strong> (<em>bool</em>) – If True, <cite>AnnData</cite> object returned will be filtered. Default is True.</p></li>
<li><p><strong>filter_poorqualitybcr</strong> (<em>bool</em>) – If True, barcodes marked with poor quality BCR contigs will be filtered. Default is False; only relevant contigs are removed and RNA barcodes are kept.</p></li>
<li><p><strong>rescue_igh</strong> (<em>bool</em>) – If True, rescues IGH contigs with highest umi counts with a requirement that it passes the <cite>umi_foldchange_cutoff</cite> option. In addition, the sum of the all the heavy chain contigs must be greater than 3 umi or all contigs will be filtered. Default is True.</p></li>
<li><p><strong>umi_foldchange_cutoff</strong> (<em>int</em>) – related to minimum fold change required to rescue heavy chain contigs/barcode otherwise they will be marked as doublets. Default is empirically set at 2-fold.</p></li>
<li><p><strong>filter_lightchains</strong> (<em>bool</em>) – cells with multiple light chains will be marked to filter. Default is True.</p></li>
<li><p><strong>filter_missing</strong> (<em>bool</em>) – cells in V(D)J data not found in <cite>AnnData</cite> object will be marked to filter. Default is True. This may be useful for toggling to False if integrating with bulk data.</p></li>
<li><p><strong>parallel</strong> (<em>bool</em>) – whether or not to use parallelization. Default is True.</p></li>
<li><p><strong>ncpu</strong> (<em>int</em>) – number of cores to use if parallel is True. Default is all available - 1.</p></li>
<li><p><strong>save</strong> (<em>str, optional</em>) – Only used if a pandas dataframe or dandelion object is provided. Specifying will save the formatted vdj table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>V(D)J <cite>DataFrame</cite> object in airr/changeo format and <cite>AnnData</cite> object.</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="dandelion.preprocessing.quantify_mutations.html" class="btn btn-neutral float-right" title="dandelion.preprocessing.quantify_mutations" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="dandelion.preprocessing.create_germlines.html" class="btn btn-neutral float-left" title="dandelion.preprocessing.create_germlines" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, zktuong

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>