

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>dandelion.preprocessing.external._preprocessing &mdash; dandelion  documentation</title>
  

  
  <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../../../../_static/dandelion_logo.png"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../../" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/jquery.js"></script>
        <script src="../../../../_static/underscore.js"></script>
        <script src="../../../../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../../index.html">
          

          
            
            <img src="../../../../_static/dandelion_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../README.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../README.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../README.html#required-database">Required database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../README.html#external-softwares">External softwares</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../README.html#basic-requirements">Basic requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../README.html#acknowledgements">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../README.html#citation">Citation</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/singularity_preprocessing.html">Dandelion preprocessing with Singularity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/gamma_delta.html">Dandelion for TCR gamma/delta reannotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/1_dandelion_preprocessing-10x_data.html">Pre-processing (Re-annotation)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/2_dandelion_filtering-10x_data.html">Filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/1b_dandelion_noreannotation-10x_data.html">Reading 10X Cell Ranger output directly</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/3_dandelion_findingclones-10x_data.html">BCR clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/1c_dandelion_scirpy.html">Interoperability with <code class="docutils literal notranslate"><span class="pre">scirpy</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/4_dandelion_visualization-10x_data.html">Visualizing BCR data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/5_dandelion_diversity_and_mutation-10x_data.html">Calculating diversity and mutation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/0_dandelion_primer.html"><em>Dandelion</em> class</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/6_dandelion_running_from_R-10x_data.html">Running from R</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../notebooks/7_dandelion_TCR_data_10x_data.html">Analyzing TCR data</a></li>
</ul>
<p class="caption"><span class="caption-text">API:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html">Preprocessing: <cite>pp</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html#module-dandelion.preprocessing.external">Preprocessing (external): <cite>pp.external</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html#module-dandelion.tools">Tools: <cite>tl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html#module-dandelion.plotting">Plotting: <cite>pl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html#module-dandelion.utilities">Utilities: <cite>utl</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html#module-dandelion.Dandelion">Dandelion</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html#module-dandelion.logging">Logging</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">dandelion</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../../index.html">Module code</a> &raquo;</li>
        
      <li>dandelion.preprocessing.external._preprocessing</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for dandelion.preprocessing.external._preprocessing</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python</span>
<span class="c1"># @Author: kt16</span>
<span class="c1"># @Date:   2020-05-12 17:56:02</span>
<span class="c1"># @Last Modified by:   Kelvin</span>
<span class="c1"># @Last Modified time: 2021-07-16 22:10:43</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">subprocess</span> <span class="kn">import</span> <span class="n">run</span>
<span class="kn">from</span> <span class="nn">datetime</span> <span class="kn">import</span> <span class="n">timedelta</span>
<span class="kn">from</span> <span class="nn">anndata</span> <span class="kn">import</span> <span class="n">AnnData</span>
<span class="kn">from</span> <span class="nn">time</span> <span class="kn">import</span> <span class="n">time</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">OrderedDict</span>
<span class="kn">from</span> <span class="nn">sklearn</span> <span class="kn">import</span> <span class="n">mixture</span>
<span class="kn">from</span> <span class="nn">time</span> <span class="kn">import</span> <span class="n">time</span>
<span class="kn">from</span> <span class="nn">...utilities._utilities</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">scipy.stats</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">from</span> <span class="nn">os</span> <span class="kn">import</span> <span class="n">PathLike</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">Tuple</span>


<div class="viewcode-block" id="assigngenes_igblast"><a class="viewcode-back" href="../../../../modules/dandelion.preprocessing.external.assigngenes_igblast.html#dandelion.preprocessing.external.assigngenes_igblast">[docs]</a><span class="k">def</span> <span class="nf">assigngenes_igblast</span><span class="p">(</span><span class="n">fasta</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">],</span> <span class="n">igblast_db</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">org</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">loci</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;ig&#39;</span><span class="p">,</span> <span class="s1">&#39;tr&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;ig&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reannotate with IgBLASTn.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fasta : PathLike</span>
<span class="sd">        fasta file for reannotation.</span>
<span class="sd">    igblast_db : PathLike, optional</span>
<span class="sd">        path to igblast database.</span>
<span class="sd">    org : str</span>
<span class="sd">        organism for germline sequences.</span>
<span class="sd">    loci : str</span>
<span class="sd">        `ig` or `tr` mode for running igblastn.</span>
<span class="sd">    verbose : bool</span>
<span class="sd">        whether or not to print the command used in terminal. Default is False.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">env</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">igblast_db</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">igdb</span> <span class="o">=</span> <span class="n">env</span><span class="p">[</span><span class="s1">&#39;IGDATA&#39;</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">OSError</span><span class="p">(</span>
                <span class="s1">&#39;Environmental variable IGDATA must be set. Otherwise, please provide path to igblast database&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">env</span><span class="p">[</span><span class="s1">&#39;IGDATA&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">igblast_db</span>
        <span class="n">igdb</span> <span class="o">=</span> <span class="n">env</span><span class="p">[</span><span class="s1">&#39;IGDATA&#39;</span><span class="p">]</span>

    <span class="n">outfolder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">fasta</span><span class="p">))</span> <span class="o">+</span> <span class="s1">&#39;/tmp&#39;</span>
    <span class="n">informat_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;blast&#39;</span><span class="p">:</span> <span class="s1">&#39;_igblast.fmt7&#39;</span><span class="p">,</span> <span class="s1">&#39;airr&#39;</span><span class="p">:</span> <span class="s1">&#39;_igblast.tsv&#39;</span><span class="p">}</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">outfolder</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">outfolder</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">fileformat</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;blast&#39;</span><span class="p">,</span> <span class="s1">&#39;airr&#39;</span><span class="p">]:</span>
        <span class="n">outfile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">fasta</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span>
            <span class="s1">&#39;.fasta&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">informat_dict</span><span class="p">[</span><span class="n">fileformat</span><span class="p">]</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AssignGenes.py&#39;</span><span class="p">,</span> <span class="s1">&#39;igblast&#39;</span><span class="p">,</span>
               <span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="n">fasta</span><span class="p">,</span>
               <span class="s1">&#39;-b&#39;</span><span class="p">,</span> <span class="n">igdb</span><span class="p">,</span>
               <span class="s1">&#39;--organism&#39;</span><span class="p">,</span> <span class="n">org</span><span class="p">,</span>
               <span class="s1">&#39;--loci&#39;</span><span class="p">,</span> <span class="n">loci</span><span class="p">,</span>
               <span class="s1">&#39;--format&#39;</span><span class="p">,</span> <span class="n">fileformat</span><span class="p">,</span>
               <span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">/</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">outfolder</span><span class="p">,</span> <span class="n">outfile</span><span class="p">)</span>
               <span class="p">]</span>

        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Running command: </span><span class="si">%s</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmd</span><span class="p">)))</span>
        <span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">env</span><span class="o">=</span><span class="n">env</span><span class="p">)</span>  <span class="c1"># logs are printed to terminal</span></div>


<div class="viewcode-block" id="makedb_igblast"><a class="viewcode-back" href="../../../../modules/dandelion.preprocessing.external.makedb_igblast.html#dandelion.preprocessing.external.makedb_igblast">[docs]</a><span class="k">def</span> <span class="nf">makedb_igblast</span><span class="p">(</span><span class="n">fasta</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">],</span> <span class="n">igblast_output</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">germline</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">org</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">extended</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parses IgBLAST output to airr format.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fasta : PathLike</span>
<span class="sd">        fasta file use for reannotation.</span>
<span class="sd">    igblast_output : PathLike, optional</span>
<span class="sd">        igblast output file.</span>
<span class="sd">    germline : PathLike, optional</span>
<span class="sd">        path to germline database.    </span>
<span class="sd">    org : str</span>
<span class="sd">        organism of germline sequences.</span>
<span class="sd">    extended : bool</span>
<span class="sd">        whether or not to parse extended 10x annotations. Default is True.</span>
<span class="sd">    verbose : bool</span>
<span class="sd">        whether or not to print the command used in terminal. Default is False.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">env</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">gml</span> <span class="o">=</span> <span class="n">env</span><span class="p">[</span><span class="s1">&#39;GERMLINE&#39;</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">OSError</span><span class="p">(</span>
                <span class="s1">&#39;Environmental variable GERMLINE must be set. Otherwise, please provide path to folder containing germline fasta files.&#39;</span><span class="p">)</span>
        <span class="n">gml</span> <span class="o">=</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;imgt/&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;/vdj/&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">env</span><span class="p">[</span><span class="s1">&#39;GERMLINE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">germline</span>
        <span class="n">gml</span> <span class="o">=</span> <span class="n">germline</span>

    <span class="k">if</span> <span class="n">igblast_output</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">indir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">fasta</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;/tmp&#39;</span>
        <span class="n">infile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">fasta</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.fasta&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_igblast.fmt7&#39;</span>
        <span class="n">igbo</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">/</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">indir</span><span class="p">,</span> <span class="n">infile</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">igbo</span> <span class="o">=</span> <span class="n">igblast_output</span>

    <span class="n">cellranger_annotation</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">/</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span>
        <span class="n">fasta</span><span class="p">),</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">fasta</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;_annotations.csv&#39;</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">extended</span><span class="p">:</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;MakeDb.py&#39;</span><span class="p">,</span> <span class="s1">&#39;igblast&#39;</span><span class="p">,</span>
               <span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="n">igbo</span><span class="p">,</span>
               <span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="n">fasta</span><span class="p">,</span>
               <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
               <span class="s1">&#39;--10x&#39;</span><span class="p">,</span> <span class="n">cellranger_annotation</span><span class="p">,</span>
               <span class="s1">&#39;--extended&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;MakeDb.py&#39;</span><span class="p">,</span> <span class="s1">&#39;igblast&#39;</span><span class="p">,</span>
               <span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="n">igbo</span><span class="p">,</span>
               <span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="n">fasta</span><span class="p">,</span>
               <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
               <span class="s1">&#39;--10x&#39;</span><span class="p">,</span> <span class="n">cellranger_annotation</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Running command: </span><span class="si">%s</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmd</span><span class="p">)))</span>
    <span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">env</span><span class="o">=</span><span class="n">env</span><span class="p">)</span>  <span class="c1"># logs are printed to terminal</span></div>


<div class="viewcode-block" id="parsedb_heavy"><a class="viewcode-back" href="../../../../modules/dandelion.preprocessing.external.parsedb_heavy.html#dandelion.preprocessing.external.parsedb_heavy">[docs]</a><span class="k">def</span> <span class="nf">parsedb_heavy</span><span class="p">(</span><span class="n">db_file</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">],</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parses AIRR table (heavy chain contigs only).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    db_file : PathLike</span>
<span class="sd">        path to AIRR table.</span>
<span class="sd">    verbose : bool</span>
<span class="sd">        whether or not to print the command used in terminal. Default is False.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">outname</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">db_file</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.tsv&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_heavy&#39;</span>

    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ParseDb.py&#39;</span><span class="p">,</span> <span class="s1">&#39;select&#39;</span><span class="p">,</span>
           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
           <span class="s1">&#39;-f&#39;</span><span class="p">,</span> <span class="s1">&#39;locus&#39;</span><span class="p">,</span>
           <span class="s1">&#39;-u&#39;</span><span class="p">,</span> <span class="s1">&#39;IGH&#39;</span><span class="p">,</span>
           <span class="s1">&#39;--logic&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">,</span>
           <span class="s1">&#39;--regex&#39;</span><span class="p">,</span>
           <span class="s1">&#39;--outname&#39;</span><span class="p">,</span> <span class="n">outname</span>
           <span class="p">]</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Running command: </span><span class="si">%s</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmd</span><span class="p">)))</span>
    <span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>  <span class="c1"># logs are printed to terminal</span></div>


<div class="viewcode-block" id="parsedb_light"><a class="viewcode-back" href="../../../../modules/dandelion.preprocessing.external.parsedb_light.html#dandelion.preprocessing.external.parsedb_light">[docs]</a><span class="k">def</span> <span class="nf">parsedb_light</span><span class="p">(</span><span class="n">db_file</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">],</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parses AIRR table (light chain contigs only).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    db_file : PathLike</span>
<span class="sd">        path to AIRR table.</span>
<span class="sd">    verbose : bool</span>
<span class="sd">        whether or not to print the command used in terminal. Default is False.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">outname</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">db_file</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.tsv&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_light&#39;</span>

    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ParseDb.py&#39;</span><span class="p">,</span> <span class="s1">&#39;select&#39;</span><span class="p">,</span>
           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
           <span class="s1">&#39;-f&#39;</span><span class="p">,</span> <span class="s1">&#39;locus&#39;</span><span class="p">,</span>
           <span class="s1">&#39;-u&#39;</span><span class="p">,</span> <span class="s1">&#39;IG[LK]&#39;</span><span class="p">,</span>
           <span class="s1">&#39;--logic&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">,</span>
           <span class="s1">&#39;--regex&#39;</span><span class="p">,</span>
           <span class="s1">&#39;--outname&#39;</span><span class="p">,</span> <span class="n">outname</span>
           <span class="p">]</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Running command: </span><span class="si">%s</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmd</span><span class="p">)))</span>
    <span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>  <span class="c1"># logs are printed to terminal</span></div>


<div class="viewcode-block" id="creategermlines"><a class="viewcode-back" href="../../../../modules/dandelion.preprocessing.external.creategermlines.html#dandelion.preprocessing.external.creategermlines">[docs]</a><span class="k">def</span> <span class="nf">creategermlines</span><span class="p">(</span><span class="n">db_file</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">],</span> <span class="n">germtypes</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">germline</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">org</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">genotype_fasta</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">v_field</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;v_call&#39;</span><span class="p">,</span> <span class="s1">&#39;v_call_genotyped&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">cloned</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">mode</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;heavy&#39;</span><span class="p">,</span> <span class="s1">&#39;light&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Wrapper for CreateGermlines.py for reconstructing germline sequences,</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    db_file : PathLike</span>
<span class="sd">        path to AIRR table.</span>
<span class="sd">    germtypes : str, optional</span>
<span class="sd">        germline type for reconstuction.</span>
<span class="sd">    germline : PathLike, optional</span>
<span class="sd">        location to germline fasta files.</span>
<span class="sd">    org : str</span>
<span class="sd">        organism for germline sequences.</span>
<span class="sd">    genotype_fasta : PathLike, optional</span>
<span class="sd">        location to corrected v germine fasta file.</span>
<span class="sd">    v_field : str, optional</span>
<span class="sd">        name of column for v segment to perform reconstruction.</span>
<span class="sd">    cloned : bool</span>
<span class="sd">        whether or not to run with cloned option.</span>
<span class="sd">    mode : str, optional</span>
<span class="sd">        whether to run on heavy or light mode. If left as None, heavy and light will be run together.</span>
<span class="sd">    verbose : bool</span>
<span class="sd">        whether or not to print the command used in terminal. Default is False.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">env</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">gml</span> <span class="o">=</span> <span class="n">env</span><span class="p">[</span><span class="s1">&#39;GERMLINE&#39;</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">OSError</span><span class="p">(</span>
                <span class="s1">&#39;Environmental variable GERMLINE must be set. Otherwise, please provide path to folder containing germline fasta files.&#39;</span><span class="p">)</span>
        <span class="n">gml</span> <span class="o">=</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;imgt/&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;/vdj/&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">env</span><span class="p">[</span><span class="s1">&#39;GERMLINE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">germline</span>
        <span class="n">gml</span> <span class="o">=</span> <span class="n">germline</span>

    <span class="k">if</span> <span class="n">germtypes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">germ_type</span> <span class="o">=</span> <span class="s1">&#39;dmask&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">germ_type</span> <span class="o">=</span> <span class="n">germtypes</span>

    <span class="k">if</span> <span class="n">cloned</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s1">&#39;heavy&#39;</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;            Reconstructing heavy chain </span><span class="si">{}</span><span class="s1"> germline sequences with </span><span class="si">{}</span><span class="s1"> for each clone.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">germ_type</span><span class="p">,</span> <span class="n">v_field</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">genotype_fasta</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;--cloned&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span>
                           <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHD.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHJ.fasta&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;--cloned&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;--cloned&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">genotype_fasta</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span>
                           <span class="s1">&#39;_IGHD.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHJ.fasta&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;--cloned&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">genotype_fasta</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
        <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s1">&#39;light&#39;</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;            Reconstructing light chain </span><span class="si">{}</span><span class="s1"> germline sequences with </span><span class="si">{}</span><span class="s1"> for each clone.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">germ_type</span><span class="p">,</span> <span class="n">v_field</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                       <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                       <span class="s1">&#39;--cloned&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGKV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGKJ.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span>
                       <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLJ.fasta&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                       <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                       <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                       <span class="s1">&#39;--cloned&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
                       <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                       <span class="p">]</span>
        <span class="k">elif</span> <span class="n">mode</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;            Reconstructing </span><span class="si">{}</span><span class="s1"> germline sequences with </span><span class="si">{}</span><span class="s1"> for each clone.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">germ_type</span><span class="p">,</span> <span class="n">v_field</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">genotype_fasta</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;--cloned&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHD.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHJ.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span>
                           <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGKV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGKJ.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span>
                           <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLJ.fasta&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;--cloned&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;--cloned&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">genotype_fasta</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHD.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHJ.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span>
                           <span class="s1">&#39;_IGKV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGKJ.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span>
                           <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLJ.fasta&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;--cloned&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">genotype_fasta</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s1">&#39;heavy&#39;</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;            Reconstructing heavy chain </span><span class="si">{}</span><span class="s1"> germline sequences with </span><span class="si">{}</span><span class="s1">.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">germ_type</span><span class="p">,</span> <span class="n">v_field</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">genotype_fasta</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span>
                           <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHD.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHJ.fasta&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">genotype_fasta</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span>
                           <span class="s1">&#39;_IGHD.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHJ.fasta&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">genotype_fasta</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
        <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s1">&#39;light&#39;</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;            Reconstructing light chain </span><span class="si">{}</span><span class="s1"> germline sequences with </span><span class="si">{}</span><span class="s1">.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">germ_type</span><span class="p">,</span> <span class="n">v_field</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                       <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                       <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGKV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGKJ.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span>
                       <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLJ.fasta&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                       <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                       <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                       <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
                       <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                       <span class="p">]</span>
        <span class="k">elif</span> <span class="n">mode</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;            Reconstructing </span><span class="si">{}</span><span class="s1"> germline sequences with </span><span class="si">{}</span><span class="s1"> for each clone.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">germ_type</span><span class="p">,</span> <span class="n">v_field</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">genotype_fasta</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHD.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHJ.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span>
                           <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGKV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGKJ.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span>
                           <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLJ.fasta&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">genotype_fasta</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHD.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHJ.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span>
                           <span class="s1">&#39;_IGKV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGKJ.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span>
                           <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLV.fasta&#39;</span><span class="p">,</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGLJ.fasta&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CreateGermlines.py&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">db_file</span><span class="p">,</span>
                           <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">germ_type</span><span class="p">,</span>
                           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">genotype_fasta</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
                           <span class="s1">&#39;--vf&#39;</span><span class="p">,</span> <span class="n">v_field</span>
                           <span class="p">]</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Running command: </span><span class="si">%s</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmd</span><span class="p">)))</span>
    <span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">env</span><span class="o">=</span><span class="n">env</span><span class="p">)</span>  <span class="c1"># logs are printed to terminal</span></div>


<div class="viewcode-block" id="tigger_genotype"><a class="viewcode-back" href="../../../../modules/dandelion.preprocessing.external.tigger_genotype.html#dandelion.preprocessing.external.tigger_genotype">[docs]</a><span class="k">def</span> <span class="nf">tigger_genotype</span><span class="p">(</span><span class="n">data</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">],</span> <span class="n">v_germline</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">outdir</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">PathLike</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">org</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">fileformat</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;airr&#39;</span><span class="p">,</span> <span class="s1">&#39;changeo&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;airr&#39;</span><span class="p">,</span> <span class="n">novel_</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;YES&#39;</span><span class="p">,</span> <span class="s1">&#39;NO&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;YES&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reassign alleles with TIgGER in R.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data : PathLike</span>
<span class="sd">        vdj tabulated data, in Change-O (TAB) or AIRR (TSV) format.</span>
<span class="sd">    germline : PathLike, optional</span>
<span class="sd">        fasta file containing IMGT-gapped V segment reference germlines. Defaults to $GERMLINE.</span>
<span class="sd">    outdir : PathLike,  optional</span>
<span class="sd">        output directory. Will be created if it does not exist. Defaults to the current working directory.</span>
<span class="sd">    org : str</span>
<span class="sd">        organism for germline sequences.</span>
<span class="sd">    fileformat : str</span>
<span class="sd">        format for running tigger. Default is &#39;airr&#39;. Also accepts &#39;changeo&#39;.</span>
<span class="sd">    novel : str</span>
<span class="sd">        whether or not to run novel allele discovery. Default is &#39;YES&#39;.</span>
<span class="sd">    verbose : bool</span>
<span class="sd">        whether or not to print the command used in terminal. Default is False.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="p">()</span>
    <span class="n">env</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">v_germline</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">gml</span> <span class="o">=</span> <span class="n">env</span><span class="p">[</span><span class="s1">&#39;GERMLINE&#39;</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">OSError</span><span class="p">(</span><span class="s1">&#39;Environmental variable GERMLINE is not set. Please provide either the path to the folder containing the germline IGHV fasta file, or direct path to the germline IGHV fasta file.&#39;</span><span class="p">)</span>
        <span class="n">gml</span> <span class="o">=</span> <span class="n">gml</span> <span class="o">+</span> <span class="s1">&#39;imgt/&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;/vdj/imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHV.fasta&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">v_germline</span><span class="p">):</span>
            <span class="n">gml</span> <span class="o">=</span> <span class="n">v_germline</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;imgt_&#39;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s1">&#39;_IGHV.fasta&#39;</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">gml</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">OSError</span><span class="p">(</span><span class="s2">&quot;Input for germline is incorrect. Please rename IGHV germline file to &#39;</span><span class="si">{}</span><span class="s2">&#39;. Otherwise, please provide path to folder containing the germline IGHV fasta file, or direct path to the germline IGHV fasta file.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gml</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">v_germline</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.fasta&#39;</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">OSError</span><span class="p">(</span>
                    <span class="s1">&#39;Input for germline is incorrect </span><span class="si">{}</span><span class="s1">. Please provide path to folder containing the germline IGHV fasta file, or direct path to the germline IGHV fasta file.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">v_germline</span><span class="p">))</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">v_germline</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="s1">&#39;ighv&#39;</span> <span class="ow">in</span> <span class="n">v_germline</span><span class="o">.</span><span class="n">lower</span><span class="p">()):</span>
                <span class="n">gml</span> <span class="o">=</span> <span class="n">v_germline</span>

    <span class="k">if</span> <span class="n">outdir</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">out_dir</span> <span class="o">=</span> <span class="n">outdir</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">out_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;tigger-genotype.R&#39;</span><span class="p">,</span>
           <span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span>
           <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">gml</span><span class="p">,</span>
           <span class="s1">&#39;-n&#39;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.tsv&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
           <span class="s1">&#39;-N&#39;</span><span class="p">,</span> <span class="n">novel_</span><span class="p">,</span>
           <span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">,</span>
           <span class="s1">&#39;-f&#39;</span><span class="p">,</span> <span class="n">fileformat</span><span class="p">]</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;      Reassigning alleles&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Running command: </span><span class="si">%s</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmd</span><span class="p">)))</span>
    <span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">env</span><span class="o">=</span><span class="n">env</span><span class="p">)</span>  <span class="c1"># logs are printed to terminal</span>
    <span class="n">elapsed_time_secs</span> <span class="o">=</span> <span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span>
    <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;tigger-genotype execution took: </span><span class="si">%s</span><span class="s2"> secs (Wall clock time)</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">timedelta</span><span class="p">(</span>
        <span class="n">seconds</span><span class="o">=</span><span class="nb">round</span><span class="p">(</span><span class="n">elapsed_time_secs</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div>


<div class="viewcode-block" id="recipe_scanpy_qc"><a class="viewcode-back" href="../../../../modules/dandelion.preprocessing.external.recipe_scanpy_qc.html#dandelion.preprocessing.external.recipe_scanpy_qc">[docs]</a><span class="k">def</span> <span class="nf">recipe_scanpy_qc</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span> <span class="n">mito_startswith</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;MT&#39;</span><span class="p">,</span> <span class="n">max_genes</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2500</span><span class="p">,</span> <span class="n">min_genes</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">mito_cutoff</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">pval_cutoff</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">min_counts</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">max_counts</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">blacklist</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">vdj_pattern</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;^TR[AB][VDJ]|^IG[HKL][VDJC]&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">AnnData</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Recipe for running a standard scanpy QC workflow.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : AnnData</span>
<span class="sd">        annotated data matrix of shape n_obs × n_vars. Rows correspond to cells and columns to genes.</span>
<span class="sd">    mito_startswith : str</span>
<span class="sd">        string pattern used for searching mitochondrial genes.</span>
<span class="sd">    max_genes : int</span>
<span class="sd">        naximum number of genes expressed required for a cell to pass filtering. Default is 2500.</span>
<span class="sd">    min_genes : int</span>
<span class="sd">        minimum number of genes expressed  required for a cell to pass filtering. Default is 200.</span>
<span class="sd">    mito_cutoff : float</span>
<span class="sd">        maximum percentage mitochondrial content allowed for a cell to pass filtering. Default is 5.</span>
<span class="sd">    pval_cutoff : float</span>
<span class="sd">        maximum Benjamini-Hochberg corrected p value from doublet detection protocol allowed for a cell to pass filtering. Default is 0.05.</span>
<span class="sd">    min_counts : int, optional</span>
<span class="sd">        minimum number of counts required for a cell to pass filtering. Default is None.</span>
<span class="sd">    max_counts : int, optional</span>
<span class="sd">        maximum number of counts required for a cell to pass filtering. Default is None.</span>
<span class="sd">    blacklist : sequence, optional</span>
<span class="sd">        if provided, will exclude these genes from highly variable genes list.</span>
<span class="sd">    vdj_pattern : str</span>
<span class="sd">        string pattern for search VDJ genes to exclude from highly variable genes.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    `AnnData` of shape n_obs × n_vars where obs now contain filtering information. Rows correspond to cells and columns to genes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">_adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1"># run scrublet</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">scrublet</span> <span class="k">as</span> <span class="nn">scr</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span><span class="s1">&#39;Please install scrublet with pip install scrublet.&#39;</span><span class="p">)</span>

    <span class="n">scrub</span> <span class="o">=</span> <span class="n">scr</span><span class="o">.</span><span class="n">Scrublet</span><span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
    <span class="n">doublet_scores</span><span class="p">,</span> <span class="n">predicted_doublets</span> <span class="o">=</span> <span class="n">scrub</span><span class="o">.</span><span class="n">scrub_doublets</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;scrublet_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">doublet_scores</span>
    <span class="c1"># overcluster prep. run basic scanpy pipeline</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">_adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">_adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;mt&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">mito_startswith</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">_adata</span><span class="p">,</span> <span class="n">qc_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;mt&#39;</span><span class="p">],</span> <span class="n">percent_top</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">log1p</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># use a model-based method to determine the cut off for mitochondrial content</span>
    <span class="n">gmm</span> <span class="o">=</span> <span class="n">mixture</span><span class="o">.</span><span class="n">GaussianMixture</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">max_iter</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">covariance_type</span><span class="o">=</span><span class="s1">&#39;full&#39;</span><span class="p">)</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">,</span> <span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]]</span>
    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;gmm_pct_count_clusters&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gmm</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">)</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
    <span class="c1"># use a simple metric to workout which cluster is the one that contains lower mito content?</span>
    <span class="n">A1</span> <span class="o">=</span> <span class="n">_adata</span><span class="p">[</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;gmm_pct_count_clusters&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
    <span class="n">B1</span> <span class="o">=</span> <span class="n">_adata</span><span class="p">[</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;gmm_pct_count_clusters&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
    <span class="n">A2</span> <span class="o">=</span> <span class="n">_adata</span><span class="p">[</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;gmm_pct_count_clusters&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
    <span class="n">B2</span> <span class="o">=</span> <span class="n">_adata</span><span class="p">[</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;gmm_pct_count_clusters&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
    <span class="k">if</span> <span class="p">((</span><span class="n">A1</span> <span class="o">&gt;</span> <span class="n">B1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">A2</span> <span class="o">&lt;</span> <span class="n">B2</span><span class="p">)):</span>
        <span class="n">keepdict</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">keepdict</span> <span class="o">=</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;gmm_pct_count_clusters_keep&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">keepdict</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;gmm_pct_count_clusters&#39;</span><span class="p">]]</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">_adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">_adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span>
        <span class="n">_adata</span><span class="p">,</span> <span class="n">min_mean</span><span class="o">=</span><span class="mf">0.0125</span><span class="p">,</span> <span class="n">max_mean</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">min_disp</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">vdj_pattern</span><span class="p">,</span> <span class="n">i</span><span class="p">):</span>
            <span class="n">_adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="n">blacklist</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">blacklist</span><span class="p">:</span>
                <span class="n">_adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">_adata</span> <span class="o">=</span> <span class="n">_adata</span><span class="p">[:,</span> <span class="n">_adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;highly_variable&#39;</span><span class="p">]]</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">_adata</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">_adata</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">_adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
    <span class="c1"># overclustering proper - do basic clustering first, then cluster each cluster</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">_adata</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">clus</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">]))[</span><span class="mi">0</span><span class="p">]:</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">_adata</span><span class="p">,</span> <span class="n">restrict_to</span><span class="o">=</span><span class="p">(</span>
            <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="p">[</span><span class="n">clus</span><span class="p">]),</span> <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;leiden_R&#39;</span><span class="p">)</span>
    <span class="c1"># weird how the new anndata/scanpy is forcing this</span>
    <span class="k">for</span> <span class="n">clus</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">]))[</span><span class="mi">1</span><span class="p">:]:</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">_adata</span><span class="p">,</span> <span class="n">restrict_to</span><span class="o">=</span><span class="p">(</span>
            <span class="s1">&#39;leiden_R&#39;</span><span class="p">,</span> <span class="p">[</span><span class="n">clus</span><span class="p">]),</span> <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;leiden_R&#39;</span><span class="p">)</span>
    <span class="c1"># compute the cluster scores - the median of Scrublet scores per overclustered cluster</span>
    <span class="k">for</span> <span class="n">clus</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden_R&#39;</span><span class="p">]):</span>
        <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden_R&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">clus</span><span class="p">,</span> <span class="s1">&#39;scrublet_cluster_score&#39;</span><span class="p">]</span> <span class="o">=</span> \
            <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span>
                <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden_R&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">clus</span><span class="p">,</span> <span class="s1">&#39;scrublet_score&#39;</span><span class="p">])</span>
    <span class="c1"># now compute doublet p-values. figure out the median and mad (from above-median values) for the distribution</span>
    <span class="n">med</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;scrublet_cluster_score&#39;</span><span class="p">])</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;scrublet_cluster_score&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">med</span>
    <span class="n">mad</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;scrublet_cluster_score&#39;</span><span class="p">][</span><span class="n">mask</span><span class="p">]</span> <span class="o">-</span> <span class="n">med</span><span class="p">)</span>
    <span class="c1"># 1 sided test for catching outliers</span>
    <span class="n">pvals</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> \
        <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span>
            <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;scrublet_cluster_score&#39;</span><span class="p">],</span> <span class="n">loc</span><span class="o">=</span><span class="n">med</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="mf">1.4826</span> <span class="o">*</span> <span class="n">mad</span><span class="p">)</span>
    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;scrublet_score_bh_pval&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bh</span><span class="p">(</span><span class="n">pvals</span><span class="p">)</span>
    <span class="c1"># threshold the p-values to get doublet calls.</span>
    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;is_doublet&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;scrublet_score_bh_pval&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">pval_cutoff</span>    
    <span class="k">if</span> <span class="n">mito_cutoff</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">min_counts</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">max_counts</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;filter_rna&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">min_genes</span> <span class="o">&lt;</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_genes</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">mito_cutoff</span><span class="p">)</span> <span class="o">|</span> \
                <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">is_doublet</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">min_counts</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">max_counts</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;filter_rna&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">min_genes</span> <span class="o">&lt;</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_genes</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">min_counts</span> <span class="o">&lt;</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_counts</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">mito_cutoff</span><span class="p">)</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">is_doublet</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;filter_rna&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">min_genes</span> <span class="o">&lt;</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_genes</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">n</span> <span class="o">&lt;</span> <span class="n">min_counts</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">mito_cutoff</span><span class="p">)</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">is_doublet</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">max_counts</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;filter_rna&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">min_genes</span> <span class="o">&lt;</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_genes</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_counts</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">mito_cutoff</span><span class="p">)</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">is_doublet</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">min_counts</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">max_counts</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;filter_rna&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">min_genes</span> <span class="o">&lt;</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_genes</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                <span class="o">~</span><span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">gmm_pct_count_clusters_keep</span><span class="p">)</span> <span class="o">|</span> \
                <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">is_doublet</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">min_counts</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">max_counts</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;filter_rna&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">min_genes</span> <span class="o">&lt;</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_genes</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">min_counts</span> <span class="o">&lt;</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_counts</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="o">~</span><span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">gmm_pct_count_clusters_keep</span><span class="p">)</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">is_doublet</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;filter_rna&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">min_genes</span> <span class="o">&lt;</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_genes</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">n</span> <span class="o">&lt;</span> <span class="n">min_counts</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="o">~</span><span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">gmm_pct_count_clusters_keep</span><span class="p">)</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">is_doublet</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">max_counts</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;filter_rna&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">min_genes</span> <span class="o">&lt;</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_genes</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">n</span> <span class="o">&gt;</span> <span class="n">max_counts</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">]],</span> <span class="n">index</span><span class="o">=</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span> <span class="o">|</span> \
                        <span class="o">~</span><span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">gmm_pct_count_clusters_keep</span><span class="p">)</span> <span class="o">|</span> \
                        <span class="p">(</span><span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">is_doublet</span><span class="p">)</span>
    <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;is_doublet&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;is_doublet&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>

    <span class="c1"># removing columns that probably don&#39;t need anymore</span>
    <span class="k">if</span> <span class="n">mito_cutoff</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span>
            <span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden_R&#39;</span><span class="p">,</span> <span class="s1">&#39;scrublet_cluster_score&#39;</span><span class="p">,</span> <span class="s1">&#39;scrublet_score_bh_pval&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span>
            <span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden_R&#39;</span><span class="p">,</span> <span class="s1">&#39;scrublet_cluster_score&#39;</span><span class="p">,</span> <span class="s1">&#39;scrublet_score_bh_pval&#39;</span><span class="p">,</span> <span class="s1">&#39;gmm_pct_count_clusters&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, zktuong.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>