

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Preprocessing: pp &mdash; dandelion  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="_static/dandelion_logo.png"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="dandelion.preprocessing.assign_isotype" href="modules/dandelion.preprocessing.assign_isotype.html" />
    <link rel="prev" title="Analyzing TCR data" href="notebooks/7_dandelion_TCR_data_10x_data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/dandelion_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="README.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html#required-database">Required database</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html#external-softwares">External softwares</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html#basic-requirements">Basic requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html#acknowledgements">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html#citation">Citation</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="notebooks/singularity_preprocessing.html">Dandelion preprocessing with Singularity</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/gamma_delta.html">Dandelion for TCR gamma/delta reannotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/1_dandelion_preprocessing-10x_data.html">Pre-processing (Re-annotation)</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/2_dandelion_filtering-10x_data.html">Filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/1b_dandelion_noreannotation-10x_data.html">Reading 10X Cell Ranger output directly</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/3_dandelion_findingclones-10x_data.html">BCR clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/1c_dandelion_scirpy.html">Interoperability with <code class="docutils literal notranslate"><span class="pre">scirpy</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/4_dandelion_visualization-10x_data.html">Visualizing BCR data</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/5_dandelion_diversity_and_mutation-10x_data.html">Calculating diversity and mutation</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/0_dandelion_primer.html"><em>Dandelion</em> class</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/6_dandelion_running_from_R-10x_data.html">Running from R</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/7_dandelion_TCR_data_10x_data.html">Analyzing TCR data</a></li>
</ul>
<p class="caption"><span class="caption-text">API:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Preprocessing: <cite>pp</cite></a><ul>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.assign_isotype.html">dandelion.preprocessing.assign_isotype</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.assign_isotypes.html">dandelion.preprocessing.assign_isotypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.calculate_threshold.html">dandelion.preprocessing.calculate_threshold</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.create_germlines.html">dandelion.preprocessing.create_germlines</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.filter_contigs.html">dandelion.preprocessing.filter_contigs</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.format_fasta.html">dandelion.preprocessing.format_fasta</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.format_fastas.html">dandelion.preprocessing.format_fastas</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.quantify_mutations.html">dandelion.preprocessing.quantify_mutations</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.reannotate_genes.html">dandelion.preprocessing.reannotate_genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.reassign_alleles.html">dandelion.preprocessing.reassign_alleles</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#module-dandelion.preprocessing.external">Preprocessing (external): <cite>pp.external</cite></a><ul>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.external.assigngenes_igblast.html">dandelion.preprocessing.external.assigngenes_igblast</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.external.creategermlines.html">dandelion.preprocessing.external.creategermlines</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.external.makedb_igblast.html">dandelion.preprocessing.external.makedb_igblast</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.external.parsedb_heavy.html">dandelion.preprocessing.external.parsedb_heavy</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.external.parsedb_light.html">dandelion.preprocessing.external.parsedb_light</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.external.recipe_scanpy_qc.html">dandelion.preprocessing.external.recipe_scanpy_qc</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.preprocessing.external.tigger_genotype.html">dandelion.preprocessing.external.tigger_genotype</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#module-dandelion.tools">Tools: <cite>tl</cite></a><ul>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.tools.clone_centrality.html">dandelion.tools.clone_centrality</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.tools.clone_degree.html">dandelion.tools.clone_degree</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.tools.clone_diversity.html">dandelion.tools.clone_diversity</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.tools.clone_overlap.html">dandelion.tools.clone_overlap</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.tools.clone_rarefaction.html">dandelion.tools.clone_rarefaction</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.tools.clone_size.html">dandelion.tools.clone_size</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.tools.define_clones.html">dandelion.tools.define_clones</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.tools.extract_edge_weights.html">dandelion.tools.extract_edge_weights</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.tools.find_clones.html">dandelion.tools.find_clones</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.tools.generate_network.html">dandelion.tools.generate_network</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.tools.transfer.html">dandelion.tools.transfer</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#module-dandelion.plotting">Plotting: <cite>pl</cite></a><ul>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.plotting.barplot.html">dandelion.plotting.barplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.plotting.clone_network.html">dandelion.plotting.clone_network</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.plotting.clone_overlap.html">dandelion.plotting.clone_overlap</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.plotting.clone_rarefaction.html">dandelion.plotting.clone_rarefaction</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.plotting.spectratype.html">dandelion.plotting.spectratype</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.plotting.stackedbarplot.html">dandelion.plotting.stackedbarplot</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#module-dandelion.utilities">Utilities: <cite>utl</cite></a><ul>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.utilities.load_data.html">dandelion.utilities.load_data</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.utilities.makeblastdb.html">dandelion.utilities.makeblastdb</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.utilities.read_h5.html">dandelion.utilities.read_h5</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.utilities.read_pkl.html">dandelion.utilities.read_pkl</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.utilities.read_10x_airr.html">dandelion.utilities.read_10x_airr</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.utilities.read_10x_vdj.html">dandelion.utilities.read_10x_vdj</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.utilities.update_metadata.html">dandelion.utilities.update_metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.utilities.concat.html">dandelion.utilities.concat</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.utilities.to_scirpy.html">dandelion.utilities.to_scirpy</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.utilities.from_scirpy.html">dandelion.utilities.from_scirpy</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#module-dandelion.Dandelion">Dandelion</a><ul>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.Dandelion.copy.html">dandelion.Dandelion.copy</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.Dandelion.update_germline.html">dandelion.Dandelion.update_germline</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.Dandelion.write_h5.html">dandelion.Dandelion.write_h5</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.Dandelion.write_pkl.html">dandelion.Dandelion.write_pkl</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#module-dandelion.logging">Logging</a><ul>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.logging.print_header.html">dandelion.logging.print_header</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/dandelion.logging.print_versions.html">dandelion.logging.print_versions</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">dandelion</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Preprocessing: <cite>pp</cite></li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="module-dandelion.preprocessing">
<span id="preprocessing-pp"></span><h1>Preprocessing: <cite>pp</cite><a class="headerlink" href="#module-dandelion.preprocessing" title="Permalink to this headline">¶</a></h1>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.assign_isotype.html#dandelion.preprocessing.assign_isotype" title="dandelion.preprocessing.assign_isotype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">assign_isotype</span></code></a>(fasta[, fileformat, org, …])</p></td>
<td><p>Annotate contigs with constant region call using blastn.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.assign_isotypes.html#dandelion.preprocessing.assign_isotypes" title="dandelion.preprocessing.assign_isotypes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">assign_isotypes</span></code></a>(fastas[, fileformat, org, …])</p></td>
<td><p>Annotate contigs with constant region call using blastn.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.calculate_threshold.html#dandelion.preprocessing.calculate_threshold" title="dandelion.preprocessing.calculate_threshold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calculate_threshold</span></code></a>(self[, mode, …])</p></td>
<td><p>Calculating nearest neighbor distances for tuning clonal assignment with <cite>shazam</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.create_germlines.html#dandelion.preprocessing.create_germlines" title="dandelion.preprocessing.create_germlines"><code class="xref py py-obj docutils literal notranslate"><span class="pre">create_germlines</span></code></a>(self[, germline, org, …])</p></td>
<td><p>Run CreateGermlines.py to reconstruct the germline V(D)J sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.filter_contigs.html#dandelion.preprocessing.filter_contigs" title="dandelion.preprocessing.filter_contigs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_contigs</span></code></a>(data, adata[, filter_contig, …])</p></td>
<td><p>Filter doublets and poor quality cells and corresponding contigs based on provided V(D)J <cite>DataFrame</cite> and <cite>AnnData</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.format_fasta.html#dandelion.preprocessing.format_fasta" title="dandelion.preprocessing.format_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">format_fasta</span></code></a>(fasta[, prefix, suffix, sep, …])</p></td>
<td><p>Add prefix to the headers/contig ids in cellranger fasta and annotation file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.format_fastas.html#dandelion.preprocessing.format_fastas" title="dandelion.preprocessing.format_fastas"><code class="xref py py-obj docutils literal notranslate"><span class="pre">format_fastas</span></code></a>(fastas[, prefix, suffix, sep, …])</p></td>
<td><p>Add prefix to the headers/contig ids in cellranger fasta and annotation file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.quantify_mutations.html#dandelion.preprocessing.quantify_mutations" title="dandelion.preprocessing.quantify_mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">quantify_mutations</span></code></a>(self[, split_locus, …])</p></td>
<td><p>Run basic mutation load analysis.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.reannotate_genes.html#dandelion.preprocessing.reannotate_genes" title="dandelion.preprocessing.reannotate_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reannotate_genes</span></code></a>(data[, igblast_db, …])</p></td>
<td><p>Reannotate cellranger fasta files with igblastn and parses to airr/changeo data format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.reassign_alleles.html#dandelion.preprocessing.reassign_alleles" title="dandelion.preprocessing.reassign_alleles"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reassign_alleles</span></code></a>(data, combined_folder[, …])</p></td>
<td><p>Correct allele calls based on a personalized genotype using tigger-reassignAlleles.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="module-dandelion.preprocessing.external">
<span id="preprocessing-external-pp-external"></span><h1>Preprocessing (external): <cite>pp.external</cite><a class="headerlink" href="#module-dandelion.preprocessing.external" title="Permalink to this headline">¶</a></h1>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.external.assigngenes_igblast.html#dandelion.preprocessing.external.assigngenes_igblast" title="dandelion.preprocessing.external.assigngenes_igblast"><code class="xref py py-obj docutils literal notranslate"><span class="pre">assigngenes_igblast</span></code></a>(fasta[, igblast_db, …])</p></td>
<td><p>Reannotate with IgBLASTn.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.external.creategermlines.html#dandelion.preprocessing.external.creategermlines" title="dandelion.preprocessing.external.creategermlines"><code class="xref py py-obj docutils literal notranslate"><span class="pre">creategermlines</span></code></a>(db_file[, germtypes, …])</p></td>
<td><p>Wrapper for CreateGermlines.py for reconstructing germline sequences,</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.external.makedb_igblast.html#dandelion.preprocessing.external.makedb_igblast" title="dandelion.preprocessing.external.makedb_igblast"><code class="xref py py-obj docutils literal notranslate"><span class="pre">makedb_igblast</span></code></a>(fasta[, igblast_output, …])</p></td>
<td><p>Parses IgBLAST output to airr format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.external.parsedb_heavy.html#dandelion.preprocessing.external.parsedb_heavy" title="dandelion.preprocessing.external.parsedb_heavy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parsedb_heavy</span></code></a>(db_file[, verbose])</p></td>
<td><p>Parses AIRR table (heavy chain contigs only).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.external.parsedb_light.html#dandelion.preprocessing.external.parsedb_light" title="dandelion.preprocessing.external.parsedb_light"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parsedb_light</span></code></a>(db_file[, verbose])</p></td>
<td><p>Parses AIRR table (light chain contigs only).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.external.recipe_scanpy_qc.html#dandelion.preprocessing.external.recipe_scanpy_qc" title="dandelion.preprocessing.external.recipe_scanpy_qc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">recipe_scanpy_qc</span></code></a>(self[, mito_startswith, …])</p></td>
<td><p>Recipe for running a standard scanpy QC workflow.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.preprocessing.external.tigger_genotype.html#dandelion.preprocessing.external.tigger_genotype" title="dandelion.preprocessing.external.tigger_genotype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigger_genotype</span></code></a>(data[, v_germline, outdir, …])</p></td>
<td><p>Reassign alleles with TIgGER in R.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="module-dandelion.tools">
<span id="tools-tl"></span><h1>Tools: <cite>tl</cite><a class="headerlink" href="#module-dandelion.tools" title="Permalink to this headline">¶</a></h1>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.tools.clone_centrality.html#dandelion.tools.clone_centrality" title="dandelion.tools.clone_centrality"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clone_centrality</span></code></a>(self[, verbose])</p></td>
<td><p>Calculates node closeness centrality in BCR network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.tools.clone_degree.html#dandelion.tools.clone_degree" title="dandelion.tools.clone_degree"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clone_degree</span></code></a>(self[, weight, verbose])</p></td>
<td><p>Calculates node degree in BCR network.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.tools.clone_diversity.html#dandelion.tools.clone_diversity" title="dandelion.tools.clone_diversity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clone_diversity</span></code></a>(self, groupby[, method, …])</p></td>
<td><p>Compute B cell clones diversity : Gini indices, Chao1 estimates, or Shannon entropy.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.tools.clone_overlap.html#dandelion.tools.clone_overlap" title="dandelion.tools.clone_overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clone_overlap</span></code></a>(self, groupby, colorby[, …])</p></td>
<td><p>A function to tabulate clonal overlap for input as a circos-style plot.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.tools.clone_rarefaction.html#dandelion.tools.clone_rarefaction" title="dandelion.tools.clone_rarefaction"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clone_rarefaction</span></code></a>(self, groupby[, …])</p></td>
<td><p>Returns rarefaction predictions for cell numbers vs clone size.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.tools.clone_size.html#dandelion.tools.clone_size" title="dandelion.tools.clone_size"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clone_size</span></code></a>(self[, max_size, clone_key, …])</p></td>
<td><p>Quantifies size of clones</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.tools.define_clones.html#dandelion.tools.define_clones" title="dandelion.tools.define_clones"><code class="xref py py-obj docutils literal notranslate"><span class="pre">define_clones</span></code></a>(self[, dist, action, model, …])</p></td>
<td><p>Find clones using changeo’s <a class="reference external" href="https://changeo.readthedocs.io/en/stable/tools/DefineClones.html">DefineClones.py</a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.tools.extract_edge_weights.html#dandelion.tools.extract_edge_weights" title="dandelion.tools.extract_edge_weights"><code class="xref py py-obj docutils literal notranslate"><span class="pre">extract_edge_weights</span></code></a>(self[, expanded_only])</p></td>
<td><p>Retrieves edge weights (BCR levenshtein distance) from graph.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.tools.find_clones.html#dandelion.tools.find_clones" title="dandelion.tools.find_clones"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_clones</span></code></a>(self[, identity, key, locus, …])</p></td>
<td><p>Find clones based on heavy chain and light chain CDR3 junction hamming distance.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.tools.generate_network.html#dandelion.tools.generate_network" title="dandelion.tools.generate_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">generate_network</span></code></a>(self[, key, clone_key, …])</p></td>
<td><p>Generates a Levenshtein distance network based on full length VDJ sequence alignments for heavy and light chain(s).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.tools.transfer.html#dandelion.tools.transfer" title="dandelion.tools.transfer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">transfer</span></code></a>(self, dandelion[, expanded_only, …])</p></td>
<td><p>Transfer data in <cite>Dandelion</cite> slots to <cite>AnnData</cite> object, updating the <cite>.obs</cite>, <cite>.uns</cite>, <cite>.obsm</cite> and <a href="#id1"><span class="problematic" id="id2">`</span></a>.obsp`slots.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="module-dandelion.plotting">
<span id="plotting-pl"></span><h1>Plotting: <cite>pl</cite><a class="headerlink" href="#module-dandelion.plotting" title="Permalink to this headline">¶</a></h1>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.plotting.barplot.html#dandelion.plotting.barplot" title="dandelion.plotting.barplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">barplot</span></code></a>(self, color[, palette, figsize, …])</p></td>
<td><p>A barplot function to plot usage of V/J genes in the data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.plotting.clone_network.html#dandelion.plotting.clone_network" title="dandelion.plotting.clone_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clone_network</span></code></a>(adata[, basis, edges])</p></td>
<td><p>Using scanpy’s plotting module to plot the network.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.plotting.clone_overlap.html#dandelion.plotting.clone_overlap" title="dandelion.plotting.clone_overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clone_overlap</span></code></a>(self, groupby, colorby[, …])</p></td>
<td><p>A plot function to visualise clonal overlap as a circos-style plot.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.plotting.clone_rarefaction.html#dandelion.plotting.clone_rarefaction" title="dandelion.plotting.clone_rarefaction"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clone_rarefaction</span></code></a>(self, color[, clone_key, …])</p></td>
<td><p>Plots rarefaction curve for cell numbers vs clone size.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.plotting.spectratype.html#dandelion.plotting.spectratype" title="dandelion.plotting.spectratype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">spectratype</span></code></a>(self, color, groupby, locus[, …])</p></td>
<td><p>A spectratype function to plot usage of CDR3 length in the data split by groups.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.plotting.stackedbarplot.html#dandelion.plotting.stackedbarplot" title="dandelion.plotting.stackedbarplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">stackedbarplot</span></code></a>(self, color, groupby[, …])</p></td>
<td><p>A stackedbarplot function to plot usage of V/J genes in the data split by groups.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="module-dandelion.utilities">
<span id="utilities-utl"></span><h1>Utilities: <cite>utl</cite><a class="headerlink" href="#module-dandelion.utilities" title="Permalink to this headline">¶</a></h1>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.utilities.load_data.html#dandelion.utilities.load_data" title="dandelion.utilities.load_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_data</span></code></a>(obj)</p></td>
<td><p>Reads in or copy dataframe object and set sequence_id as index without dropping.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.utilities.makeblastdb.html#dandelion.utilities.makeblastdb" title="dandelion.utilities.makeblastdb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">makeblastdb</span></code></a>(ref)</p></td>
<td><p>Run makeblastdb on constant region fasta file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.utilities.read_h5.html#dandelion.utilities.read_h5" title="dandelion.utilities.read_h5"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_h5</span></code></a>([filename])</p></td>
<td><p>Read in and returns a <cite>Dandelion</cite> class from .h5 format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.utilities.read_pkl.html#dandelion.utilities.read_pkl" title="dandelion.utilities.read_pkl"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_pkl</span></code></a>([filename])</p></td>
<td><p>Read in and returns a <cite>Dandelion</cite> class saved using pickle format.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.utilities.read_10x_airr.html#dandelion.utilities.read_10x_airr" title="dandelion.utilities.read_10x_airr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_10x_airr</span></code></a>(file)</p></td>
<td><p>Read the 10x AIRR rearrangement .tsv directly and returns a <cite>Dandelion</cite> object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.utilities.read_10x_vdj.html#dandelion.utilities.read_10x_vdj" title="dandelion.utilities.read_10x_vdj"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_10x_vdj</span></code></a>(path[, filename_prefix, …])</p></td>
<td><p>A parser to read .csv and .json files directly from folder containing 10x cellranger-outouts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.utilities.update_metadata.html#dandelion.utilities.update_metadata" title="dandelion.utilities.update_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_metadata</span></code></a>(self[, retrieve, locus, …])</p></td>
<td><p>A <cite>Dandelion</cite> initialisation function to update and populate the <cite>.metadata</cite> slot.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.utilities.concat.html#dandelion.utilities.concat" title="dandelion.utilities.concat"><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code></a>(arrays[, check_unique])</p></td>
<td><p>Concatenate dataframe and return as <cite>Dandelion</cite> object.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.utilities.to_scirpy.html#dandelion.utilities.to_scirpy" title="dandelion.utilities.to_scirpy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_scirpy</span></code></a>(data[, transfer])</p></td>
<td><p>Convert a <cite>Dandelion</cite> object to scirpy’s format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.utilities.from_scirpy.html#dandelion.utilities.from_scirpy" title="dandelion.utilities.from_scirpy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_scirpy</span></code></a>(adata)</p></td>
<td><p>Read a <cite>scirpy</cite> initialized <cite>AnnData</cite> oject and returns a <cite>Dandelion</cite> object.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="module-dandelion.Dandelion">
<span id="dandelion"></span><h1>Dandelion<a class="headerlink" href="#module-dandelion.Dandelion" title="Permalink to this headline">¶</a></h1>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.Dandelion.copy.html#dandelion.Dandelion.copy" title="dandelion.Dandelion.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>Performs a deep copy of all slots in <cite>Dandelion</cite> class.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.Dandelion.update_germline.html#dandelion.Dandelion.update_germline" title="dandelion.Dandelion.update_germline"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_germline</span></code></a>([corrected, germline, org])</p></td>
<td><p>Update germline reference with corrected sequences and store in <cite>Dandelion</cite> object.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.Dandelion.write_h5.html#dandelion.Dandelion.write_h5" title="dandelion.Dandelion.write_h5"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_h5</span></code></a>([filename, complib, compression, …])</p></td>
<td><p>Writes a <cite>Dandelion</cite> class to .h5 format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.Dandelion.write_pkl.html#dandelion.Dandelion.write_pkl" title="dandelion.Dandelion.write_pkl"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_pkl</span></code></a>([filename])</p></td>
<td><p>Writes a <cite>Dandelion</cite> class to .pkl format.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="module-dandelion.logging">
<span id="logging"></span><h1>Logging<a class="headerlink" href="#module-dandelion.logging" title="Permalink to this headline">¶</a></h1>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="modules/dandelion.logging.print_header.html#dandelion.logging.print_header" title="dandelion.logging.print_header"><code class="xref py py-obj docutils literal notranslate"><span class="pre">print_header</span></code></a>([dependencies])</p></td>
<td><p>Versions that are essential for dandelion’s operation.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="modules/dandelion.logging.print_versions.html#dandelion.logging.print_versions" title="dandelion.logging.print_versions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">print_versions</span></code></a>([dependencies])</p></td>
<td><p>Versions that are essential for dandelion’s operation.</p></td>
</tr>
</tbody>
</table>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="modules/dandelion.preprocessing.assign_isotype.html" class="btn btn-neutral float-right" title="dandelion.preprocessing.assign_isotype" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="notebooks/7_dandelion_TCR_data_10x_data.html" class="btn btn-neutral float-left" title="Analyzing TCR data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, zktuong.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>