

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Overview &mdash; dandelion 0.0.26 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="_static/dandelion_logo.png"/>
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Pre-processing" href="notebooks/1_dandelion_preprocessing-10x%20data.html" />
    <link rel="prev" title="Welcome to dandelion’s documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/dandelion_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">dandelion:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="#required-database">Required database</a></li>
<li class="toctree-l1"><a class="reference internal" href="#external-softwares">External softwares</a></li>
<li class="toctree-l1"><a class="reference internal" href="#basic-requirements">Basic requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="#acknowledgements">Acknowledgements</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="notebooks/1_dandelion_preprocessing-10x%20data.html">Pre-processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/2_dandelion_filtering-10x%20data.html">Filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/3_dandelion_findingclones_and_analysis-10x%20data.html">BCR clustering and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/5_dandelion_running_from_R.html">Running from R</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">dandelion</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Overview</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/README.md.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<p><img alt="dandelion_logo" src="_images/dandelion_logo_illustration.png" /></p>
<p>Version = 0.0.26</p>
<p>Hi there! I have put together a python package for analyzing single cell BCR/V(D)J data from 10x Genomics 5’ solution! It streamlines the pre-processing, leveraging some tools from immcantation suite, and integrates with scanpy/anndata for single-cell BCR analysis. It also includes a couple of functions for visualization.</p>
<div class="section" id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<p><img alt="dandelion_overview" src="_images/dandelion_overview.png" /></p>
<p>Illustration of the <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> class slots</p>
<p><img alt="dandelion_overview" src="_images/dandelion_class.png" /></p>
<!-- <img src="https://github.com/zktuong/dandelion/blob/master/docs/notebooks/img/dandelion_class.png" alt="dandelion_class" width="471.0648" height="341.3572"> --><p>Please refer to the <a class="reference external" href="https://sc-dandelion.readthedocs.io/">documentation</a> or the notebooks <a class="reference external" href="https://nbviewer.jupyter.org/github/zktuong/dandelion/tree/devel/docs/notebooks/">here</a>:</p>
<p>The raw files for the examples can be downloaded from 10X’s Single Cell Immune Profiling datasets <a class="reference external" href="https://support.10xgenomics.com/single-cell-vdj/datasets">website</a>.</p>
</div>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>I would reccomend installing this in order:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># in bash/zsh terminal</span>
<span class="c1"># create a conda environment with specific modules</span>
conda create --name dandelion <span class="nv">python</span><span class="o">=</span><span class="m">3</span>.7 
conda activate dandelion
</pre></div>
</div>
<p>First, install <a class="reference external" href="https://scanpy.readthedocs.io/en/latest/installation.html">scanpy</a>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># these are required by dandelion</span>
conda install -c conda-forge distance joblib plotnine adjustText
conda install -c bioconda igblast blast <span class="c1"># if this doesn&#39;t work, download them manually (see below)</span>
conda install -c conda-forge <span class="s2">&quot;rpy2&gt;=3.3.2,&lt;3.3.5&quot;</span> <span class="c1"># to make compatible for R version 4</span>
<span class="c1"># or pip install &quot;rpy2&gt;=3.3.2,&lt;3.3.5&quot;</span>
<span class="c1"># If it fails because it&#39;s compiling using clang, first, work out where the path is to your gcc compiler (use brew to install gcc if needed):</span>
<span class="c1"># then run</span>
<span class="c1"># env CC=/path/to/location/of/bin/gcc-9 pip install &quot;rpy2&gt;=3.3.2,&lt;3.3.5&quot;</span>

<span class="c1"># Use pip to install the following with --no-cache-dir --upgrade if necessary</span>
<span class="c1"># and then lastly install this</span>
pip install sc-dandelion
<span class="c1"># or</span>
pip install git+https://github.com/zktuong/dandelion.git

<span class="c1"># for the development branch, run this:</span>
pip install git+https://github.com/zktuong/dandelion.git@devel
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">dandelion</span></code> also requires some R packages intalled.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># in R</span>
<span class="nf">install.packages</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;optparse&quot;</span><span class="p">,</span> <span class="s">&quot;alakazam&quot;</span><span class="p">,</span> <span class="s">&quot;tigger&quot;</span><span class="p">,</span> <span class="s">&quot;airr&quot;</span><span class="p">,</span> <span class="s">&quot;shazam&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p>or the following if using conda to manage R:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># in bash/zsh terminal</span>
conda install -c conda-forge r-optparse r-alakazam r-tigger r-airr r-shazam
</pre></div>
</div>
<p>The package should now be properly installed and when starting up jupyter notebook in the virtual environment, the kernel <code class="docutils literal notranslate"><span class="pre">python3</span></code> should work. Otherwise, you might need to add it manually:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># in bash/zsh terminal</span>
python -m ipykernel install --user --name dandelion --display-name <span class="s2">&quot;Python (dandelion)&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="required-database">
<h1>Required database<a class="headerlink" href="#required-database" title="Permalink to this headline">¶</a></h1>
<p>Last but not least, you will need to download the database folder in this repository and place them somewhere accessible. The igblast and germline database folders were originally downloaded with <a class="reference external" href="https://immcantation.readthedocs.io/en/4.1.0/docker/intro.html">immcantation’s</a> docker image (4.1.0). The blast database were downloaded from IMGT and manually curated. I have uploaded a copy of the required databases in a separate <a class="reference external" href="https://github.com/zktuong/databases_for_vdj">repository</a>(Last update: 28/11/2020). Once you’ve unpacked the folders, export the the path to the database folders as environmental variables in your <code class="docutils literal notranslate"><span class="pre">~/.bash_profile</span></code> or <code class="docutils literal notranslate"><span class="pre">~/.zshenv</span></code> like below. This will allow dandelion to access them easily. In the future, the databases will have to be updated accordingly.</p>
<p>So for example, if I unpack into <code class="docutils literal notranslate"><span class="pre">~/Documents</span></code></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># in bash/zsh terminal</span>
<span class="c1"># set up environmental variables in ~/.bash_profile</span>
<span class="nb">echo</span> <span class="s1">&#39;export GERMLINE=~/Documents/dandelion/database/germlines/&#39;</span> &gt;&gt; ~/.bash_profile <span class="c1"># or ~/.zshenv</span>
<span class="nb">echo</span> <span class="s1">&#39;export IGDATA=~/Documents/dandelion/database/igblast/&#39;</span> &gt;&gt; ~/.bash_profile <span class="c1"># or ~/.zshenv</span>
<span class="nb">echo</span> <span class="s1">&#39;export BLASTDB=~/Documents/dandelion/database/blast/&#39;</span> &gt;&gt; ~/.bash_profile <span class="c1"># or ~/.zshenv</span>
<span class="nb">source</span> ~/.bash_profile <span class="c1"># or ~/.zshenv</span>
</pre></div>
</div>
</div>
<div class="section" id="external-softwares">
<h1>External softwares<a class="headerlink" href="#external-softwares" title="Permalink to this headline">¶</a></h1>
<p>While blast and igblast executables are managed through conda, you can also download <a class="reference external" href="https://ftp.ncbi.nih.gov/blast/executables/igblast/release/LATEST/">igblast</a> and <a class="reference external" href="https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">blast+</a> manually, and store the softwares somewhere accessible. Just make sure to set the paths to them appropriately.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># in bash/zsh terminal</span>
<span class="c1"># unpack where relevant and export the path to the softwares, e.g. ~/Documents/</span>
<span class="nb">echo</span> <span class="s1">&#39;export PATH=~/Documents/software/bin:$PATH&#39;</span> &gt;&gt; ~/.bash_profile <span class="c1"># or ~/.zshenv</span>
<span class="nb">source</span> ~/.bash_profile <span class="c1"># or ~/.zshenv</span>
</pre></div>
</div>
</div>
<div class="section" id="basic-requirements">
<h1>Basic requirements<a class="headerlink" href="#basic-requirements" title="Permalink to this headline">¶</a></h1>
<p>Python packages</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># conda</span>
<span class="n">python</span><span class="o">&gt;=</span><span class="mf">3.7</span><span class="p">,</span><span class="o">&lt;</span><span class="mf">3.8</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>
<span class="n">numpy</span><span class="o">&gt;=</span><span class="mf">1.18</span><span class="o">.</span><span class="mi">4</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>
<span class="n">pandas</span><span class="o">&gt;=</span><span class="mf">1.0</span><span class="o">.</span><span class="mi">3</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>
<span class="n">distance</span><span class="o">&gt;=</span><span class="mf">0.1</span><span class="o">.</span><span class="mi">3</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>
<span class="n">joblib</span><span class="o">==</span><span class="mf">0.14</span><span class="o">.</span><span class="mi">1</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>
<span class="n">jupyter</span><span class="o">==</span><span class="mf">1.0</span><span class="o">.</span><span class="mi">0</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>
<span class="n">scikit</span><span class="o">-</span><span class="n">learn</span><span class="o">&gt;=</span><span class="mf">0.23</span><span class="o">.</span><span class="mi">0</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>
<span class="n">numba</span><span class="o">&gt;=</span><span class="mf">0.48</span><span class="o">.</span><span class="mi">0</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>
<span class="n">pytables</span><span class="o">==</span><span class="mf">3.6</span><span class="o">.</span><span class="mi">1</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>
<span class="n">seaborn</span><span class="o">&gt;=</span><span class="mf">0.10</span><span class="o">.</span><span class="mi">1</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>
<span class="n">leidenalg</span><span class="o">&gt;=</span><span class="mf">0.8</span><span class="o">.</span><span class="mi">0</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>
<span class="n">plotnine</span><span class="o">&gt;=</span><span class="mf">0.6</span><span class="o">.</span><span class="mi">0</span> <span class="p">(</span><span class="n">conda</span><span class="o">-</span><span class="n">forge</span><span class="p">)</span>

<span class="c1"># Other executables (through conda)</span>
<span class="n">blast</span><span class="o">&gt;=</span><span class="mf">2.10</span><span class="o">.</span><span class="mi">0</span> <span class="p">(</span><span class="n">bioconda</span><span class="p">)</span> <span class="c1"># depends on the database version as well</span>
<span class="n">igblast</span><span class="o">==</span><span class="mf">1.15</span><span class="o">.</span><span class="mi">0</span> <span class="p">(</span><span class="n">bioconda</span><span class="p">)</span>

<span class="c1"># pip</span>
<span class="n">anndata</span><span class="o">&gt;=</span><span class="mf">0.7</span><span class="o">.</span><span class="mi">1</span>
<span class="n">scanpy</span><span class="o">&gt;=</span><span class="mf">1.4</span><span class="o">.</span><span class="mi">6</span>
<span class="n">scrublet</span><span class="o">&gt;=</span><span class="mf">0.2</span><span class="o">.</span><span class="mi">1</span>
<span class="n">scikit</span><span class="o">-</span><span class="n">bio</span><span class="o">&gt;=</span><span class="mf">0.5</span><span class="o">.</span><span class="mi">6</span> 
<span class="n">changeo</span><span class="o">&gt;=</span><span class="mf">1.0</span><span class="o">.</span><span class="mi">0</span>
<span class="n">presto</span><span class="o">&gt;=</span><span class="mf">0.6</span><span class="o">.</span><span class="mi">0</span>
<span class="n">polyleven</span><span class="o">&gt;=</span><span class="mf">0.5</span>
<span class="n">networkx</span><span class="o">&gt;=</span><span class="mf">2.4</span>
<span class="n">rpy2</span><span class="o">&gt;=</span><span class="mf">3.3</span><span class="o">.</span><span class="mi">2</span><span class="p">,</span><span class="o">&lt;</span><span class="mf">3.3</span><span class="o">.</span><span class="mi">5</span>
</pre></div>
</div>
<p>R packages</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">alakazam_1.0.1</span>
<span class="n">tigger_1.0.0</span>
<span class="n">airr_1.2.0</span>
<span class="n">shazam_1.0.0</span>
<span class="n">ggplot2</span>
</pre></div>
</div>
</div>
<div class="section" id="acknowledgements">
<h1>Acknowledgements<a class="headerlink" href="#acknowledgements" title="Permalink to this headline">¶</a></h1>
<p>I would like to acknowledge the contributions from Dr. Ondrej Suschanek, Dr. Benjamin Stewart, Dr. Rachel Bashford-Rogers and Prof. Menna Clatworthy, who helped with the initial conception of the project and for all discussions.</p>
<p>I would also like to acknowledge Dr. Jongeun Park, Dr. Cecilia-Dominguez Conde, Dr. Hamish King, Dr. Krysztof Polanksi and Dr. Peng He with whom I have had very useful discussions. I would also like to thank my wife who helped name the package, because she thought the plots looked like a dandelion =D.</p>
<p>If there are any ideas, comments, suggestions, thing you would like to know more etc., please feel free to email me at kt16&#64;sanger.ac.uk or post in the issue tracker and I will get back to you.</p>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="notebooks/1_dandelion_preprocessing-10x%20data.html" class="btn btn-neutral float-right" title="Pre-processing" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to dandelion’s documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, zktuong

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>